<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>7&nbsp; Cell localisation between pairs of cell types – Spatial analysis playbook</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./04b-cell_localisation_parent.html" rel="next">
<link href="./03a-cell_annotation.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script><script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script><link rel="stylesheet" href="styles.css">
</head>
<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="./index.html" class="navbar-brand navbar-brand-logo">
    <img src="./images/USydLogo.svg" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Spatial analysis playbook</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./04a-cell_localisation.html">Cell localisation</a></li><li class="breadcrumb-item"><a href="./04a-cell_localisation.html"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Cell localisation between pairs of cell types</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Overview</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Datasets</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./getting_started.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Getting Started</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Processing</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01-processing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Cell segmentation and pre-processing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02-quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Quality Control</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Cell annotation</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03a-cell_annotation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Unsupervised clustering for cell annotation</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Cell localisation</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04a-cell_localisation.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Cell localisation between pairs of cell types</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04b-cell_localisation_parent.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Cell relationships relative to expected behaviour</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">Spatial domains</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05-cellular_niches.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Identifying spatial domains with unsupervised clustering</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true">
 <span class="menu-text">Marker expression</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06-changes_in_marker_expression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Changes in marker expression</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true">
 <span class="menu-text">Classification</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07-classification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Finding associations between clinical variables and spatial features</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" role="navigation" aria-expanded="true">
 <span class="menu-text">Case studies</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-8" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08-case_study1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Case Study: Head and Neck Squamous Cell Carcinoma (Ferguson et al., 2022)</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li>
<a href="#quantifying-cell-type-co-localisation-with-spicyr" id="toc-quantifying-cell-type-co-localisation-with-spicyr" class="nav-link active" data-scroll-target="#quantifying-cell-type-co-localisation-with-spicyr"><span class="header-section-number">7.1</span> Quantifying cell type co-localisation with spicyR</a>
  <ul class="collapse">
<li><a href="#linear-modelling" id="toc-linear-modelling" class="nav-link" data-scroll-target="#linear-modelling"><span class="header-section-number">7.1.1</span> Linear modelling</a></li>
  <li><a href="#test-for-changes-in-localisation-for-a-specific-pair-of-cells" id="toc-test-for-changes-in-localisation-for-a-specific-pair-of-cells" class="nav-link" data-scroll-target="#test-for-changes-in-localisation-for-a-specific-pair-of-cells"><span class="header-section-number">7.1.2</span> Test for changes in localisation for a specific pair of cells</a></li>
  <li><a href="#test-for-changes-in-localisation-for-all-pairwise-cell-combinations" id="toc-test-for-changes-in-localisation-for-all-pairwise-cell-combinations" class="nav-link" data-scroll-target="#test-for-changes-in-localisation-for-all-pairwise-cell-combinations"><span class="header-section-number">7.1.3</span> Test for changes in localisation for all pairwise cell combinations</a></li>
  <li><a href="#linear-modelling-for-custom-metrics" id="toc-linear-modelling-for-custom-metrics" class="nav-link" data-scroll-target="#linear-modelling-for-custom-metrics"><span class="header-section-number">7.1.4</span> Linear modelling for custom metrics</a></li>
  <li><a href="#mixed-effects-modelling" id="toc-mixed-effects-modelling" class="nav-link" data-scroll-target="#mixed-effects-modelling"><span class="header-section-number">7.1.5</span> Mixed effects modelling</a></li>
  <li><a href="#performing-survival-analysis" id="toc-performing-survival-analysis" class="nav-link" data-scroll-target="#performing-survival-analysis"><span class="header-section-number">7.1.6</span> Performing survival analysis</a></li>
  <li><a href="#accounting-for-tissue-inhomogeneity" id="toc-accounting-for-tissue-inhomogeneity" class="nav-link" data-scroll-target="#accounting-for-tissue-inhomogeneity"><span class="header-section-number">7.1.7</span> Accounting for tissue inhomogeneity</a></li>
  <li><a href="#adjusting-for-cell-count" id="toc-adjusting-for-cell-count" class="nav-link" data-scroll-target="#adjusting-for-cell-count"><span class="header-section-number">7.1.8</span> Adjusting for cell count</a></li>
  </ul>
</li>
  <li><a href="#sessioninfo" id="toc-sessioninfo" class="nav-link" data-scroll-target="#sessioninfo"><span class="header-section-number">7.2</span> sessionInfo</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./04a-cell_localisation.html">Cell localisation</a></li><li class="breadcrumb-item"><a href="./04a-cell_localisation.html"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Cell localisation between pairs of cell types</span></a></li></ol></nav><div class="quarto-title">
<h1 class="title">
<span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Cell localisation between pairs of cell types</span>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><p>Spatial co-localisation analysis helps uncover whether specific cell types are found near each other more (or less) often than expected by chance. This can reveal insights into immune infiltration, tumour-stroma interactions, or the organisation of functional niches.</p>
<p>Before drawing conclusions from co-localisation patterns, it’s important to recognise the potential confounding factors that can distort results. Tissue architecture is rarely random—cells are often organised into structured regions, influenced by development, disease states, or sampling. If this spatial inhomogeneity isn’t accounted for, standard co-localisation methods can misinterpret natural density variations as meaningful interactions. Similarly, differences in cell abundance across images—due to biology or technical variability—can bias association metrics. And finally, variation between samples or replicates must be considered to ensure that observed patterns are consistent and not driven by a few outliers. Any robust co-localisation analysis must address all three of these challenges: local spatial structure, uneven cell-type frequencies, and replicate-level variability.</p>
<p>In this section, we demonstrate how to use the <code>spicyR</code> package to detect co-localisation patterns while accounting for tissue inhomogeneity, image-to-image variability, and cell abundance—ensuring results that are both statistically sound and biologically meaningful.</p>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># load required libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ellispatrick.github.io/spicyR/">spicyR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/Statial">Statial</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment">SpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/SydneyBioX/SpatialDatasets">SpatialDatasets</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BodenmillerGroup/imcRtools">imcRtools</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/therneau/survival">survival</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/">tibble</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">treekoR</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pharmaverse/ggsurvfit">ggsurvfit</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># set parameters</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span><span class="va">use_mc</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">use_mc</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">nCores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html">detectCores</a></span><span class="op">(</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="va">nCores</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="op">}</span></span>
<span><span class="va">BPPARAM</span> <span class="op">&lt;-</span> <span class="fu">simpleSeg</span><span class="fu">:::</span><span class="fu"><a href="https://sydneybiox.github.io/simpleSeg/reference/generateBPParam.html">generateBPParam</a></span><span class="op">(</span><span class="va">nCores</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="quantifying-cell-type-co-localisation-with-spicyr" class="level2" data-number="7.1"><h2 data-number="7.1" class="anchored" data-anchor-id="quantifying-cell-type-co-localisation-with-spicyr">
<span class="header-section-number">7.1</span> Quantifying cell type co-localisation with spicyR</h2>
<p><a href="https://www.bioconductor.org/packages/release/bioc/html/spicyR.html">spicyR</a> provides a metric to quantify the degree of localisation or dispersion between two cell types. It then tests for changes in this co-localisation metric across different disease states or groups.</p>
<p><img src="images/spicyR_fig1.jpeg" align="center" style="height: 300px; border: 0px"></p>
<p>Here, we will use the <a href="./datasets.html">Keren 2018</a> dataset to demonstrate the use of spicyR. The data is stored as a <code>SpatialExperiment</code> object within the <code>SpatialDatasets</code> package and contains single-cell spatial data from 41 images for three types of breast cancer tumours (cold, compartmentalised, and mixed).</p>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">kerenSPE</span> <span class="op">&lt;-</span> <span class="fu">SpatialDatasets</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialDatasets/man/SpatialDatasets.html">spe_Keren_2018</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># remove any missing data in our outcome columns</span></span>
<span><span class="va">kerenSPE</span> <span class="op">=</span> <span class="va">kerenSPE</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/stats/complete.cases.html">complete.cases</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Censored"</span>, <span class="st">"Survival_days_capped*"</span>,</span>
<span>                                                           <span class="st">"tumour_type"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Eleven immune cell types were identified in this dataset by the original authors: (double negative CD3 T cells, CD4 T cells, B cells, monocytes, macrophages, CD8 T cells, neutrophils, natural killer cells, dendritic cells, regulatory T cells), 2 structural cell types (endothelial, mesenchymal), 2 tumour cell types (keratin+ tumour, tumour) and one unidentified category.</p>
<section id="linear-modelling" class="level3" data-number="7.1.1"><h3 data-number="7.1.1" class="anchored" data-anchor-id="linear-modelling">
<span class="header-section-number">7.1.1</span> Linear modelling</h3>
<p>We use the L-function to measure the degree of co-localisation between two cell types. The L-function is a statistical tool used to quantify spatial relationships between two cell types—specifically, whether they tend to be closer together (co-localised) or further apart (dispersed) than would be expected by chance.</p>
<p>For each cell of type A, we count how many cells of type B are found within a certain radius. This is done across all A–B cell pairs in the image, and for a range of radii (e.g., 10, 20, 30 pixels or microns). This gives us an observed pattern of how frequently B cells appear near A cells at different spatial scales.</p>
<p>We then compare this observed pattern to what we would expect if the B cells were distributed completely at random throughout the tissue—this is called the null model, usually based on a Poisson process. If there is no biological interaction between the two cell types, their spatial arrangement should look random.</p>
<p>If the observed number of B cells near A cells is consistently higher than expected under this null model, it suggests positive spatial association—that is, the two cell types are co-localised. If it’s consistently lower, it indicates negative spatial association, or spatial dispersion. The L-function transforms and scales these comparisons in a way that accounts for variance and edge effects, making the differences easier to interpret across images and datasets.</p>
</section><section id="test-for-changes-in-localisation-for-a-specific-pair-of-cells" class="level3" data-number="7.1.2"><h3 data-number="7.1.2" class="anchored" data-anchor-id="test-for-changes-in-localisation-for-a-specific-pair-of-cells">
<span class="header-section-number">7.1.2</span> Test for changes in localisation for a specific pair of cells</h3>
<p>Firstly, we can test whether one cell type tends to be more localised with another cell type in one condition compared to the other. This can be done using the <code>spicy</code> function, where we specify the <code>condition</code> parameter.</p>
<p>In this example, we want to see whether or not neutrophils (<code>to</code>) tend to be found around CD8 T cells (<code>from</code>) in compartmentalised tumours compared to cold tumours. Given that there are 3 conditions, we can specify the desired conditions by setting the order of our <code>condition</code> factor. <code>spicy</code> will choose the first level of the factor as the base condition and the second level as the comparison condition. <code>spicy</code> will also naturally coerce the <code>condition</code> column into a factor if it is not already a factor.</p>
<p>To draw meaningful conclusions from co-localisation patterns, it’s important to carefully choose the proximity radius <code>r</code> — the distance that defines when two cells are considered “near” each other. This radius should reflect the biological scale of interaction, such as the typical signaling range or physical cell contact. If the radius is too small, real associations might be missed due to slight spatial variability or measurement noise. On the other hand, if the radius is too large, the analysis may pick up false signals driven by overall regional cell density rather than genuine biological interactions. Selecting the right radius is therefore key to accurately capturing the spatial relationships of interest.</p>
<p>By default, <code>spicy</code> calculates the L-function over a range of radii. In the example below, we calculate the L-function for three values of <code>r</code>: 20 pixels, 50 pixels, and 100 pixels to capture both local and global relationships.</p>
<p>The column containing cell type annotations and image IDs can be specified using the <code>cellType</code> and <code>imageID</code> arguments respectively. By default, <code>spicy</code> uses the columns named <code>cellType</code> and <code>imageID</code>.</p>
<p>We obtain a <code>spicy</code> object which details the results of the modelling performed. The <code>topPairs</code> function can be used to obtain the associated coefficients and p-value.</p>
<div class="cell" data-time_it="true">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">spicyTestPair</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/spicy.html">spicy</a></span><span class="op">(</span></span>
<span>  <span class="va">kerenSPE</span>,</span>
<span>  r <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">50</span>, <span class="fl">100</span><span class="op">)</span>,</span>
<span>  condition <span class="op">=</span> <span class="st">"tumour_type"</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"CD8_T_cell"</span>,</span>
<span>  to <span class="op">=</span> <span class="st">"Neutrophils"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The `BPPARAM` argument of `spicy()` is deprecated as of spicyR 1.18.0.
ℹ Please use the `cores` argument instead.
ℹ The deprecated feature was likely used in the spicyR package.
  Please report the issue at &lt;https://github.com/SydneyBioX/spicyR/issues&gt;.</code></pre>
</div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/topPairs.html">topPairs</a></span><span class="op">(</span><span class="va">spicyTestPair</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                        intercept coefficient      p.value   adj.pvalue
CD8_T_cell__Neutrophils  -91.5703     94.5402 0.0006792206 0.0006792206
                              from          to
CD8_T_cell__Neutrophils CD8_T_cell Neutrophils</code></pre>
</div>
<div style="text-align: right;">
<em>Time for this code chunk to run with 5.5 cores: 3.5 s
</em>
</div>
<p></p>
</div>
<p>As the <code>coefficient</code> in <code>spicyTestPair</code> is positive, we find that neutrophils are significantly more likely to be found near CD8 T cells in the compartmentalised tumours group compared to the cold tumour group.</p>
<div class="callout callout-style-default callout-tip callout-titled" title="Choosing parameters for spicy">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Choosing parameters for spicy
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>How do we select an optimal value for <code>r</code>?`</strong></p>
<ul>
<li><p>The choice of <code>r</code> will depend on the degree of co-localistion we expect to see and the biological context. Choosing a small value of <code>r</code> is optimal for examining local spatial relationships, and larger values of <code>r</code> will reveal global spatial relationships.</p></li>
<li><p>When the degree of localistion is unknown, it is best to choose a range of radii to define the co-localisation statistic to capture both local and global relationships.</p></li>
</ul>
</div>
</div>
</section><section id="test-for-changes-in-localisation-for-all-pairwise-cell-combinations" class="level3" data-number="7.1.3"><h3 data-number="7.1.3" class="anchored" data-anchor-id="test-for-changes-in-localisation-for-all-pairwise-cell-combinations">
<span class="header-section-number">7.1.3</span> Test for changes in localisation for all pairwise cell combinations</h3>
<p>We can perform what we did above for all pairwise combinations of cell types by excluding the <code>from</code> and <code>to</code> parameters in <code>spicy</code>. Additional covariates can be added using the <code>covariates</code> argument.</p>
<div class="cell" data-time_it="true">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">spicyTest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/spicy.html">spicy</a></span><span class="op">(</span></span>
<span>  <span class="va">kerenSPE</span>,</span>
<span>  condition <span class="op">=</span> <span class="st">"tumour_type"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/topPairs.html">topPairs</a></span><span class="op">(</span><span class="va">spicyTest</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                         intercept coefficient      p.value   adj.pvalue
Macrophages__dn_T_CD3     64.26984   -57.90199 2.240072e-08 6.160198e-06
dn_T_CD3__Macrophages     62.44605   -55.84542 5.987026e-08 8.232161e-06
Macrophages__CD8_T_cell  105.28040   -98.30184 2.151824e-05 1.525685e-03
CD8_T_cell__Macrophages  103.94483   -96.91337 2.219178e-05 1.525685e-03
Macrophages__DC_or_Mono   74.32427   -60.87678 3.696109e-05 2.032860e-03
DC_or_Mono__Macrophages   72.47461   -59.14177 5.498388e-05 2.520095e-03
Neutrophils__dn_T_CD3    -60.45705    66.94703 1.268374e-04 4.534466e-03
dn_T_CD3__Neutrophils    -60.45215    67.47179 1.319117e-04 4.534466e-03
dn_T_CD3__dn_T_CD3       -60.82745    97.65301 2.359087e-04 7.208323e-03
Macrophages__Macrophages  70.42380   -40.78211 9.662288e-04 1.791030e-02
                                from          to
Macrophages__dn_T_CD3    Macrophages    dn_T_CD3
dn_T_CD3__Macrophages       dn_T_CD3 Macrophages
Macrophages__CD8_T_cell  Macrophages  CD8_T_cell
CD8_T_cell__Macrophages   CD8_T_cell Macrophages
Macrophages__DC_or_Mono  Macrophages  DC_or_Mono
DC_or_Mono__Macrophages   DC_or_Mono Macrophages
Neutrophils__dn_T_CD3    Neutrophils    dn_T_CD3
dn_T_CD3__Neutrophils       dn_T_CD3 Neutrophils
dn_T_CD3__dn_T_CD3          dn_T_CD3    dn_T_CD3
Macrophages__Macrophages Macrophages Macrophages</code></pre>
</div>
<div style="text-align: right;">
<em>Time for this code chunk to run with 5.5 cores: 24.67 s
</em>
</div>
<p></p>
</div>
<p>Again, we obtain a <code>spicy</code> object which outlines the result of the linear models performed for each pairwise combination of cell types.</p>
<p>We can also examine the L-function metrics of individual images by using the convenient <code>bind</code> function on our <code>spicyTest</code> results object.</p>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/bind.html">bind</a></span><span class="op">(</span><span class="va">spicyTest</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  imageID         condition Keratin_Tumour__Keratin_Tumour
1       1             mixed                      -2.300602
2       2             mixed                      -1.989699
3       3 compartmentalised                      11.373530
4       4 compartmentalised                      33.931133
5       5 compartmentalised                      17.922818
  dn_T_CD3__Keratin_Tumour B_cell__Keratin_Tumour
1                -5.298543             -20.827279
2               -16.020022               3.025815
3               -21.857447             -24.962913
4               -36.438476             -40.470221
5               -20.816783             -38.138076</code></pre>
</div>
</div>
<p>The results can be represented as a bubble plot using the <code>signifPlot</code> function.</p>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/signifPlot.html">signifPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">spicyTest</span>,</span>
<span>  breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3</span>, <span class="fl">3</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  marksToPlot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Macrophages"</span>, <span class="st">"DC_or_Mono"</span>, <span class="st">"dn_T_CD3"</span>, <span class="st">"Neutrophils"</span>,</span>
<span>                  <span class="st">"CD8_T_cell"</span>, <span class="st">"Keratin_Tumour"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="04a-cell_localisation_files/figure-html/04a-signifPlot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Here, we can observe that the most significant relationships occur between macrophages and double negative CD3 T cells, suggesting that the two cell types are far more dispersed in compartmentalised tumours compared to cold tumours. In general, it appears that our immune cell types become more dispersed in compartmentalised tumours compared to cold tumours.</p>
<p>To examine a specific cell type-cell type relationship in more detail, we can use <code>spicyBoxplot</code> and specify either <code>from = "Macrophages"</code> and <code>to = "dn_T_CD3"</code> or <code>rank = 1</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/spicyBoxPlot.html">spicyBoxPlot</a></span><span class="op">(</span>results <span class="op">=</span> <span class="va">spicyTest</span>, </span>
<span>             <span class="co"># from = "Macrophages",</span></span>
<span>             <span class="co"># to = "dn_T_CD3"</span></span>
<span>             rank <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="04a-cell_localisation_files/figure-html/04a-spicyBoxPlot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The boxplot confirms what we originally found in the bubble plot. Macrophages and double negative CD3 T cells are significantly more dispersed (lower L-function score) in compartmentalised tumours compared to cold tumours.</p>
</section><section id="linear-modelling-for-custom-metrics" class="level3" data-number="7.1.4"><h3 data-number="7.1.4" class="anchored" data-anchor-id="linear-modelling-for-custom-metrics">
<span class="header-section-number">7.1.4</span> Linear modelling for custom metrics</h3>
<p>spicyR can also be applied to custom distance or abundance metrics. A kNN interactions graph can be generated with the function <code>buildSpatialGraph</code> from the <code>imcRtools</code> package. This generates a <code>colPairs</code> object inside of the <code>SpatialExperiment</code> object.</p>
<p>spicyR provides the function <code>convPairs</code> for converting a <code>colPairs</code> object into an abundance matrix by calculating the average number of nearby cells types for every cell type for a given <code>k</code>. For example, if there exists on average 5 neutrophils for every macrophage in image 1, the column <code>Neutrophil__Macrophage</code> would have a value of 5 for image 1.</p>
<div class="cell">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">kerenSPE</span> <span class="op">&lt;-</span> <span class="fu">imcRtools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/imcRtools/man/buildSpatialGraph.html">buildSpatialGraph</a></span><span class="op">(</span><span class="va">kerenSPE</span>, </span>
<span>                                         img_id <span class="op">=</span> <span class="st">"imageID"</span>, </span>
<span>                                         type <span class="op">=</span> <span class="st">"knn"</span>, k <span class="op">=</span> <span class="fl">20</span>,</span>
<span>                                        coords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pairAbundances</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/convPairs.html">convPairs</a></span><span class="op">(</span><span class="va">kerenSPE</span>,</span>
<span>                  colPair <span class="op">=</span> <span class="st">"knn_interaction_graph"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">pairAbundances</span><span class="op">[</span><span class="st">"B_cell__B_cell"</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   B_cell__B_cell
1      12.7349608
10      0.2777778
11      0.0000000
12      1.3333333
13      1.2200957
14      0.0000000</code></pre>
</div>
</div>
<p>The custom distance or abundance metrics can then be included in the analysis with the <code>alternateResult</code> parameter.</p>
<div class="cell">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">spicyTestColPairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/spicy.html">spicy</a></span><span class="op">(</span></span>
<span>  <span class="va">kerenSPE</span>,</span>
<span>  condition <span class="op">=</span> <span class="st">"tumour_type"</span>,</span>
<span>  alternateResult <span class="op">=</span> <span class="va">pairAbundances</span>,</span>
<span>  weights <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/topPairs.html">topPairs</a></span><span class="op">(</span><span class="va">spicyTestColPairs</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                             intercept coefficient    p.value adj.pvalue
Unidentified__CD8_T_cell   0.151346975  0.60159494 0.01687494  0.7034781
Keratin_Tumour__CD8_T_cell 0.117283437 -0.09850513 0.01870063  0.7034781
NK__CD8_T_cell             0.113835243  0.53296190 0.02304383  0.7034781
B_cell__Tumour             0.016326531  0.23053781 0.02626946  0.7034781
Other_Immune__DC_or_Mono   0.144559342 -0.12839560 0.02895838  0.7034781
dn_T_CD3__Neutrophils      0.029731407  0.29756814 0.03260677  0.7034781
Tregs__CD4_T_cell          0.195913616  0.51002005 0.03536184  0.7034781
dn_T_CD3__NK               0.005405405  0.18354680 0.04525721  0.7034781
Other_Immune__NK           0.013333333  0.22116787 0.04660829  0.7034781
CD4_T_cell__DC             2.600122100 -1.36259931 0.04845418  0.7034781
                                     from          to
Unidentified__CD8_T_cell     Unidentified  CD8_T_cell
Keratin_Tumour__CD8_T_cell Keratin_Tumour  CD8_T_cell
NK__CD8_T_cell                         NK  CD8_T_cell
B_cell__Tumour                     B_cell      Tumour
Other_Immune__DC_or_Mono     Other_Immune  DC_or_Mono
dn_T_CD3__Neutrophils            dn_T_CD3 Neutrophils
Tregs__CD4_T_cell                   Tregs  CD4_T_cell
dn_T_CD3__NK                     dn_T_CD3          NK
Other_Immune__NK             Other_Immune          NK
CD4_T_cell__DC                 CD4_T_cell          DC</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/signifPlot.html">signifPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">spicyTestColPairs</span>,</span>
<span>  marksToPlot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Macrophages"</span>, <span class="st">"dn_T_CD3"</span>, <span class="st">"CD4_T_cell"</span>, </span>
<span>                  <span class="st">"B_cell"</span>, <span class="st">"DC_or_Mono"</span>, <span class="st">"Neutrophils"</span>, <span class="st">"CD8_T_cell"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="04a-cell_localisation_files/figure-html/04a-signifPlot for kNN-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Using abudance metric yields no cell type pairs which are significantly dispersed or localised in compartmentalised tumours compared to cold tumours.</p>
</section><section id="mixed-effects-modelling" class="level3" data-number="7.1.5"><h3 data-number="7.1.5" class="anchored" data-anchor-id="mixed-effects-modelling">
<span class="header-section-number">7.1.5</span> Mixed effects modelling</h3>
<p>spicyR supports mixed effects modelling when multiple images are obtained for each subject. In this case, <code>subject</code> is treated as a random effect and <code>condition</code> is treated as a fixed effect. To perform mixed effects modelling, we can specify the <code>subject</code> parameter in the <code>spicy</code> function.</p>
<p>To demonstrate spicyR’s functionality with mixed effects models, we will use the <a href="./datasets.html">Damond 2019</a> dataset.</p>
<div class="cell" data-time_it="true">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># load in data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"diabetesData"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># mixed effects modelling with spicy</span></span>
<span><span class="va">spicyMixedTest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/spicy.html">spicy</a></span><span class="op">(</span></span>
<span>  <span class="va">diabetesData</span>,</span>
<span>  condition <span class="op">=</span> <span class="st">"stage"</span>,</span>
<span>  subject <span class="op">=</span> <span class="st">"case"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div style="text-align: right;">
<em>Time for this code chunk to run with 5.5 cores: 116 s
</em>
</div>
<p></p>
</div>
<p>As before, we generate a <code>spicy</code> results object, and we can use <code>topPairs</code> to identify the most significant cell type pairs.</p>
<div class="cell">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/topPairs.html">topPairs</a></span><span class="op">(</span><span class="va">spicyMixedTest</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                         intercept coefficient      p.value adj.pvalue
beta__delta           6.081500e+01  -15.433244 0.0006616410 0.09171134
delta__beta           6.090722e+01  -15.276364 0.0007164949 0.09171134
B__Th                 2.553231e-15   10.480124 0.0127535339 0.42316692
delta__delta          7.021912e+01  -16.358092 0.0155005315 0.42316692
Th__B                -3.418098e-15   10.003451 0.0173028442 0.42316692
B__unknown            4.199775e-16    4.584274 0.0182158635 0.42316692
otherimmune__naiveTc -3.179087e+00   11.944638 0.0200373935 0.42316692
unknown__macrophage   4.339584e+00   -5.274671 0.0222620877 0.42316692
unknown__B            8.620167e-16    4.680750 0.0244004305 0.42316692
macrophage__unknown   4.305429e+00   -4.886438 0.0249693456 0.42316692
                            from         to
beta__delta                 beta      delta
delta__beta                delta       beta
B__Th                          B         Th
delta__delta               delta      delta
Th__B                         Th          B
B__unknown                     B    unknown
otherimmune__naiveTc otherimmune    naiveTc
unknown__macrophage      unknown macrophage
unknown__B               unknown          B
macrophage__unknown   macrophage    unknown</code></pre>
</div>
</div>
<p>We can use <code>signifPlot</code> to visualise the results.</p>
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/signifPlot.html">signifPlot</a></span><span class="op">(</span><span class="va">spicyMixedTest</span>,</span>
<span>           marksToPlot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"beta"</span>, <span class="st">"delta"</span>, <span class="st">"B"</span>, <span class="st">"Th"</span>, <span class="st">"otherimmune"</span>, </span>
<span>                           <span class="st">"naiveTc"</span>, <span class="st">"macrophage"</span>, <span class="st">"Tc"</span>, <span class="st">"stromal"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="04a-cell_localisation_files/figure-html/04a-signifPlot on mixed model-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The graph shows a significant decrease in co-localisation between delta and beta cells in the pancreas within the onset diabetes group compared to the non-diabetes group. Additionally, there is a significant increase in co-localisation among certain immune cell groups, including B cells and Th cells, as well as naive Tc cells and other immune cells. These findings align with the results reported in the original study.</p>
</section><section id="performing-survival-analysis" class="level3" data-number="7.1.6"><h3 data-number="7.1.6" class="anchored" data-anchor-id="performing-survival-analysis">
<span class="header-section-number">7.1.6</span> Performing survival analysis</h3>
<p><code>spicy</code> can also be used to perform survival analysis to asses whether changes in co-localisation between cell types are associated with survival probability. <code>spicy</code> fits a Cox proportional hazards model to assess the risk of death with the L-function as the explanatory variable. If there are multiple images provided per subject, <code>spicy</code> fits a Cox mixed effects model instead.</p>
<p>To perform survival analysis, <code>spicy</code> requires the <code>SingleCellExperiment</code> object being used to contain a column called <code>survival</code> as a <code>Surv</code> object.</p>
<div class="cell">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">event</span> <span class="op">=</span> <span class="fl">1</span> <span class="op">-</span> <span class="va">kerenSPE</span><span class="op">$</span><span class="va">Censored</span></span>
<span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">survival</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html">Surv</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">`Survival_days_capped*`</span>, <span class="va">kerenSPE</span><span class="op">$</span><span class="va">event</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can then perform survival analysis using the <code>spicy</code> function by specifying <code>condition = "survival"</code>. The corresponding coefficients and p-values can be accessed through the <code>survivalResults</code> slot in the <code>spicy</code> results object.</p>
<div class="cell" data-time_it="true">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Running survival analysis</span></span>
<span><span class="va">spicySurvival</span> <span class="op">=</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/spicy.html">spicy</a></span><span class="op">(</span><span class="va">kerenSPE</span>,</span>
<span>                      condition <span class="op">=</span> <span class="st">"survival"</span>,</span>
<span>                      BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># top 10 significant pairs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">spicySurvival</span><span class="op">$</span><span class="va">survivalResults</span>, <span class="fl">10</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 4
   test                       coef se.coef    p.value
   &lt;chr&gt;                     &lt;dbl&gt;   &lt;dbl&gt;      &lt;dbl&gt;
 1 Other_Immune__Tregs     0.0236  0.00868 0.00000850
 2 CD4_T_cell__Tregs       0.0178  0.00688 0.0000119 
 3 Tregs__Other_Immune     0.0237  0.00875 0.0000124 
 4 Tregs__CD4_T_cell       0.0171  0.00677 0.0000290 
 5 CD8_T_cell__CD8_T_cell  0.00605 0.00273 0.000329  
 6 Tumour__CD8_T_cell     -0.0307  0.0115  0.000582  
 7 CD8_T_cell__Tumour     -0.0307  0.0117  0.000667  
 8 CD4_T_cell__dn_T_CD3    0.00850 0.00354 0.000710  
 9 dn_T_CD3__CD4_T_cell    0.00841 0.00354 0.000925  
10 DC__Other_Immune       -0.0288  0.0123  0.00105   </code></pre>
</div>
<div style="text-align: right;">
<em>Time for this code chunk to run with 5.5 cores: 24.42 s
</em>
</div>
<p></p>
</div>
<div class="cell">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/signifPlot.html">signifPlot</a></span><span class="op">(</span><span class="va">spicySurvival</span>,</span>
<span>           marksToPlot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Tumour"</span>, <span class="st">"Tregs"</span>, <span class="st">"NK"</span>, <span class="st">"Neutrophils"</span>, <span class="st">"Mono_or_Neu"</span>, </span>
<span>                           <span class="st">"Macrophages"</span>, <span class="st">"Keratin_Tumour"</span>, <span class="st">"dn_T_CD3"</span>, <span class="st">"DC_or_Mono"</span>,</span>
<span>                           <span class="st">"DC"</span>, <span class="st">"CD8_T_cell"</span>, <span class="st">"CD4_T_cell"</span>, <span class="st">"B_cell"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="04a-cell_localisation_files/figure-html/04a-signifPlot for survival-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>From the table and the graph above, we can see that the coefficient for <code>Tumour__CD8_T_cell</code> is negative, indicating that localisation between the two cell types is associated with a better prognosis for the patient. We can also see that localisation between most immune cell types (<code>Neutrophils__CD8_T_cell</code>, <code>Tregs__CD4_T_cell</code>, <code>dn_T_CD3__CD4_T_cell</code>) is associated with worse outcomes for the patient.</p>
<p>We can examine the relationship for one pair of cell types (<code>Tumour__CD8_T_cell</code>) more closely using a Kaplan-Meier curve. Below, we extract the survival data from <code>kerenSPE</code> and create a <code>Surv</code> object.</p>
<div class="cell">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># extracting survival data</span></span>
<span><span class="va">survData</span> <span class="op">&lt;-</span> <span class="va">kerenSPE</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">colData</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">imageID</span>, <span class="va">survival</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># creating survival vector</span></span>
<span><span class="va">kerenSurv</span> <span class="op">&lt;-</span> <span class="va">survData</span><span class="op">$</span><span class="va">survival</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">kerenSurv</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">survData</span><span class="op">$</span><span class="va">imageID</span></span>
<span></span>
<span><span class="va">kerenSurv</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    1     2     3     4     5     6     7     8     9    10    11    12    13 
 2612   745 3130+ 2523+ 1683+ 2275+   584   946 3767+ 3822+ 3774+ 4353+  1072 
   14    15    16    17    18    19    20    21    23    24    25    26    27 
4145+  1754   530  2842 5063+ 3725+ 4761+   635    91   194 4785+ 4430+  3658 
   28    29    31    32    33    34    35    36    37    39    40    41 
3767+  1319  1009 1568+ 1738+ 2832+ 2759+ 3063+ 2853+ 2096+  3573 3355+ </code></pre>
</div>
</div>
<p>We can then convert our L-function metrics into a binary metric with two categories: Localised and Dispersed, and plot a Kaplan-Meier curve to view its relationship to survival probability.</p>
<div class="cell">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># obtain L-function values for a specific cell type</span></span>
<span><span class="co"># and convert into localised/dispersed based on the median</span></span>
<span><span class="va">survRelationship</span> <span class="op">=</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/bind.html">bind</a></span><span class="op">(</span><span class="va">spicySurvival</span><span class="op">)</span><span class="op">[[</span><span class="st">"Tumour__CD8_T_cell"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">survRelationship</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">survRelationship</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">survRelationship</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>                          <span class="st">"Localised"</span>, <span class="st">"Dispersed"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># ensuring consistency and removing missing values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">survRelationship</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">kerenSurv</span><span class="op">)</span></span>
<span><span class="va">survRelationship</span> <span class="op">=</span> <span class="va">survRelationship</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">survRelationship</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">kerenSurv</span> <span class="op">=</span> <span class="va">kerenSurv</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">kerenSurv</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">survRelationship</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># plotting Kaplan-Meier curve</span></span>
<span><span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/survfit2.html">survfit2</a></span><span class="op">(</span><span class="va">kerenSurv</span> <span class="op">~</span> <span class="va">survRelationship</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/ggsurvfit.html">ggsurvfit</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Tumour__CD8_T_cell"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="04a-cell_localisation_files/figure-html/04a-KM curve-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The KM curve aligns with that we observed from the bubble plot.</p>
</section><section id="accounting-for-tissue-inhomogeneity" class="level3" data-number="7.1.7"><h3 data-number="7.1.7" class="anchored" data-anchor-id="accounting-for-tissue-inhomogeneity">
<span class="header-section-number">7.1.7</span> Accounting for tissue inhomogeneity</h3>
<p>The <code>spicy</code> function can also account for tissue inhomogeneity to avoid false positives or negatives. This can be done by setting the <code>sigma =</code> parameter within the <code>spicy</code> function. By default, <code>sigma</code> is set to <code>NULL</code>, and <code>spicy</code> assumes a homogeneous tissue structure.</p>
<p>To demonstrate why <code>sigma</code> is a useful parameter, we examine the degree of co-localisation between <code>Keratin_Tumour__Neutrophils</code> in one image using the <code>getPairwise</code> function, which returns the L-function values for each cell type pair. We set the radius over which the L-function should be calculated (<code>Rs = 100</code>) and specify <code>sigma = NULL</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># filter SPE object to obtain image 24 data</span></span>
<span><span class="va">kerenSubset</span> <span class="op">=</span> <span class="va">kerenSPE</span><span class="op">[</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span><span class="op">$</span><span class="va">imageID</span> <span class="op">==</span> <span class="st">"24"</span><span class="op">]</span></span>
<span></span>
<span><span class="va">pairwiseAssoc</span> <span class="op">=</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/getPairwise.html">getPairwise</a></span><span class="op">(</span><span class="va">kerenSubset</span>, </span>
<span>                            sigma <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>                            r <span class="op">=</span> <span class="fl">100</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pairwiseAssoc</span><span class="op">[[</span><span class="st">"Keratin_Tumour__Neutrophils"</span><span class="op">]</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 10.88892</code></pre>
</div>
</div>
<p>The calculated L-function is positive, indicating attraction between the two cell types.</p>
<p>When we specify <code>sigma = 20</code> and re-calculate the L-function, it indicates that there is no relationship between <code>Keratin_Tumour</code> and <code>Neutrophils</code>, i.e., there is no major attraction or dispersion, as it now takes into account tissue inhomogeneity.</p>
<div class="cell">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pairwiseAssoc</span> <span class="op">=</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/getPairwise.html">getPairwise</a></span><span class="op">(</span><span class="va">kerenSubset</span>, </span>
<span>                            sigma <span class="op">=</span> <span class="fl">20</span>, </span>
<span>                            r <span class="op">=</span> <span class="fl">100</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pairwiseAssoc</span><span class="op">[[</span><span class="st">"Keratin_Tumour__Neutrophils"</span><span class="op">]</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.9024836</code></pre>
</div>
</div>
<p>To understand why this might be happening, we can take a closer look at the relationship between <code>Keratin_Tumour</code> and <code>Neutrophils</code>. The <code>plotImage</code> function allows us to plot any two cell types for a specific image. Below, we plot image 24 for the <code>Keratin_Tumour__Neutrophils</code> relationship by specifying <code>from = Keratin_Tumour</code> and <code>to = Neutrophils.</code></p>
<div class="cell">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">kerenSPE</span>, imageToPlot <span class="op">=</span> <span class="st">"24"</span>, </span>
<span>          from <span class="op">=</span> <span class="st">"Keratin_Tumour"</span>, </span>
<span>          to <span class="op">=</span> <span class="st">"Neutrophils"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="04a-cell_localisation_files/figure-html/04a-plot cell types-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Plotting image 24 shows that the supposed co-localisation occurs due to the dense cluster of cells near the bottom of the image, and when we take this into account, the localisation disappears.</p>
</section><section id="adjusting-for-cell-count" class="level3" data-number="7.1.8"><h3 data-number="7.1.8" class="anchored" data-anchor-id="adjusting-for-cell-count">
<span class="header-section-number">7.1.8</span> Adjusting for cell count</h3>
<p>The L-function (and by extension, the co-localisation score) is sensitive to the number of cells in an image. Too few cells can cause the L-function to become unstable and obscure meaningful spatial relationships. The issue is particularly pronounced when there is a skewed ratio of cell types in an image.</p>
<p>In the plot below, we can see that the variance of the co-localisation metric is greater when the number of cells is low, and becomes more stable as the number of cells increases.</p>
<p>To address this issue, spicyR uses a shape constrained generalised additive model (GAM) to model the co-localisation metric <span class="math inline">\(u\)</span> as a function of the number of cells per cell type. The inverse of this fitted curve is used to generate weights which are applied to each image. Images with fewer cells have a lower weight. spicyR can perform image weighting in two ways: by fitting the GAM on the scores from all cell types at once, or by fitting the GAM on each pair of cell types. To perform image weighting by cell type pair, set <code>weightsByPair = TRUE</code> when using <code>spicy</code>.</p>
<p>From the boxplot below, we can identify images that may have very few cells or a skewed cell type distribution. For consistency, we will use the cell type pair <code>Macrophages__dn_T_CD3</code>, which showed significant differences in co-localisation between cold and compartmentalised tumours in our initial analysis.</p>
<div class="cell">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/spicyBoxPlot.html">spicyBoxPlot</a></span><span class="op">(</span><span class="va">spicyTest</span>,</span>
<span>             from <span class="op">=</span> <span class="st">"Macrophages"</span>,</span>
<span>             to <span class="op">=</span> <span class="st">"dn_T_CD3"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 2 rows containing non-finite outside the scale range
(`stat_boxplot()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="04a-cell_localisation_files/figure-html/04a-spicyBoxplot 2-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Using the graph, we can filter for compartmentalised images that had an L-function value greater than 20 and examine the structure of these images.</p>
<div class="cell">
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/bind.html">bind</a></span><span class="op">(</span><span class="va">spicyTest</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">imageID</span>, <span class="va">condition</span>, <span class="va">Macrophages__dn_T_CD3</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="co"># filter for compartmenatalised images with L-function &gt; 20 (an outlier)</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">condition</span> <span class="op">==</span> <span class="st">"compartmentalised"</span> <span class="op">&amp;</span> <span class="va">Macrophages__dn_T_CD3</span> <span class="op">&gt;</span> <span class="fl">20</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  imageID         condition Macrophages__dn_T_CD3
1      34 compartmentalised              34.12149</code></pre>
</div>
</div>
<p>We can use <code>plotImage</code> to examine image 34 more closely.</p>
<div class="cell">
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">kerenSPE</span>, imageToPlot <span class="op">=</span> <span class="fl">34</span>, from <span class="op">=</span> <span class="st">"Macrophages"</span>, to <span class="op">=</span> <span class="st">"dn_T_CD3"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="04a-cell_localisation_files/figure-html/04a-plotImage 34-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The value of the L-function tells us that there is localisation between macrophages and double negative CD3 T cells in this image. However, when we examine the image in question, it appears that the number of <code>dn_T_CD3</code> cells is low compared to the number of <code>Macrophages</code>. The L-function is therefore not capturing the full context of the spatial relationship.</p>
<p>We can repeat the process above for mixed tumours.</p>
<div class="cell">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/bind.html">bind</a></span><span class="op">(</span><span class="va">spicyTest</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">imageID</span>, <span class="va">condition</span>, <span class="va">Macrophages__dn_T_CD3</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">condition</span> <span class="op">==</span> <span class="st">"mixed"</span> <span class="op">&amp;</span> <span class="va">Macrophages__dn_T_CD3</span> <span class="op">&gt;</span> <span class="fl">20</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  imageID condition Macrophages__dn_T_CD3
1      14     mixed              38.71105
2      23     mixed              46.91940
3      33     mixed              36.41111</code></pre>
</div>
</div>
<p>Here, three images have high L-function values. We will examine image 33.</p>
<div class="cell">
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">kerenSPE</span>, imageToPlot <span class="op">=</span> <span class="fl">33</span>, from <span class="op">=</span> <span class="st">"Macrophages"</span>, to <span class="op">=</span> <span class="st">"dn_T_CD3"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="04a-cell_localisation_files/figure-html/04a-plotImage 33-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>As before, we can see that the number of <code>dn_T_CD3</code> cells is low, driving the L-function up even though true localisation may not be occurring.</p>
<div class="callout callout-style-default callout-tip callout-titled" title="Assessing image quality">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Assessing image quality
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>How do we distinguish “good” images from “bad” ones?</strong></p>
<ul>
<li><p>Typically, images with very few cells can skew the L-function, and make it appear as though there is localisation or dispersion when, in reality, no significant spatial pattern exists. These images usually have very high or very low L-function values compared to other images in the same group.</p></li>
<li><p>Skewed images must be further examined to understand the reason behind the abnormal L-function values. Factors such as imaging artifacts, poor segmentation, or uneven cell distribution could contribute to this skew. In contrast, good quality images typically exhibit a more balanced ratio of cell types and a consistent number of cells across images.</p></li>
</ul>
</div>
</div>
<p>In the next section, we will demonstrate how and why we can derive spatial contexts within tissue samples to produce robust spatial quantifications.</p>
</section></section><section id="sessioninfo" class="level2" data-number="7.2"><h2 data-number="7.2" class="anchored" data-anchor-id="sessioninfo">
<span class="header-section-number">7.2</span> sessionInfo</h2>
<div class="cell">
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.5.0 (2025-04-11)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.4.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.1

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Australia/Sydney
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] spicyR_1.19.4               testthat_3.2.3             
 [3] ggsurvfit_1.1.0             treekoR_1.16.0             
 [5] tibble_3.2.1                survival_3.8-3             
 [7] dplyr_1.1.4                 imcRtools_1.14.0           
 [9] SpatialDatasets_1.6.3       ExperimentHub_2.16.0       
[11] AnnotationHub_3.16.0        BiocFileCache_2.16.0       
[13] dbplyr_2.5.0                SpatialExperiment_1.18.1   
[15] SingleCellExperiment_1.30.1 SummarizedExperiment_1.38.1
[17] Biobase_2.68.0              GenomicRanges_1.60.0       
[19] GenomeInfoDb_1.44.0         IRanges_2.42.0             
[21] S4Vectors_0.46.0            BiocGenerics_0.54.0        
[23] generics_0.1.4              MatrixGenerics_1.20.0      
[25] matrixStats_1.5.0           ggplot2_3.5.2              
[27] Statial_1.10.0             

loaded via a namespace (and not attached):
  [1] dichromat_2.0-0.1           vroom_1.6.5                
  [3] urlchecker_1.0.1            tiff_0.1-12                
  [5] dcanr_1.24.0                goftest_1.2-3              
  [7] DT_0.33                     Biostrings_2.76.0          
  [9] HDF5Array_1.36.0            TH.data_1.1-3              
 [11] vctrs_0.6.5                 spatstat.random_3.4-1      
 [13] digest_0.6.37               png_0.1-8                  
 [15] shape_1.4.6.1               proxy_0.4-27               
 [17] BiocBaseUtils_1.10.0        ggrepel_0.9.6              
 [19] deldir_2.0-4                magick_2.8.5               
 [21] MASS_7.3-65                 reshape2_1.4.4             
 [23] httpuv_1.6.16               foreach_1.5.2              
 [25] withr_3.0.2                 ggfun_0.1.8                
 [27] xfun_0.52                   ellipsis_0.3.2             
 [29] ggpubr_0.6.0                doRNG_1.8.6.2              
 [31] memoise_2.0.1               RTriangle_1.6-0.15         
 [33] cytomapper_1.20.0           ggbeeswarm_0.7.2           
 [35] RProtoBufLib_2.20.0         profvis_0.4.0              
 [37] systemfonts_1.2.3           tidytree_0.4.6             
 [39] zoo_1.8-14                  GlobalOptions_0.1.2        
 [41] Formula_1.2-5               KEGGREST_1.48.0            
 [43] promises_1.3.2              httr_1.4.7                 
 [45] rstatix_0.7.2               rhdf5filters_1.20.0        
 [47] rhdf5_2.52.0                rstudioapi_0.17.1          
 [49] miniUI_0.1.1.1              UCSC.utils_1.4.0           
 [51] units_0.8-7                 concaveman_1.1.0           
 [53] curl_6.2.3                  ggraph_2.2.1               
 [55] h5mread_1.0.1               polyclip_1.10-7            
 [57] GenomeInfoDbData_1.2.14     SparseArray_1.8.0          
 [59] fftwtools_0.9-11            desc_1.4.3                 
 [61] xtable_1.8-4                stringr_1.5.1              
 [63] doParallel_1.0.17           evaluate_1.0.3             
 [65] S4Arrays_1.8.0              hms_1.1.3                  
 [67] colorspace_2.1-1            filelock_1.0.3             
 [69] spatstat.data_3.1-6         magrittr_2.0.3             
 [71] readr_2.1.5                 later_1.4.2                
 [73] viridis_0.6.5               ggtree_3.16.0              
 [75] lattice_0.22-6              spatstat.geom_3.4-1        
 [77] XML_3.99-0.18               scuttle_1.18.0             
 [79] ggupset_0.4.1               class_7.3-23               
 [81] svgPanZoom_0.3.4            pillar_1.10.2              
 [83] simpleSeg_1.10.0            nlme_3.1-168               
 [85] iterators_1.0.14            EBImage_4.50.0             
 [87] compiler_4.5.0              beachmat_2.24.0            
 [89] stringi_1.8.7               sf_1.0-21                  
 [91] devtools_2.4.5              tensor_1.5                 
 [93] minqa_1.2.8                 ClassifyR_3.12.0           
 [95] plyr_1.8.9                  crayon_1.5.3               
 [97] abind_1.4-8                 gridGraphics_0.5-1         
 [99] locfit_1.5-9.12             sp_2.2-0                   
[101] graphlayouts_1.2.2          bit_4.6.0                  
[103] terra_1.8-50                sandwich_3.1-1             
[105] codetools_0.2-20            multcomp_1.4-28            
[107] textshaping_1.0.1           e1071_1.7-16               
[109] GetoptLong_1.0.5            plotly_4.10.4              
[111] mime_0.13                   MultiAssayExperiment_1.34.0
[113] splines_4.5.0               circlize_0.4.16            
[115] Rcpp_1.0.14                 utf8_1.2.5                 
[117] knitr_1.50                  blob_1.2.4                 
[119] clue_0.3-66                 BiocVersion_3.21.1         
[121] lme4_1.1-37                 fs_1.6.6                   
[123] nnls_1.6                    Rdpack_2.6.4               
[125] pkgbuild_1.4.8              ggplotify_0.1.2            
[127] ggsignif_0.6.4              Matrix_1.7-3               
[129] scam_1.2-19                 statmod_1.5.0              
[131] tzdb_0.5.0                  svglite_2.2.1              
[133] tweenr_2.0.3                pkgconfig_2.0.3            
[135] pheatmap_1.0.12             tools_4.5.0                
[137] cachem_1.1.0                rbibutils_2.3              
[139] RSQLite_2.3.11              viridisLite_0.4.2          
[141] DBI_1.2.3                   numDeriv_2016.8-1.1        
[143] fastmap_1.2.0               rmarkdown_2.29             
[145] scales_1.4.0                grid_4.5.0                 
[147] usethis_3.1.0               shinydashboard_0.7.3       
[149] broom_1.0.8                 patchwork_1.3.0            
[151] BiocManager_1.30.25         carData_3.0-5              
[153] farver_2.1.2                reformulas_0.4.1           
[155] tidygraph_1.3.1             mgcv_1.9-3                 
[157] yaml_2.3.10                 ggthemes_5.1.0             
[159] cli_3.6.5                   purrr_1.0.4                
[161] hopach_2.68.0               lifecycle_1.0.4            
[163] mvtnorm_1.3-3               sessioninfo_1.2.3          
[165] backports_1.5.0             BiocParallel_1.42.0        
[167] cytolib_2.20.0              gtable_0.3.6               
[169] rjson_0.2.23                parallel_4.5.0             
[171] ape_5.8-1                   limma_3.64.1               
[173] jsonlite_2.0.0              edgeR_4.6.2                
[175] bitops_1.0-9                bit64_4.6.0-1              
[177] brio_1.1.5                  Rtsne_0.17                 
[179] yulab.utils_0.2.0           FlowSOM_2.16.0             
[181] spatstat.utils_3.1-4        BiocNeighbors_2.2.0        
[183] ranger_0.17.0               flowCore_2.20.0            
[185] bdsmatrix_1.3-7             spatstat.univar_3.1-3      
[187] lazyeval_0.2.2              shiny_1.10.0               
[189] ConsensusClusterPlus_1.72.0 htmltools_0.5.8.1          
[191] diffcyt_1.28.0              rappdirs_0.3.3             
[193] distances_0.1.12            glue_1.8.0                 
[195] XVector_0.48.0              RCurl_1.98-1.17            
[197] rprojroot_2.0.4             treeio_1.32.0              
[199] classInt_0.4-11             coxme_2.2-22               
[201] jpeg_0.1-11                 gridExtra_2.3              
[203] boot_1.3-31                 igraph_2.1.4               
[205] R6_2.6.1                    tidyr_1.3.1                
[207] ggiraph_0.8.13              labeling_0.4.3             
[209] ggh4x_0.3.0                 cluster_2.1.8.1            
[211] rngtools_1.5.2              pkgload_1.4.0              
[213] Rhdf5lib_1.30.0             aplot_0.2.5                
[215] nloptr_2.2.1                DelayedArray_0.34.1        
[217] tidyselect_1.2.1            vipor_0.4.7                
[219] ggforce_0.4.2               raster_3.6-32              
[221] car_3.1-3                   AnnotationDbi_1.70.0       
[223] KernSmooth_2.23-26          data.table_1.17.4          
[225] htmlwidgets_1.6.4           ComplexHeatmap_2.24.0      
[227] RColorBrewer_1.1-3          rlang_1.1.6                
[229] spatstat.sparse_3.1-0       spatstat.explore_3.4-3     
[231] remotes_2.5.0               lmerTest_3.1-3             
[233] uuid_1.2-1                  colorRamps_2.3.4           
[235] ggnewscale_0.5.1            beeswarm_0.4.0             </code></pre>
</div>
</div>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./03a-cell_annotation.html" class="pagination-link" aria-label="Unsupervised clustering for cell annotation">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Unsupervised clustering for cell annotation</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./04b-cell_localisation_parent.html" class="pagination-link" aria-label="Cell relationships relative to expected behaviour">
        <span class="nav-page-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Cell relationships relative to expected behaviour</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>