<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>10&nbsp; Changes in marker expression – Spatial analysis playbook</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./07-classification.html" rel="next">
<link href="./05-cellular_niches.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><link rel="stylesheet" href="styles.css">
</head>
<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="./index.html" class="navbar-brand navbar-brand-logo">
    <img src="./images/USydLogo.svg" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Spatial analysis playbook</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./06-changes_in_marker_expression.html">Marker expression</a></li><li class="breadcrumb-item"><a href="./06-changes_in_marker_expression.html"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Changes in marker expression</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Overview</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Datasets</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./getting_started.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Getting Started</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Processing</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01-processing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Cell segmentation and pre-processing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02-quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Quality Control</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Cell annotation</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03a-cell_annotation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Unsupervised clustering for cell annotation</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Cell localisation</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04a-cell_localisation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Cell localisation between pairs of cell types</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04b-cell_localisation_parent.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Cell relationships relative to expected behaviour</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">Spatial domains</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05-cellular_niches.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Identifying spatial domains with unsupervised clustering</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true">
 <span class="menu-text">Marker expression</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06-changes_in_marker_expression.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Changes in marker expression</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true">
 <span class="menu-text">Classification</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07-classification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Finding associations between clinical variables and spatial features</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" role="navigation" aria-expanded="true">
 <span class="menu-text">Case studies</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-8" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08-case_study1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Case Study: Head and Neck Squamous Cell Carcinoma (Ferguson et al., 2022)</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#statial-marker-means" id="toc-statial-marker-means" class="nav-link active" data-scroll-target="#statial-marker-means"><span class="header-section-number">10.1</span> Statial: Marker means</a></li>
  <li>
<a href="#spatiomark-identifying-continuous-changes-in-cell-state" id="toc-spatiomark-identifying-continuous-changes-in-cell-state" class="nav-link" data-scroll-target="#spatiomark-identifying-continuous-changes-in-cell-state"><span class="header-section-number">10.2</span> SpatioMark: Identifying continuous changes in cell state</a>
  <ul class="collapse">
<li><a href="#continuous-cell-state-changes-within-a-single-image" id="toc-continuous-cell-state-changes-within-a-single-image" class="nav-link" data-scroll-target="#continuous-cell-state-changes-within-a-single-image"><span class="header-section-number">10.2.1</span> Continuous cell state changes within a single image</a></li>
  <li><a href="#continuous-cell-state-changes-across-all-images" id="toc-continuous-cell-state-changes-across-all-images" class="nav-link" data-scroll-target="#continuous-cell-state-changes-across-all-images"><span class="header-section-number">10.2.2</span> Continuous cell state changes across all images</a></li>
  <li><a href="#contamination-lateral-marker-spill-over" id="toc-contamination-lateral-marker-spill-over" class="nav-link" data-scroll-target="#contamination-lateral-marker-spill-over"><span class="header-section-number">10.2.3</span> Contamination (Lateral marker spill over)</a></li>
  <li><a href="#associate-continuous-state-changes-with-survival-outcomes" id="toc-associate-continuous-state-changes-with-survival-outcomes" class="nav-link" data-scroll-target="#associate-continuous-state-changes-with-survival-outcomes"><span class="header-section-number">10.2.4</span> Associate continuous state changes with survival outcomes</a></li>
  </ul>
</li>
  <li><a href="#sessioninfo" id="toc-sessioninfo" class="nav-link" data-scroll-target="#sessioninfo"><span class="header-section-number">10.3</span> sessionInfo</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./06-changes_in_marker_expression.html">Marker expression</a></li><li class="breadcrumb-item"><a href="./06-changes_in_marker_expression.html"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Changes in marker expression</span></a></li></ol></nav><div class="quarto-title">
<h1 class="title">
<span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Changes in marker expression</span>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><p>Sometimes, cells that appear similar based on marker expression alone can behave very differently depending on their surroundings. Subtle shifts in marker levels—too faint or variable to be captured through clustering—may hold important clues about a cell’s functional state. For example, immune cells may only become activated when positioned near tumour or infected cells, and these changes may not be detectable when cells are analysed in isolation. Technical noise and overlapping expression patterns further complicate efforts to distinguish such nuanced subpopulations using conventional approaches.</p>
<p>In these cases, spatial context becomes a critical lens. By analysing how marker expression varies in relation to a cell’s neighbourhood—such as its proximity to other cell types or tissue structures—we can uncover functional diversity that would otherwise remain hidden. This spatial perspective reveals how cells adapt to their microenvironment, transition between states, or participate in local interactions, offering a richer understanding of tissue dynamics and disease mechanisms.</p>
<p>The <code>SpatioMark</code> method, implemented in the <code>Statial</code> package, is designed for exactly this purpose. It detects markers whose expression levels change within a given cell type depending on their spatial context—highlighting activation, suppression, or other state changes that are missed by standard clustering.</p>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/Statial">Statial</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ellispatrick.github.io/spicyR/">spicyR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/ClassifyR/">ClassifyR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ellispatrick.github.io/lisaClust/">lisaClust</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pharmaverse/ggsurvfit">ggsurvfit</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/therneau/survival">survival</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/">tibble</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">treekoR</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># set parameters</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span><span class="va">use_mc</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">use_mc</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">nCores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html">detectCores</a></span><span class="op">(</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="va">nCores</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="op">}</span></span>
<span><span class="va">BPPARAM</span> <span class="op">&lt;-</span> <span class="fu">simpleSeg</span><span class="fu">:::</span><span class="fu"><a href="https://sydneybiox.github.io/simpleSeg/reference/generateBPParam.html">generateBPParam</a></span><span class="op">(</span><span class="va">nCores</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="statial-marker-means" class="level2" data-number="10.1"><h2 data-number="10.1" class="anchored" data-anchor-id="statial-marker-means">
<span class="header-section-number">10.1</span> Statial: Marker means</h2>
<p>Before diving into spatially contextualised marker changes with <code>SpatioMark</code>, it’s useful to understand how more conventional marker expression summaries—like marker means—can still offer valuable insight, especially when stratified by spatial features. A marker mean provides a simple but effective way to quantify expression: we calculate the average intensity of a marker across a set of cells, and then compare these averages across experimental conditions. While this can be done at the whole-image level, much more meaningful biological comparisons emerge when we stratify by cell type, spatial domain, or both.</p>
<p>For example, if you’re interested in understanding how CD163 expression changes in infiltrating macrophages located specifically within tumour regions across treatment groups, you would focus on the marker mean for macrophages confined to that domain. This allows you to ask precise questions about functional changes in well-defined spatial and cellular contexts.</p>
<p>For this demonstration, we will use the <a href="./datasets.html">Keren 2018</a> dataset.</p>
<div class="cell" data-time_it="true">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">kerenSPE</span> <span class="op">&lt;-</span> <span class="fu">SpatialDatasets</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialDatasets/man/SpatialDatasets.html">spe_Keren_2018</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Removing patients without survival data.</span></span>
<span><span class="va">kerenSPE</span> <span class="op">&lt;-</span> <span class="va">kerenSPE</span><span class="op">[</span>,<span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">`Survival_days_capped*`</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># identify spatial domains with lisaClust</span></span>
<span><span class="va">kerenSPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/lisaClust.html">lisaClust</a></span><span class="op">(</span><span class="va">kerenSPE</span>,</span>
<span>                      k <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                      BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div style="text-align: right;">
<em>Time for this code chunk to run with 5.5 cores: 35.06 s
</em>
</div>
<p></p>
</div>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">event</span> <span class="op">=</span> <span class="fl">1</span> <span class="op">-</span> <span class="va">kerenSPE</span><span class="op">$</span><span class="va">Censored</span></span>
<span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">survival</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html">Surv</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">`Survival_days_capped*`</span>, <span class="va">kerenSPE</span><span class="op">$</span><span class="va">event</span><span class="op">)</span></span>
<span><span class="co"># Extracting survival data</span></span>
<span><span class="va">survData</span> <span class="op">&lt;-</span> <span class="va">kerenSPE</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">imageID</span>, <span class="va">survival</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">survival</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span></span>
<span><span class="co"># Creating survival vector</span></span>
<span><span class="va">kerenSurv</span> <span class="op">&lt;-</span> <span class="va">survData</span><span class="op">$</span><span class="va">survival</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">kerenSurv</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">survData</span><span class="op">$</span><span class="va">imageID</span></span>
<span></span>
<span><span class="va">kerenSurv</span> <span class="op">&lt;-</span> <span class="va">kerenSurv</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">kerenSurv</span><span class="op">)</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Our <code>Statial</code> package provides functionality to identify the average marker expression of a given cell type in a given region, using the <code>getMarkerMeans</code> function. Similar to <code>spicyR</code> and <code>lisaClust</code>, these features can also be used for survival analysis.</p>
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cellTypeRegionMeans</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/getMarkerMeans.html">getMarkerMeans</a></span><span class="op">(</span><span class="va">kerenSPE</span>,</span>
<span>  imageID <span class="op">=</span> <span class="st">"imageID"</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>  region <span class="op">=</span> <span class="st">"region"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cellTypeRegionMeans</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Na__Keratin_Tumour__region_2 Si__Keratin_Tumour__region_2
1                   -0.7775969                   -0.5503253
2                    0.1489698                   -0.2596228
3                   -0.7918177                   -0.5268896
  P__Keratin_Tumour__region_2
1                  -1.1306943
2                  -0.8310180
3                  -0.4678555</code></pre>
</div>
</div>
<p>The output is a dataframe containing the average expression of each marker in each cell type in each region. The column names are formatted as: <code>marker__cell_type__region</code>.</p>
<p>We can use the <code>colTest</code> function from spicyR to check whether average marker expression in each cell type in each region is associated with survival probability. <code>colTest</code> requires three arguments: i) <code>df</code> specifies the dataframe containing marker means, ii) <code>condition</code> specifies the outcome of interest, and iii) <code>type</code> specifies the type of test to perform (wilcox, t-test, or survival). In the code below, we’ve specified <code>condition</code> to be our <code>Surv</code> vector and <code>type = survival</code> indicates we are performing survival analysis.</p>
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">survivalResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/colTest.html">colTest</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">cellTypeRegionMeans</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">kerenSurv</span><span class="op">)</span>, <span class="op">]</span>, </span>
<span>                           condition <span class="op">=</span> <span class="va">kerenSurv</span>, </span>
<span>                           type <span class="op">=</span> <span class="st">"survival"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">survivalResults</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                    coef se.coef    pval adjPval
B7H3__CD4_T_cell__region_1         270.0   76.00 0.00038    0.42
CD163__CD4_T_cell__region_1         67.0   19.00 0.00038    0.42
FoxP3__CD4_T_cell__region_1         25.0    7.20 0.00052    0.42
Si__Unidentified__region_4          -3.1    0.89 0.00053    0.42
CD56__CD4_T_cell__region_1          28.0    8.10 0.00067    0.42
Keratin6__Keratin_Tumour__region_3   1.6    0.47 0.00074    0.42
                                                              cluster
B7H3__CD4_T_cell__region_1                 B7H3__CD4_T_cell__region_1
CD163__CD4_T_cell__region_1               CD163__CD4_T_cell__region_1
FoxP3__CD4_T_cell__region_1               FoxP3__CD4_T_cell__region_1
Si__Unidentified__region_4                 Si__Unidentified__region_4
CD56__CD4_T_cell__region_1                 CD56__CD4_T_cell__region_1
Keratin6__Keratin_Tumour__region_3 Keratin6__Keratin_Tumour__region_3</code></pre>
</div>
</div>
<p>Our most significant relationship appears to be <code>B7H3__CD4_T_cell__region_1</code>, which represents the average expression of the B7H3 marker in CD4 T cells in region 1. The positive coefficient associated with this relationship indicates that higher expression of B7H3 in CD4 T cells in this region is associated with poorer survival outcomes for patients.</p>
<p>We can examine this relationship in more detail by plotting a Kaplan-Meier curve.</p>
<div class="cell">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Selecting the most significant relationship</span></span>
<span><span class="va">survRelationship</span> <span class="op">&lt;-</span> <span class="va">cellTypeRegionMeans</span><span class="op">[[</span><span class="st">"B7H3__CD4_T_cell__region_1"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">survRelationship</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">survRelationship</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">survRelationship</span><span class="op">)</span>, <span class="st">"Higher expression"</span>, <span class="st">"Lower expression"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plotting Kaplan-Meier curve</span></span>
<span><span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/survfit2.html">survfit2</a></span><span class="op">(</span><span class="va">kerenSurv</span> <span class="op">~</span> <span class="va">survRelationship</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/ggsurvfit.html">ggsurvfit</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/add_pvalue.html">add_pvalue</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"B7H3__CD4_T_cell__region_1"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="06-changes_in_marker_expression_files/figure-html/06-KM curve for region-1.png" class="img-fluid figure-img" width="480"></p>
</figure>
</div>
</div>
</div>
<p>We can also look at cell types alone, without separating by region. To do this, we simply do not specify <code>region</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cellTypeMeans</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/getMarkerMeans.html">getMarkerMeans</a></span><span class="op">(</span><span class="va">kerenSPE</span>,</span>
<span>  imageID <span class="op">=</span> <span class="st">"imageID"</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellType"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">survivalResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/colTest.html">colTest</a></span><span class="op">(</span><span class="va">cellTypeMeans</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">kerenSurv</span><span class="op">)</span>, <span class="op">]</span>, <span class="va">kerenSurv</span>, type <span class="op">=</span> <span class="st">"survival"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">survivalResults</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                          coef se.coef    pval adjPval                  cluster
CD56__Tregs               15.0    4.20 0.00051    0.41              CD56__Tregs
HLA_Class_1__Mono_or_Neu  -1.4    0.47 0.00230    0.51 HLA_Class_1__Mono_or_Neu
FoxP3__Endothelial       120.0   40.00 0.00350    0.51       FoxP3__Endothelial
CD138__Macrophages         1.1    0.39 0.00490    0.51       CD138__Macrophages
HLA_Class_1__CD8_T_cell   -1.3    0.48 0.00520    0.51  HLA_Class_1__CD8_T_cell
CD138__dn_T_CD3            0.4    0.15 0.00570    0.51          CD138__dn_T_CD3</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Selecting the most significant relationship</span></span>
<span><span class="va">survRelationship</span> <span class="op">&lt;-</span> <span class="va">cellTypeMeans</span><span class="op">[[</span><span class="st">"CD56__Tregs"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">survRelationship</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">survRelationship</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">survRelationship</span><span class="op">)</span>, <span class="st">"Higher expression"</span>, <span class="st">"Lower expression"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plotting Kaplan-Meier curve</span></span>
<span><span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/survfit2.html">survfit2</a></span><span class="op">(</span><span class="va">kerenSurv</span> <span class="op">~</span> <span class="va">survRelationship</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/ggsurvfit.html">ggsurvfit</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/add_pvalue.html">add_pvalue</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"CD56__Tregs"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="06-changes_in_marker_expression_files/figure-html/06-KM curve for celltype-1.png" class="img-fluid figure-img" width="480"></p>
</figure>
</div>
</div>
</div>
<p>The coefficient associated with <code>CD56__Tregs</code> is positive, which indicates that higher expression is associated with poorer survival outcomes for patients, and this is also reflected in the KM curve.</p>
</section><section id="spatiomark-identifying-continuous-changes-in-cell-state" class="level2" data-number="10.2"><h2 data-number="10.2" class="anchored" data-anchor-id="spatiomark-identifying-continuous-changes-in-cell-state">
<span class="header-section-number">10.2</span> SpatioMark: Identifying continuous changes in cell state</h2>
<p>Marker means summarised per cell type or per spatial region can offer valuable insights into how expression levels vary across conditions or anatomical compartments. However, these summaries treat cells as independent units and often miss finer-grained, dynamic patterns that arise from cell-cell interactions.</p>
<p>This is where <code>SpatioMark</code> becomes particularly powerful. Instead of averaging across broad groups, <code>SpatioMark</code> captures how marker expression in a given cell type changes in a continuous fashion based on proximity to other cell types. This allows us to detect context-dependent expression shifts—revealing, for instance, how the presence of neighbouring immune or tumour cells can influence a cell’s functional state. By focusing on these spatial gradients in marker expression, SpatioMark provides a framework for understanding how local microenvironments modulate cell behaviour in situ.</p>
<p><img src="images/spatiomark_fig1.jpg" align="center" style="height: 300px; border: 0px"></p>
<section id="continuous-cell-state-changes-within-a-single-image" class="level3" data-number="10.2.1"><h3 data-number="10.2.1" class="anchored" data-anchor-id="continuous-cell-state-changes-within-a-single-image">
<span class="header-section-number">10.2.1</span> Continuous cell state changes within a single image</h3>
<p>The first step in analysing these changes is to calculate the spatial proximity and abundance of each cell to every cell type. <code>getDistances</code> calculates the Euclidean distance from each cell to the nearest cell of each cell type. <code>getAbundances</code> calculates the K-function value for each cell with respect to each cell type. Both metrics are stored in the <code>reducedDims</code> slot of the <code>SpatialExperiment</code> object.</p>
<!--fix spatial coordinates, assign spatial coords to spatial coords slot-->
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># assign spatial coordinates</span></span>
<span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">x</span> <span class="op">=</span> <span class="fu">spatialCoords</span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">y</span> <span class="op">=</span> <span class="fu">spatialCoords</span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># calculate distances for a maximum distance of 200</span></span>
<span><span class="va">kerenSPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/getDistances.html">getDistances</a></span><span class="op">(</span><span class="va">kerenSPE</span>,</span>
<span>  maxDist <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># calculate K-function for a radius of 200</span></span>
<span><span class="va">kerenSPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/getAbundances.html">getAbundances</a></span><span class="op">(</span><span class="va">kerenSPE</span>,</span>
<span>  r <span class="op">=</span> <span class="fl">200</span>,</span>
<span>  nCores <span class="op">=</span> <span class="va">nCores</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDims</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of length 2
names(2): distances abundances</code></pre>
</div>
</div>
<p>First, let’s examine the same effect observed earlier with <code>Kontextual</code> - the localisation between p53+ keratin/tumour cells and macrophages in the context of total keratin/tumour cells for image 6 of the Keren 2018 dataset.</p>
<p>Statial provides two main functions to assess this relationship - <code>calcStateChanges</code> and <code>plotStateChanges</code>. We can use <code>calcStateChanges</code> to examine the relationship between two cell types for one marker in a specific image. In this case, we’re examining the relationship between keratin/tumour cells (<code>from = Keratin_Tumour</code>) and macrophages (<code>to = "Macrophages"</code>) for the marker p53 (<code>marker = "p53"</code>) in <code>image = "6"</code>. We can appreciate that the <code>fdr</code> statistic for this relationship is significant, with a negative t-value, indicating that the expression of p53 in keratin/tumour cells decreases as distance from macrophages increases.</p>
<div class="cell">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">stateChanges</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/calcStateChanges.html">calcStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSPE</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  image <span class="op">=</span> <span class="st">"6"</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"Keratin_Tumour"</span>,</span>
<span>  to <span class="op">=</span> <span class="st">"Macrophages"</span>,</span>
<span>  marker <span class="op">=</span> <span class="st">"p53"</span>,</span>
<span>  nCores <span class="op">=</span> <span class="va">nCores</span><span class="op">)</span></span>
<span></span>
<span><span class="va">stateChanges</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  imageID primaryCellType otherCellType marker         coef      tval
1       6  Keratin_Tumour   Macrophages    p53 -0.001402178 -7.010113
          pval          fdr
1 2.868257e-12 2.868257e-12</code></pre>
</div>
</div>
<p>Statial also provides a convenient function for visualising this interaction - <code>plotStateChanges</code>. Here, again we can specify <code>image = 6</code> and our main cell types of interest, keratin/tumour cells and macrophages, and our marker p53, in the same format as <code>calcStateChanges</code>.</p>
<p>Through this analysis, we can observe that keratin/tumour cells closer to a group of macrophages tend to have higher expression of p53, as observed in the first graph. This relationship is quantified with the second graph, showing an overall decrease of p53 expression in keratin/tumour cells as distance from macrophages increases.</p>
<p>These results allow us to essentially arrive at the same result as Kontextual, which calculated a localisation between p53+ keratin/tumour cells and macrophages in the wider context of keratin/tumour cells.</p>
<div class="cell">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/plotStateChanges.html">plotStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSPE</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  image <span class="op">=</span> <span class="st">"6"</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"Keratin_Tumour"</span>,</span>
<span>  to <span class="op">=</span> <span class="st">"Macrophages"</span>,</span>
<span>  marker <span class="op">=</span> <span class="st">"p53"</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  shape <span class="op">=</span> <span class="fl">19</span>,</span>
<span>  interactive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plotModelFit <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot the image</span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">image</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="06-changes_in_marker_expression_files/figure-html/06-plotStateChanges-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># plot the scatter plot</span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">scatter</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="06-changes_in_marker_expression_files/figure-html/06-scatterplot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="continuous-cell-state-changes-across-all-images" class="level3" data-number="10.2.2"><h3 data-number="10.2.2" class="anchored" data-anchor-id="continuous-cell-state-changes-across-all-images">
<span class="header-section-number">10.2.2</span> Continuous cell state changes across all images</h3>
<p>Beyond looking at single cell-to-cell interactions for a single image, we can also look at all interactions across all images. The <code>calcStateChanges</code> function provided by Statial can be expanded for this exact purpose - by not specifying cell types, a marker, or an image, <code>calcStateChanges</code> will examine the most significant correlations between distance and marker expression across the entire dataset. Here, we’ve filtered out the most significant interactions to only include those found within image 6 of the Keren 2018 dataset.</p>
<div class="cell" data-time_it="true">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">stateChanges</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/calcStateChanges.html">calcStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSPE</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  nCores <span class="op">=</span> <span class="va">nCores</span>,</span>
<span>  minCells <span class="op">=</span> <span class="fl">100</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">stateChanges</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">imageID</span> <span class="op">==</span> <span class="fl">6</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   imageID primaryCellType otherCellType       marker         coef      tval
1        6  Keratin_Tumour  Unidentified           Na  0.004218419  25.03039
2        6  Keratin_Tumour   Macrophages  HLA_Class_1 -0.003823497 -24.69629
3        6  Keratin_Tumour    CD4_T_cell  HLA_Class_1 -0.003582774 -23.87797
4        6  Keratin_Tumour  Unidentified Beta.catenin  0.005893120  23.41953
5        6  Keratin_Tumour    CD8_T_cell  HLA_Class_1 -0.003154544 -23.13804
6        6  Keratin_Tumour    DC_or_Mono  HLA_Class_1 -0.003353834 -22.98944
7        6  Keratin_Tumour      dn_T_CD3  HLA_Class_1 -0.003123446 -22.63197
8        6  Keratin_Tumour        Tumour  HLA_Class_1  0.003684079  21.94265
9        6  Keratin_Tumour    CD4_T_cell           Fe -0.003457338 -21.43550
10       6  Keratin_Tumour    CD4_T_cell   phospho.S6 -0.002892457 -20.50767
            pval           fdr
1  6.971648e-127 3.361087e-123
2  7.814253e-124 3.408521e-120
3  1.745242e-116 5.328836e-113
4  1.917245e-112 5.488145e-109
5  5.444541e-110 1.424922e-106
6  1.053130e-108 2.679645e-105
7  1.237988e-105 2.870893e-102
8  8.188258e-100  1.630540e-96
9   1.287478e-95  2.246354e-92
10  3.928912e-88  5.712544e-85</code></pre>
</div>
<div style="text-align: right;">
<em>Time for this code chunk to run with 5.5 cores: 3.73 s
</em>
</div>
<p></p>
</div>
<p>In image 6, the majority of the top 10 most significant interactions occur between keratin/tumour cells and an immune population, and many of these interactions appear to involve the HLA class I ligand.</p>
<p>We can examine some of these interactions further with the <code>plotStateChanges</code> function.</p>
<div class="cell">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/plotStateChanges.html">plotStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSPE</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  image <span class="op">=</span> <span class="st">"6"</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"Keratin_Tumour"</span>,</span>
<span>  to <span class="op">=</span> <span class="st">"Macrophages"</span>,</span>
<span>  marker <span class="op">=</span> <span class="st">"HLA_Class_1"</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  shape <span class="op">=</span> <span class="fl">19</span>,</span>
<span>  interactive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plotModelFit <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"lm"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot the image</span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">image</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="06-changes_in_marker_expression_files/figure-html/06-plotStateChanges HLA-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># plot the scatter plot</span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">scatter</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="06-changes_in_marker_expression_files/figure-html/06-scatter plot HLA-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The plot above shows us a clear visual correlation - as the distance from macrophages decreases, keratin/tumour cells increase their expression HLA class I. Biologically, HLA Class I molecules are ligands present on all nucleated cells, responsible for presenting internal cell antigens to the immune system. Their role is to mark abnormal cells for destruction by CD8+ T cells or NK cells, facilitating immune surveillance and response.</p>
<p>Next, let’s take a look at the top 10 most significant results across all images.</p>
<div class="cell">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">stateChanges</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       imageID primaryCellType otherCellType     marker         coef
69468       37     Endothelial        Tumour       Lag3 -0.001621517
150351      11     Neutrophils            NK       CD56 -0.059936866
16402       35      CD4_T_cell        B_cell       CD20 -0.029185750
16498       35      CD4_T_cell    DC_or_Mono       CD20  0.019125946
4891        35          B_cell    DC_or_Mono phospho.S6  0.005282065
16507       35      CD4_T_cell    DC_or_Mono phospho.S6  0.004033218
4885        35          B_cell    DC_or_Mono     HLA.DR  0.011120703
5043        35          B_cell  Other_Immune          P  0.011182182
16354       35      CD4_T_cell      dn_T_CD3       CD20  0.016349492
4888        35          B_cell    DC_or_Mono     H3K9ac  0.005096632
                tval          pval           fdr
69468  -1.082138e+14  0.000000e+00  0.000000e+00
150351 -1.419485e+14  0.000000e+00  0.000000e+00
16402  -4.057355e+01 7.019343e-282 4.286502e-277
16498   4.053436e+01 1.891267e-281 8.662052e-277
4891    4.041385e+01 5.306590e-278 1.944345e-273
16507   3.472882e+01 4.519947e-219 1.380098e-214
4885    3.415344e+01 8.401034e-212 2.198683e-207
5043    3.414375e+01 1.056403e-211 2.419176e-207
16354   3.391901e+01 1.219488e-210 2.482349e-206
4888    3.399856e+01 3.266533e-210 5.984320e-206</code></pre>
</div>
</div>
<p>Immediately, we can appreciate that a couple of these interactions are not biologically plausible. One of the most significant interactions occurs between B cells and CD4 T cells in image 35, where CD4 T cells are found to increase CD20 expression when in close proximity to B cells. Biologically, CD20 is a highly specific marker for B cells, and under healthy circumstances are usually not expressed in T cells.</p>
<p>Could this potentially be an artefact of <code>calcStateChanges</code>? We can examine the image through the <code>plotStateChanges</code> function, where we indeed observe a strong increase in CD20 expression in T cells nearby B cell populations.</p>
<div class="cell">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/plotStateChanges.html">plotStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSPE</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  image <span class="op">=</span> <span class="st">"35"</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"CD4_T_cell"</span>,</span>
<span>  to <span class="op">=</span> <span class="st">"B_cell"</span>,</span>
<span>  marker <span class="op">=</span> <span class="st">"CD20"</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  shape <span class="op">=</span> <span class="fl">19</span>,</span>
<span>  interactive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plotModelFit <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot the image</span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">image</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="06-changes_in_marker_expression_files/figure-html/06-plotStateChanges CD20-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># plot the scatter plot</span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">scatter</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 26 rows containing missing values or values outside the scale range
(`geom_smooth()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="06-changes_in_marker_expression_files/figure-html/06-scatter plot CD20-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>So why are T cells expressing CD20? This brings us to a key problem of cell segmentation - contamination.</p>
</section><section id="contamination-lateral-marker-spill-over" class="level3" data-number="10.2.3"><h3 data-number="10.2.3" class="anchored" data-anchor-id="contamination-lateral-marker-spill-over">
<span class="header-section-number">10.2.3</span> Contamination (Lateral marker spill over)</h3>
<p>Contamination, or lateral marker spill over, is an issue that results in a cell’s marker expressions being wrongly attributed to another adjacent cell. This issue arises from incorrect segmentation where components of one cell are wrongly determined as belonging to another cell. Alternatively, this issue can arise when antibodies used to tag and measure marker expressions don’t latch on properly to a cell of interest, thereby resulting in residual markers being wrongly assigned as belonging to a cell near the intended target cell. It is important that we either correct or account for this incorrect attribution of markers in our modelling process. This is critical in understanding whether significant cell-cell interactions detected are an artefact of technical measurement errors driven by spill over or are real biological changes that represent a shift in a cell’s state.</p>
<p>To circumvent this problem, Statial provides a function that predicts the probability that a cell is any particular cell type - <code>calcContamination</code>. <code>calcContamination</code> returns a dataframe of probabilities demarcating the chance of a cell being any particular cell type. This dataframe is stored under <code>contaminations</code> in the <code>reducedDim</code> slot of the <code>SingleCellExperiment</code> object. It also provides the <code>rfMainCellProb</code> column, which provides the probability that a cell is indeed the cell type it has been designated. For example, for a cell designated as CD8, <code>rfMainCellProb</code> could give a 80% chance that the cell is indeed CD8, due to contamination.</p>
<p>We can then introduce these probabilities as covariates into our linear model by setting <code>contamination = TRUE</code> as a parameter in our <code>calcStateChanges</code> function.</p>
<div class="cell">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">kerenSPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/calcContamination.html">calcContamination</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Growing trees.. Progress: 100%. Estimated remaining time: 0 seconds.</code></pre>
</div>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">stateChangesCorrected</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/calcStateChanges.html">calcStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">kerenSPE</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  nCores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  minCells <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  contamination <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">stateChangesCorrected</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       imageID primaryCellType otherCellType     marker         coef
69468       37     Endothelial        Tumour       Lag3 -0.001621517
150351      11     Neutrophils            NK       CD56 -0.059936866
16402       35      CD4_T_cell        B_cell       CD20 -0.025114581
16498       35      CD4_T_cell    DC_or_Mono       CD20  0.016154346
4891        35          B_cell    DC_or_Mono phospho.S6  0.004312579
16507       35      CD4_T_cell    DC_or_Mono phospho.S6  0.003593907
88516        3  Keratin_Tumour            DC         Ca -0.014151972
16354       35      CD4_T_cell      dn_T_CD3       CD20  0.013728798
16357       35      CD4_T_cell      dn_T_CD3     HLA.DR  0.010329232
3697        28          B_cell            NK         Na -0.004480367
                tval          pval           fdr
69468  -6.030700e+13  0.000000e+00  0.000000e+00
150351 -1.522794e+14  0.000000e+00  0.000000e+00
16402  -3.515983e+01 2.259264e-223 1.379665e-218
16498   3.402099e+01 1.677480e-211 7.682901e-207
4891    2.994786e+01 1.340822e-169 4.912800e-165
16507   2.958522e+01 6.665852e-167 1.959134e-162
88516  -3.004738e+01 7.485733e-167 1.959134e-162
16354   2.955635e+01 1.271559e-166 2.911886e-162
16357   2.886146e+01 6.506249e-160 1.324390e-155
3697   -2.904572e+01 1.101584e-157 2.018113e-153</code></pre>
</div>
</div>
<p>However, this is not a perfect solution for the issue of contamination. As we can see, despite factoring in contamination into our linear model, the correlation between B cell density and CD20 expression in CD4 T cells remains one of the most significant interactions in our model.</p>
<p>However, this does not mean factoring in contamination into our linear model was ineffective.</p>
<p>Whilst our correction attempts do not rectify every relationship which arises due to contamination, we show that a significant portion of these relationships are rectified. We can show this by plotting a ROC curve of true positives against false positives. In general, cell type specific markers such as CD4 (specific to T helper cells), CD8 (specific to cytotoxic T cells), and CD20 should not change in cells they are not specific to. Therefore, relationships detected to be significant involving these cell type markers are likely false positives and will be treated as such for the purposes of evaluation. Meanwhile, cell state markers are predominantly likely to be true positives.</p>
<p>Plotting the relationship between false positives and true positives, we’d expect the contamination correction to be greatest in the relationships with the top 100 lowest p values, where we indeed see more true positives than false positives with contamination correction.</p>
<div class="cell">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cellTypeMarkers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD3"</span>, <span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"CD56"</span>, <span class="st">"CD11c"</span>, <span class="st">"CD68"</span>, <span class="st">"CD45"</span>, <span class="st">"CD20"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">values</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"red"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">values</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"None"</span>, <span class="st">"Corrected"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>TP <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html">cumsum</a></span><span class="op">(</span><span class="va">stateChanges</span><span class="op">$</span><span class="va">marker</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">cellTypeMarkers</span><span class="op">)</span>, </span>
<span>             FP <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html">cumsum</a></span><span class="op">(</span><span class="op">!</span><span class="va">stateChanges</span><span class="op">$</span><span class="va">marker</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">cellTypeMarkers</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"None"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>TP <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html">cumsum</a></span><span class="op">(</span><span class="va">stateChangesCorrected</span><span class="op">$</span><span class="va">marker</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">cellTypeMarkers</span><span class="op">)</span>, </span>
<span>             FP <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html">cumsum</a></span><span class="op">(</span><span class="op">!</span><span class="va">stateChangesCorrected</span><span class="op">$</span><span class="va">marker</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">cellTypeMarkers</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"Corrected"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">TP</span>, y <span class="op">=</span> <span class="va">FP</span>, colour <span class="op">=</span> <span class="va">type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Cell state marker (FP)"</span>, x <span class="op">=</span> <span class="st">"Cell type marker (TP)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">values</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="06-changes_in_marker_expression_files/figure-html/06-ROC curve-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Below, we zoom in on the ROC curve where the top 100 lowest p values occur, where we indeed see more true positives than false positives with contamination correction.</p>
<div class="cell">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">TP</span>, y <span class="op">=</span> <span class="va">FP</span>, colour <span class="op">=</span> <span class="va">type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">xlim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">100</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">ylim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1000</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Cell state marker (FP)"</span>, x <span class="op">=</span> <span class="st">"Cell type marker (TP)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">values</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="06-changes_in_marker_expression_files/figure-html/06-top 100 ROC-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="associate-continuous-state-changes-with-survival-outcomes" class="level3" data-number="10.2.4"><h3 data-number="10.2.4" class="anchored" data-anchor-id="associate-continuous-state-changes-with-survival-outcomes">
<span class="header-section-number">10.2.4</span> Associate continuous state changes with survival outcomes</h3>
<p>Similiar to <code>Kontextual</code>, we can run a similar survival analysis using our <code>SpatioMark</code> results. Here, <code>prepMatrix</code> extracts the coefficients, or the <code>coef</code> column of <code>stateChanges</code> by default. To use the t-values instead, specify <code>column = "tval"</code> in the <code>prepMatrix</code> function. As before, we use <code>colTest</code> to build the CoxPH model.</p>
<div class="cell" data-time_it="true">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Preparing features for Statial</span></span>
<span><span class="va">stateMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/prepMatrix.html">prepMatrix</a></span><span class="op">(</span><span class="va">stateChanges</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Ensuring rownames of stateMat match up with rownames of the survival vector</span></span>
<span><span class="va">stateMat</span> <span class="op">&lt;-</span> <span class="va">stateMat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">kerenSurv</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Remove some very small values</span></span>
<span><span class="va">stateMat</span> <span class="op">&lt;-</span> <span class="va">stateMat</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">stateMat</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0.0001</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">.8</span><span class="op">]</span></span>
<span></span>
<span><span class="va">survivalResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/colTest.html">colTest</a></span><span class="op">(</span><span class="va">stateMat</span>, <span class="va">kerenSurv</span>, type <span class="op">=</span> <span class="st">"survival"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">survivalResults</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                         coef se.coef   pval adjPval
Keratin_Tumour__Mono_or_Neu__Pan.Keratin -280      89 0.0018    0.63
Macrophages__Keratin_Tumour__HLA_Class_1  220      75 0.0034    0.63
Keratin_Tumour__CD8_T_cell__Keratin6     -220      77 0.0036    0.63
Macrophages__Other_Immune__HLA_Class_1   -480     170 0.0057    0.75
Keratin_Tumour__Mesenchymal__dsDNA       -810     310 0.0094    0.80
Keratin_Tumour__Unidentified__H3K27me3    490     190 0.0100    0.80
                                                                          cluster
Keratin_Tumour__Mono_or_Neu__Pan.Keratin Keratin_Tumour__Mono_or_Neu__Pan.Keratin
Macrophages__Keratin_Tumour__HLA_Class_1 Macrophages__Keratin_Tumour__HLA_Class_1
Keratin_Tumour__CD8_T_cell__Keratin6         Keratin_Tumour__CD8_T_cell__Keratin6
Macrophages__Other_Immune__HLA_Class_1     Macrophages__Other_Immune__HLA_Class_1
Keratin_Tumour__Mesenchymal__dsDNA             Keratin_Tumour__Mesenchymal__dsDNA
Keratin_Tumour__Unidentified__H3K27me3     Keratin_Tumour__Unidentified__H3K27me3</code></pre>
</div>
<div style="text-align: right;">
<em>Time for this code chunk to run with 5.5 cores: 0.18 s
</em>
</div>
<p></p>
</div>
<p><code>Keratin_Tumour__Mono_or_Neu__Pan.Keratin</code> is the most significant pairwise relationship which contributes to patient survival. That is, the relationship between pan-keratin expression in keratin/tumour cells and their spatial proximity to monocytes/neutrophils. The negative coefficient associated with this relationship tells us that higher pan-keratin expression in keratin/tumour cells nearby monocyte/neutrophil cell populations leads to better survival outcomes for patients.</p>
<div class="cell">
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Selecting the most significant relationship</span></span>
<span><span class="va">survRelationship</span> <span class="op">&lt;-</span> <span class="va">stateMat</span><span class="op">[[</span><span class="st">"Keratin_Tumour__Mono_or_Neu__Pan.Keratin"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">survRelationship</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">survRelationship</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">survRelationship</span><span class="op">)</span>, <span class="st">"Higher expression in close cells"</span>, <span class="st">"Lower expression in close cells"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plotting Kaplan-Meier curve</span></span>
<span><span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/survfit2.html">survfit2</a></span><span class="op">(</span><span class="va">kerenSurv</span> <span class="op">~</span> <span class="va">survRelationship</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/ggsurvfit.html">ggsurvfit</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="http://www.danieldsjoberg.com/ggsurvfit/reference/add_pvalue.html">add_pvalue</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Keratin_Tumour__Mono_or_Neu__Pan.Keratin"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="06-changes_in_marker_expression_files/figure-html/06-Km curve with stateChanges-1.png" class="img-fluid figure-img" width="480"></p>
</figure>
</div>
</div>
</div>
<!--## scFeatures: Moran's I-->
<p>We conclude the section on spatial quantification metrics here. So far, we have identified 7 metrics to quantify spatial relationships -</p>
<ol type="1">
<li>Cell type proportions (FuseSOM)</li>
<li>Co-localisation between pairs of cell types using the L-function (spicyR)</li>
<li>Cell type co-localisation with respect to a parent population using <code>Kontextual</code> (Statial)</li>
<li>Regions of co-localisation, or spatial domains (lisaClust)</li>
<li>Marker means in each cell type (Statial)</li>
<li>Marker means in each cell type in each region (Statial)</li>
<li>Proximity-associated changes in marker expression using SpatioMark (Statial)</li>
</ol></section></section><section id="sessioninfo" class="level2" data-number="10.3"><h2 data-number="10.3" class="anchored" data-anchor-id="sessioninfo">
<span class="header-section-number">10.3</span> sessionInfo</h2>
<div class="cell">
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.5.0 (2025-04-11)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.4.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.1

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Australia/Sydney
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] SpatialDatasets_1.6.3       SpatialExperiment_1.18.1   
 [3] ExperimentHub_2.16.0        AnnotationHub_3.16.0       
 [5] BiocFileCache_2.16.0        dbplyr_2.5.0               
 [7] treekoR_1.16.0              tibble_3.2.1               
 [9] ggsurvfit_1.1.0             ggplot2_3.5.2              
[11] SingleCellExperiment_1.30.1 dplyr_1.1.4                
[13] lisaClust_1.16.0            ClassifyR_3.12.0           
[15] survival_3.8-3              BiocParallel_1.42.0        
[17] MultiAssayExperiment_1.34.0 SummarizedExperiment_1.38.1
[19] Biobase_2.68.0              GenomicRanges_1.60.0       
[21] GenomeInfoDb_1.44.0         IRanges_2.42.0             
[23] MatrixGenerics_1.20.0       matrixStats_1.5.0          
[25] S4Vectors_0.46.0            BiocGenerics_0.54.0        
[27] generics_0.1.4              spicyR_1.21.1              
[29] Statial_1.10.0             

loaded via a namespace (and not attached):
  [1] fs_1.6.6                    spatstat.sparse_3.1-0      
  [3] bitops_1.0-9                EBImage_4.50.0             
  [5] httr_1.4.7                  hopach_2.68.0              
  [7] RColorBrewer_1.1-3          doParallel_1.0.17          
  [9] numDeriv_2016.8-1.1         tools_4.5.0                
 [11] doRNG_1.8.6.2               backports_1.5.0            
 [13] R6_2.6.1                    HDF5Array_1.36.0           
 [15] lazyeval_0.2.2              mgcv_1.9-3                 
 [17] rhdf5filters_1.20.0         GetoptLong_1.0.5           
 [19] sp_2.2-0                    withr_3.0.2                
 [21] gridExtra_2.3               coxme_2.2-22               
 [23] textshaping_1.0.1           cli_3.6.5                  
 [25] spatstat.explore_3.4-3      sandwich_3.1-1             
 [27] labeling_0.4.3              nnls_1.6                   
 [29] mvtnorm_1.3-3               spatstat.data_3.1-6        
 [31] systemfonts_1.2.3           yulab.utils_0.2.0          
 [33] ggupset_0.4.1               svglite_2.2.1              
 [35] colorRamps_2.3.4            dichromat_2.0-0.1          
 [37] limma_3.64.1                RSQLite_2.3.11             
 [39] flowCore_2.20.0             rstudioapi_0.17.1          
 [41] simpleSeg_1.10.0            gridGraphics_0.5-1         
 [43] shape_1.4.6.1               spatstat.random_3.4-1      
 [45] car_3.1-3                   scam_1.2-19                
 [47] Matrix_1.7-3                RProtoBufLib_2.20.0        
 [49] ggbeeswarm_0.7.2            abind_1.4-8                
 [51] terra_1.8-50                lifecycle_1.0.4            
 [53] yaml_2.3.10                 multcomp_1.4-28            
 [55] edgeR_4.6.2                 carData_3.0-5              
 [57] rhdf5_2.52.0                SparseArray_1.8.0          
 [59] Rtsne_0.17                  blob_1.2.4                 
 [61] grid_4.5.0                  promises_1.3.2             
 [63] shinydashboard_0.7.3        crayon_1.5.3               
 [65] bdsmatrix_1.3-7             lattice_0.22-6             
 [67] KEGGREST_1.48.0             magick_2.8.5               
 [69] cytomapper_1.20.0           pillar_1.10.2              
 [71] knitr_1.50                  ComplexHeatmap_2.24.0      
 [73] dcanr_1.24.0                rjson_0.2.23               
 [75] boot_1.3-31                 codetools_0.2-20           
 [77] glue_1.8.0                  ggiraph_0.8.13             
 [79] ggfun_0.1.8                 spatstat.univar_3.1-3      
 [81] data.table_1.17.4           vctrs_0.6.5                
 [83] png_0.1-8                   treeio_1.32.0              
 [85] Rdpack_2.6.4                gtable_0.3.6               
 [87] cachem_1.1.0                xfun_0.52                  
 [89] mime_0.13                   rbibutils_2.3              
 [91] S4Arrays_1.8.0              ConsensusClusterPlus_1.72.0
 [93] reformulas_0.4.1            pheatmap_1.0.12            
 [95] iterators_1.0.14            cytolib_2.20.0             
 [97] statmod_1.5.0               TH.data_1.1-3              
 [99] nlme_3.1-168                ggtree_3.16.0              
[101] bit64_4.6.0-1               filelock_1.0.3             
[103] svgPanZoom_0.3.4            vipor_0.4.7                
[105] DBI_1.2.3                   colorspace_2.1-1           
[107] raster_3.6-32               tidyselect_1.2.1           
[109] curl_6.2.3                  bit_4.6.0                  
[111] compiler_4.5.0              diffcyt_1.28.0             
[113] h5mread_1.0.1               DelayedArray_0.34.1        
[115] plotly_4.10.4               scales_1.4.0               
[117] rappdirs_0.3.3              tiff_0.1-12                
[119] stringr_1.5.1               digest_0.6.37              
[121] goftest_1.2-3               fftwtools_0.9-11           
[123] spatstat.utils_3.1-4        minqa_1.2.8                
[125] rmarkdown_2.29              XVector_0.48.0             
[127] htmltools_0.5.8.1           pkgconfig_2.0.3            
[129] jpeg_0.1-11                 lme4_1.1-37                
[131] fastmap_1.2.0               rlang_1.1.6                
[133] GlobalOptions_0.1.2         htmlwidgets_1.6.4          
[135] ggthemes_5.1.0              UCSC.utils_1.4.0           
[137] shiny_1.10.0                ggh4x_0.3.0                
[139] farver_2.1.2                zoo_1.8-14                 
[141] jsonlite_2.0.0              RCurl_1.98-1.17            
[143] magrittr_2.0.3              Formula_1.2-5              
[145] GenomeInfoDbData_1.2.14     ggplotify_0.1.2            
[147] patchwork_1.3.0             Rhdf5lib_1.30.0            
[149] Rcpp_1.0.14                 viridis_0.6.5              
[151] ape_5.8-1                   ggnewscale_0.5.1           
[153] stringi_1.8.7               MASS_7.3-65                
[155] plyr_1.8.9                  parallel_4.5.0             
[157] deldir_2.0-4                Biostrings_2.76.0          
[159] splines_4.5.0               tensor_1.5                 
[161] circlize_0.4.16             locfit_1.5-9.12            
[163] igraph_2.1.4                ggpubr_0.6.0               
[165] uuid_1.2-1                  ranger_0.17.0              
[167] spatstat.geom_3.4-1         ggsignif_0.6.4             
[169] rngtools_1.5.2              reshape2_1.4.4             
[171] BiocVersion_3.21.1          XML_3.99-0.18              
[173] evaluate_1.0.3              BiocManager_1.30.25        
[175] nloptr_2.2.1                foreach_1.5.2              
[177] tweenr_2.0.3                httpuv_1.6.16              
[179] tidyr_1.3.1                 purrr_1.0.4                
[181] polyclip_1.10-7             clue_0.3-66                
[183] BiocBaseUtils_1.10.0        ggforce_0.4.2              
[185] xtable_1.8-4                broom_1.0.8                
[187] tidytree_0.4.6              rstatix_0.7.2              
[189] later_1.4.2                 viridisLite_0.4.2          
[191] class_7.3-23                lmerTest_3.1-3             
[193] aplot_0.2.5                 AnnotationDbi_1.70.0       
[195] memoise_2.0.1               beeswarm_0.4.0             
[197] FlowSOM_2.16.0              cluster_2.1.8.1            
[199] concaveman_1.1.0           </code></pre>
</div>
</div>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./05-cellular_niches.html" class="pagination-link" aria-label="Identifying spatial domains with unsupervised clustering">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Identifying spatial domains with unsupervised clustering</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./07-classification.html" class="pagination-link" aria-label="Finding associations between clinical variables and spatial features">
        <span class="nav-page-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Finding associations between clinical variables and spatial features</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>