<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>11&nbsp; Finding associations between clinical variables and spatial features â€“ Spatial analysis playbook</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./08-case_study1.html" rel="next">
<link href="./06-changes_in_marker_expression.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script><script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script><link rel="stylesheet" href="styles.css">
</head>
<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="./index.html" class="navbar-brand navbar-brand-logo">
    <img src="./images/USydLogo.svg" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Spatial analysis playbook</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./07-classification.html">Classification</a></li><li class="breadcrumb-item"><a href="./07-classification.html"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Finding associations between clinical variables and spatial features</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Overview</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Datasets</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./getting_started.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Getting Started</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Processing</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01-processing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Cell segmentation and pre-processing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02-quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Quality Control</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Cell annotation</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03a-cell_annotation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Unsupervised clustering for cell annotation</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Cell localisation</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04a-cell_localisation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Cell localisation between pairs of cell types</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04b-cell_localisation_parent.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Cell relationships relative to expected behaviour</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">Spatial domains</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05-cellular_niches.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Identifying spatial domains with unsupervised clustering</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true">
 <span class="menu-text">Marker expression</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06-changes_in_marker_expression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Changes in marker expression</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true">
 <span class="menu-text">Classification</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07-classification.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Finding associations between clinical variables and spatial features</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" role="navigation" aria-expanded="true">
 <span class="menu-text">Case studies</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-8" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08-case_study1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Case Study: Head and Neck Squamous Cell Carcinoma (Ferguson et al., 2022)</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#introduction-to-classifyr" id="toc-introduction-to-classifyr" class="nav-link active" data-scroll-target="#introduction-to-classifyr"><span class="header-section-number">11.1</span> Introduction to <code>ClassifyR</code></a></li>
  <li><a href="#introduction-to-scfeatures" id="toc-introduction-to-scfeatures" class="nav-link" data-scroll-target="#introduction-to-scfeatures"><span class="header-section-number">11.2</span> Introduction to <code>scFeatures</code></a></li>
  <li><a href="#classification-of-patients-by-condition" id="toc-classification-of-patients-by-condition" class="nav-link" data-scroll-target="#classification-of-patients-by-condition"><span class="header-section-number">11.3</span> Classification of patients by condition</a></li>
  <li><a href="#classification-of-patients-by-survival" id="toc-classification-of-patients-by-survival" class="nav-link" data-scroll-target="#classification-of-patients-by-survival"><span class="header-section-number">11.4</span> Classification of patients by survival</a></li>
  <li>
<a href="#easy-and-hard-to-classify-patients" id="toc-easy-and-hard-to-classify-patients" class="nav-link" data-scroll-target="#easy-and-hard-to-classify-patients"><span class="header-section-number">11.5</span> Easy and hard to classify patients</a>
  <ul class="collapse">
<li><a href="#predicting-tumour-type" id="toc-predicting-tumour-type" class="nav-link" data-scroll-target="#predicting-tumour-type"><span class="header-section-number">11.5.1</span> Predicting tumour type</a></li>
  <li><a href="#predicting-survival-outcomes" id="toc-predicting-survival-outcomes" class="nav-link" data-scroll-target="#predicting-survival-outcomes"><span class="header-section-number">11.5.2</span> Predicting survival outcomes</a></li>
  </ul>
</li>
  <li><a href="#sessioninfo" id="toc-sessioninfo" class="nav-link" data-scroll-target="#sessioninfo"><span class="header-section-number">12</span> sessionInfo</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./07-classification.html">Classification</a></li><li class="breadcrumb-item"><a href="./07-classification.html"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Finding associations between clinical variables and spatial features</span></a></li></ol></nav><div class="quarto-title">
<h1 class="title">
<span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Finding associations between clinical variables and spatial features</span>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><p>Classification models play a central role in spatial omics by identifying features that distinguish between patient groupsâ€”advancing both mechanistic understanding and clinical stratification. These distinguishing features can capture a wide range of biological signals, from differences in tissue architecture and cell-cell interactions to variations in cell state or microenvironment composition. Depending on the aim, analyses may seek to explain the underlying biology of a condition or to build predictive models for clinical outcomes such as progression, metastasis, or therapeutic response.</p>
<p>However, due to biological heterogeneity, not all features are informative across all individuals. To address this, robust modelling frameworks must accommodate complex, subgroup-specific signals. This ensures that models not only perform well statistically but also yield biologically and clinically meaningful insights. <code>ClassifyR</code> provides such a framework, offering tools for feature selection, repeated cross-validation, performance evaluation, survival analysis, and integration across multiple data types. It is designed to rigorously assess which features predict outcomes, and for which subgroups, thereby enhancing both discovery and translational potential.</p>
<p>To build these models, it is essential to first summarise each tissue sample in terms of meaningful, comparable features. This is where the <code>scFeatures</code> package plays a complementary role. <code>scFeatures</code> generates a rich and diverse set of sample-level features that capture both spatial and non-spatial characteristicsâ€”including spatial organisation, cell-type proportions, expression variability, pathway activity, and more. Many of these features are derived from the spatial metrics computed earlier in this playbook. By creating a comprehensive feature set, <code>scFeatures</code> increases the likelihood of identifying biologically relevant patterns and supports multi-view modelling strategies.</p>
<p>In this section, we demonstrate how <code>scFeatures</code> can be used to generate interpretable, high-dimensional representations of spatial omics data, and how these can be combined with <code>ClassifyR</code> to construct and evaluate predictive models of patient outcome.</p>
<!-- Steps: -->
<!-- 1.  Introduction to ClassifyR -->
<!-- 2.  Classification of patients by condition -->
<!-- 3.  Classification of patients by survival -->
<!-- 4.  Easy and Hard to classify patients (samplesMetricMap) -->
<!-- 5.  Maximising accuracy during classification (parameter tuning for crossValidate) -->
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/ClassifyR/">ClassifyR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ellispatrick.github.io/lisaClust/">lisaClust</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/Statial">Statial</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ellispatrick.github.io/spicyR/">spicyR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/scFeatures/">scFeatures</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># set parameters</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># whether to use multiple cores (recommended)</span></span>
<span><span class="va">use_mc</span> <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="va">is_windows</span> <span class="op">=</span> <span class="va">.Platform</span><span class="op">$</span><span class="va">OS.type</span> <span class="op">==</span> <span class="st">"windows"</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">use_mc</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">nCores</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html">ceiling</a></span><span class="op">(</span><span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span><span class="op">)</span>, <span class="fl">1</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">nCores</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">BPPARAM</span> <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html">SerialParam</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">is_windows</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">BPPARAM</span> <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html">SnowParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="va">nCores</span>, type <span class="op">=</span> <span class="st">"SOCK"</span><span class="op">)</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">BPPARAM</span> <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="va">nCores</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="va">BPPARAM</span> <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html">SerialParam</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="introduction-to-classifyr" class="level2" data-number="11.1"><h2 data-number="11.1" class="anchored" data-anchor-id="introduction-to-classifyr">
<span class="header-section-number">11.1</span> Introduction to <code>ClassifyR</code>
</h2>
<p><a href="https://www.bioconductor.org/packages/release/bioc/html/ClassifyR.html">ClassifyR</a> provides a structured pipeline for cross-validated classification. Classification is viewed in terms of four stages: data transformation, feature selection, classifier training, and prediction. The driver functions <code>crossValidate</code> and <code>runTests</code> implements varieties of cross-validation. They are:</p>
<ul>
<li>Permutation of the order of samples followed by k-fold cross-validation</li>
<li>Repeated <span class="math inline">\(x\%\)</span> test set cross-validation</li>
<li>leave-k-out cross-validation</li>
</ul>
<p>We will use the Keren 2018 dataset to perform two predictions: 1) predict the patientâ€™s tumour type (compartmentalised or mixed) and 2) predict the patientâ€™s survival outcome.</p>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># load the Keren 2018 dataset</span></span>
<span><span class="va">kerenSPE</span> <span class="op">&lt;-</span> <span class="fu">SpatialDatasets</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialDatasets/man/SpatialDatasets.html">spe_Keren_2018</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># remove any missing data in our outcome columns</span></span>
<span><span class="va">kerenSPE</span> <span class="op">=</span> <span class="va">kerenSPE</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/stats/complete.cases.html">complete.cases</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Censored"</span>, <span class="st">"Survival_days_capped*"</span>,</span>
<span>                                                           <span class="st">"tumour_type"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co"># load pre-computed data for efficiency</span></span>
<span><span class="va">kontextMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"data/kontextMat.rds"</span><span class="op">)</span></span>
<span><span class="va">stateMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"data/stateMat.rds"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="introduction-to-scfeatures" class="level2" data-number="11.2"><h2 data-number="11.2" class="anchored" data-anchor-id="introduction-to-scfeatures">
<span class="header-section-number">11.2</span> Introduction to <code>scFeatures</code>
</h2>
<p><a href="https://www.bioconductor.org/packages/release/bioc/html/scFeatures.html">scFeatures</a> allows us to generate up to 17 sample-level metrics, including cell type proportions, cell type-specific marker expressions, cell type-specific pathway expressions, cell type-specific cell-cell communications, and overall aggregated marker expressions.</p>
<p><img src="images/scFeatures_fig1.jpeg" align="center" style="height: 300px; border: 0px"></p>
<p>Here, we will use <code>scFeatures</code> to generate two sample-level metrics: <code>proportion_raw</code>, the proportion of cell types within each sample or image, and <code>gene_mean_celltype</code>, which captures cell type-specific marker expression.</p>
</section><section id="classification-of-patients-by-condition" class="level2" data-number="11.3"><h2 data-number="11.3" class="anchored" data-anchor-id="classification-of-patients-by-condition">
<span class="header-section-number">11.3</span> Classification of patients by condition</h2>
<p>We will use the Keren 2018 dataset to classify a patientâ€™s tumour into one of two types: compartmentalised and mixed. First, we will filter out all cases which had cold tumours.</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># filter out cold tumours</span></span>
<span><span class="va">kerenSPE</span> <span class="op">=</span> <span class="va">kerenSPE</span><span class="op">[</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span><span class="op">$</span><span class="va">tumour_type</span> <span class="op">!=</span> <span class="st">"cold"</span><span class="op">]</span></span>
<span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">tumour_type</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/droplevels.html">droplevels</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">tumour_type</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Factor-class.html">levels</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">tumour_type</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"compartmentalised"</span>, <span class="st">"mixed"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">kontextMat</span> <span class="op">=</span> <span class="va">kontextMat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">kontextMat</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">imageID</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">stateMat</span> <span class="op">=</span> <span class="va">stateMat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">stateMat</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">imageID</span><span class="op">)</span>, <span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We will build a list of feature matrices using the features weâ€™ve generated in the previous chapters, including the two metrics generated through <code>scFeatures</code>:</p>
<ol type="1">
<li>Co-localisation between pairs of cell types using the L-function (<code>spicyR</code>)</li>
<li>Cell type co-localisation with respect to a parent population using <code>Kontextual</code> (<code>Statial</code>)</li>
<li>Regions of co-localisation, or spatial domains (<code>lisaClust</code>)</li>
<li>Marker means in each cell type (<code>Statial</code>)</li>
<li>Proximity-associated changes in marker expression using SpatioMark (<code>Statial</code>)</li>
<li>Cell type proportions (<code>scFeatures</code>)</li>
<li>Cell type-specific marker expression (<code>scFeatures</code>)</li>
</ol>
<div class="cell" data-time_it="true">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># pre-generated spatial metrics</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Cell localisation (spicyR)</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"spicyR"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/getPairwise.html">getPairwise</a></span><span class="op">(</span><span class="va">kerenSPE</span>,</span>
<span>                                BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The `BPPARAM` argument of `getPairwise()` is deprecated as of spicyR 1.18.0.
â„¹ Please use the `cores` argument instead.
â„¹ The deprecated feature was likely used in the spicyR package.
  Please report the issue at &lt;https://github.com/SydneyBioX/spicyR/issues&gt;.</code></pre>
</div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Cell localisation with respect to a parent (Kontextual)</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"Kontextual"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">kontextMat</span></span>
<span></span>
<span><span class="co"># Spatial Domains (lisaClust)</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"lisaClust"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/getProp.html">getProp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/lisaClust.html">lisaClust</a></span><span class="op">(</span><span class="va">kerenSPE</span>, k <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                                         BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span><span class="op">)</span>, <span class="st">"region"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The `BPPARAM` argument of `lisaClust()` is deprecated as of lisaClust 1.14.4.
â„¹ Please use the `cores` argument instead.
â„¹ The deprecated feature was likely used in the lisaClust package.
  Please report the issue at
  &lt;https://github.com/ellispatrick/lisaClust/issues&gt;.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Generating local L-curves.</code></pre>
</div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Marker means in each cell type (Statial)</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"cellTypeMarkerMeans"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/getMarkerMeans.html">getMarkerMeans</a></span><span class="op">(</span><span class="va">kerenSPE</span>, imageID <span class="op">=</span> <span class="st">"imageID"</span>,</span>
<span>                                      cellType <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>                                        region <span class="op">=</span> <span class="st">"cellType"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Proximity-associated changes in marker expression (SpatioMark)</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"SpatioMark"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">stateMat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div style="text-align: right;">
<em>Time for this code chunk to run with 6 cores: 25.19 s
</em>
</div>
<p></p>
</div>
<p>We will now use <code>scFeatures</code> to generate our two selected features. Below, we use the <code>scFeatures</code> function and specify <code>type = "spatial_p"</code> to indicate we are using spatial proteomics data.</p>
<div class="cell">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># generate assay matrix and add spatial features</span></span>
<span><span class="va">keren_mat</span> <span class="op">=</span> <span class="fu">assay</span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># generate spatial coords vectors</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu">spatialCoords</span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">as.vector</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">=</span> <span class="fu">spatialCoords</span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">)</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">as.vector</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># scFeatures</span></span>
<span><span class="va">scfeatures_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/scFeatures/reference/scFeatures.html">scFeatures</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">keren_mat</span>, </span>
<span>                                sample <span class="op">=</span> <span class="va">kerenSPE</span><span class="op">$</span><span class="va">imageID</span>, </span>
<span>                                celltype <span class="op">=</span> <span class="va">kerenSPE</span><span class="op">$</span><span class="va">cellType</span>,</span>
<span>                                spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span>,</span>
<span>                                feature_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"proportion_raw"</span>, <span class="st">"gene_mean_celltype"</span><span class="op">)</span>  , </span>
<span>                                type <span class="op">=</span> <span class="st">"spatial_p"</span>,  </span>
<span>                                ncores <span class="op">=</span> <span class="va">nCores</span>,  </span>
<span>                                species <span class="op">=</span> <span class="st">"Homo sapiens"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>generating proportion raw features</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: spatial_p does not currently support parallel computation</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>generating gene mean celltype features</code></pre>
</div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># add to our list of feature matrices</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"proportions"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">proportion_raw</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"cellTypeMarkerMeans"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">gene_mean_celltype</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We will then generate a factor vector of our outcome variable.</p>
<div class="cell">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># outcome vector</span></span>
<span><span class="va">outcome</span> <span class="op">=</span> <span class="va">kerenSPE</span><span class="op">$</span><span class="va">tumour_type</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">imageID</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">outcome</span><span class="op">)</span> <span class="op">=</span> <span class="va">kerenSPE</span><span class="op">$</span><span class="va">imageID</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">imageID</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">outcome</span>, <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                1                 2                 3                 4 
            mixed             mixed compartmentalised compartmentalised 
                5 
compartmentalised 
Levels: compartmentalised mixed</code></pre>
</div>
</div>
<p><code>ClassifyR</code> provides a convenient function, <code>crossValidate</code>, to build and test models. <code>crossValidate</code> must be supplied with <code>measurements</code>, a simple tabular data container or a list-like structure of such related tabular data on common samples. It can be in the form of a <code>matrix</code>, <code>data.frame</code>, <code>DataFrame</code>, <code>MultiAssayExperiment</code> or a list of <code>data.frames</code>.</p>
<p><code>crossValidate</code> must also be supplied with <code>outcome</code>, which represents the prediction to be made. <code>outcome</code> can be either a factor containing the class labels for each observation, or a <code>character</code> of length 1 that matches a column name in <code>measurements</code> which holds the classes. If a <code>character</code> is provided, <code>crossValidate</code> will automatically remove the classes before training.</p>
<p>By default, <code>crossValidate</code> will build and train a random forest. Alternative classifiers can be specified using the <code>classifier</code> argument. To view all available feature selection and classification approaches, use the <code>available</code> function.</p>
<div class="cell" data-time_it="true">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># perform 50 repeats of 5-fold cross-validation</span></span>
<span><span class="va">cv</span> <span class="op">=</span> <span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/crossValidate.html">crossValidate</a></span><span class="op">(</span>measurements <span class="op">=</span> <span class="va">data</span>,</span>
<span>                   outcome <span class="op">=</span> <span class="va">outcome</span>,</span>
<span>                   nFolds <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                   nRepeats <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                   nCores <span class="op">=</span> <span class="va">nCores</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div style="text-align: right;">
<em>Time for this code chunk to run with 6 cores: 76.01 s
</em>
</div>
<p></p>
</div>
<p>We can use <code>performancePlot</code> to visualise performance metrics for all our features. Here, we visualise the AUC for each of the seven feature matrices we tested. Additional performance metrics can be specified in the <code>metric</code> argument.</p>
<div class="cell">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/performancePlot.html">performancePlot</a></span><span class="op">(</span></span>
<span>  <span class="va">cv</span>,</span>
<span>  metric <span class="op">=</span> <span class="st">"AUC"</span>,</span>
<span>  characteristicsList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Assay Name"</span><span class="op">)</span>,</span>
<span>  orderingList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"Assay Name"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"spicyR"</span>, <span class="st">"lisaClust"</span>, <span class="st">"Kontextual"</span>, <span class="st">"SpatioMark"</span>, <span class="st">"proportions"</span>, <span class="st">"cellTypeMarkerMeans"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="07-classification_files/figure-html/07-performance plot-1.png" class="img-fluid figure-img" width="1344"></p>
</figure>
</div>
</div>
</div>
<p>From the graph, both <code>lisaClust</code> and <code>proportions</code> appear to capture information which is predictive of the tumour type of patients.</p>
</section><section id="classification-of-patients-by-survival" class="level2" data-number="11.4"><h2 data-number="11.4" class="anchored" data-anchor-id="classification-of-patients-by-survival">
<span class="header-section-number">11.4</span> Classification of patients by survival</h2>
<p><code>crossValidate</code> also has the capacity to test classification performance for a survival outcome. In this case, <code>outcome</code> must be a <code>Surv</code> object of the same length as the number of samples in the feature matrix and should contain information about the time and censoring of the samples. Alternatively, we can specify <code>outcome</code> to be a <code>character</code> of length 2 or 3 that each match a column name in a data frame which holds information about the time and censoring of the samples. The time-to-event columns will automatically be removed before training is done.</p>
<p>We will first add a <code>survival</code> column to the <code>kerenSPE</code> object.</p>
<div class="cell">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># create a Surv object named "survival"</span></span>
<span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">event</span> <span class="op">=</span> <span class="fl">1</span> <span class="op">-</span> <span class="va">kerenSPE</span><span class="op">$</span><span class="va">Censored</span></span>
<span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">survival</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html">Surv</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">`Survival_days_capped*`</span>, <span class="va">kerenSPE</span><span class="op">$</span><span class="va">event</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># outcome vector</span></span>
<span><span class="va">surv_outcome</span> <span class="op">=</span> <span class="va">kerenSPE</span><span class="op">$</span><span class="va">survival</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">imageID</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">surv_outcome</span><span class="op">)</span> <span class="op">=</span> <span class="va">kerenSPE</span><span class="op">$</span><span class="va">imageID</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">kerenSPE</span><span class="op">$</span><span class="va">imageID</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">surv_outcome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    1     2     3     4     5     6     7     8     9    10    11    12    13 
 2612   745 3130+ 2523+ 1683+ 2275+   584   946 3767+ 3822+ 3774+ 4353+  1072 
   14    16    17    18    20    21    23    27    28    29    31    32    33 
4145+   530  2842 5063+ 4761+   635    91  3658 3767+  1319  1009 1568+ 1738+ 
   34    35    36    37    39    40    41 
2832+ 2759+ 3063+ 2853+ 2096+  3573 3355+ </code></pre>
</div>
</div>
<p>We can then run <code>crossValidate</code> and specify the outcome to be <code>surv_outcome</code>, and use <code>performancePlot</code> to visualise the performance of the cross-validation. Since we are performing survival analysis, we will specify <code>metric = "C-index"</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># perform 50 repeats of 5-fold cross-validation</span></span>
<span><span class="va">surv_cv</span> <span class="op">=</span> <span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/crossValidate.html">crossValidate</a></span><span class="op">(</span>measurements <span class="op">=</span> <span class="va">data</span>,</span>
<span>                   outcome <span class="op">=</span> <span class="va">surv_outcome</span>,</span>
<span>                   nFolds <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                   nRepeats <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                   nCores <span class="op">=</span> <span class="va">nCores</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/performancePlot.html">performancePlot</a></span><span class="op">(</span><span class="va">surv_cv</span>,</span>
<span>  metric <span class="op">=</span> <span class="st">"C-index"</span>,</span>
<span>  characteristicsList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Assay Name"</span><span class="op">)</span>,</span>
<span>  orderingList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"Assay Name"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"spicyR"</span>, <span class="st">"lisaClust"</span>, <span class="st">"Kontextual"</span>, <span class="st">"SpatioMark"</span>, <span class="st">"proportions"</span>, <span class="st">"cellTypeMarkerMeans"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="07-classification_files/figure-html/07-survival cv-1.png" class="img-fluid figure-img" width="1152"></p>
</figure>
</div>
</div>
</div>
<p>From the graph, we can see that <code>lisaClust</code> appears to capture information that is predictive of survival outcomes comparatively well.</p>
</section><section id="easy-and-hard-to-classify-patients" class="level2" data-number="11.5"><h2 data-number="11.5" class="anchored" data-anchor-id="easy-and-hard-to-classify-patients">
<span class="header-section-number">11.5</span> Easy and hard to classify patients</h2>
<p>The <code>samplesMetricMap</code> function allows visual comparison of sample-wise error rates or accuracy measures from the cross-validation process, helping to identify which samples are consistently well-classified and which are not.</p>
<section id="predicting-tumour-type" class="level3" data-number="11.5.1"><h3 data-number="11.5.1" class="anchored" data-anchor-id="predicting-tumour-type">
<span class="header-section-number">11.5.1</span> Predicting tumour type</h3>
<div class="cell">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/samplesMetricMap.html">samplesMetricMap</a></span><span class="op">(</span><span class="va">cv</span>,  </span>
<span>                 classColours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#3F48CC"</span>, <span class="st">"#880015"</span><span class="op">)</span>,</span>
<span>                 metricColours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#FFFFFF"</span>, <span class="st">"#CFD1F2"</span>, <span class="st">"#9FA3E5"</span>, <span class="st">"#6F75D8"</span>, <span class="st">"#3F48CC"</span><span class="op">)</span>,</span>
<span>                                      <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#FFFFFF"</span>, <span class="st">"#E1BFC4"</span>, <span class="st">"#C37F8A"</span>, <span class="st">"#A53F4F"</span>, <span class="st">"#880015"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="07-classification_files/figure-html/07-samples metric map-1.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>TableGrob (2 x 1) "arrange": 2 grobs
  z     cells    name                grob
1 1 (2-2,1-1) arrange      gtable[layout]
2 2 (1-1,1-1) arrange text[GRID.text.399]</code></pre>
</div>
</div>
<p>This plot highlights samples that are challenging to classify, providing an opportunity to investigate sources of ambiguity. For instance, patients 36 and 13 were particularly difficult to classify. These cases might benefit from additional clinical review or closer inspection of spatial and molecular features not captured in the current feature set. Overall, both cell type proportions (<code>proportion_raw</code>) and spatial domains (<code>lisaClust</code>) performed well for tumour type classification. Interestingly, compartmentalised tumours were generally easier to classify than mixed tumours, suggesting a stronger or more consistent spatial signal.</p>
</section><section id="predicting-survival-outcomes" class="level3" data-number="11.5.2"><h3 data-number="11.5.2" class="anchored" data-anchor-id="predicting-survival-outcomes">
<span class="header-section-number">11.5.2</span> Predicting survival outcomes</h3>
<p>We can also use <code>sampleMetricMap</code> to identify samples that were difficult to classify with respect to a survival outcome.</p>
<div class="cell">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/samplesMetricMap.html">samplesMetricMap</a></span><span class="op">(</span><span class="va">surv_cv</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="07-classification_files/figure-html/07-samples metric map survival-1.png" class="img-fluid figure-img" width="1152"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>TableGrob (2 x 1) "arrange": 2 grobs
  z     cells    name                grob
1 1 (2-2,1-1) arrange      gtable[layout]
2 2 (1-1,1-1) arrange text[GRID.text.551]</code></pre>
</div>
</div>
<p>In the survival classification task, patient 35 stood out as a difficult case, while samples 31 through 20 were better classified, particularly by <code>lisaClust</code>-derived features. This suggests that spatial domain organisation may carry prognostic value in this dataset.</p>
<p>Samples identified as hard to classify offer valuable opportunities for further exploration:</p>
<ul>
<li>
<strong>Feature enrichment</strong>: Re-examine these samples to see if they possess atypical spatial structures or marker expression profiles that were not well captured by the current feature set. Incorporating additional spatial or morphological features () may improve model sensitivity.</li>
</ul>
<p>Biological subgrouping: Difficult-to-classify samples may represent intermediate or mixed phenotypes. Cluster these cases separately to explore whether they define a distinct subgroup or transition state, especially relevant in diseases with high heterogeneity.</p>
<p>External validation: If another cohort is available, assess whether the same samples or subgroups remain difficult to classify, which may help distinguish model-specific from biologically meaningful ambiguity.</p>
<p>Manual annotation: Engage domain experts to manually review spatial images of ambiguous cases for insights into unmodeled variation, such as rare cell types, artefacts, or unexpected tissue features.</p>
<p>Integrative modelling: Combine spatial features with other data modalities (e.g., genomics, transcriptomics, pathology annotations) to test whether multi-modal integration resolves classification uncertainty in these samples.</p>
 DataFrame(), 
                       outcome, crossValParams, SVMparams)
```
:::

The index of chosen of the parameters, as well as all combinations of parameters and their associated performance metric, are stored for every validation, and can be accessed with the `tunedParameters` function.

::: {.cell}

```{.r .cell-code}
tunedParameters(SVMresults)[1:5]
```
:::
--&gt;
<p>In this way, we can predict clinical outcomes and identify samples that are hard to classify.</p>
<p>In subsequent sections, we will demonstrate how the end-to-end workflow can be applied to a single dataset in the form of case studies.</p>
</section></section><section id="sessioninfo" class="level1" data-number="12"><h1 data-number="12">
<span class="header-section-number">12</span> sessionInfo</h1>
<div class="cell">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.5.0 (2025-04-11)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.4.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.1

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Australia/Sydney
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] SpatialDatasets_1.6.3       SpatialExperiment_1.18.1   
 [3] SingleCellExperiment_1.30.1 ExperimentHub_2.16.0       
 [5] AnnotationHub_3.16.0        BiocFileCache_2.16.0       
 [7] dbplyr_2.5.0                scFeatures_1.8.0           
 [9] ggplot2_3.5.2               spicyR_1.20.1              
[11] Statial_1.10.0              lisaClust_1.16.0           
[13] ClassifyR_3.12.0            survival_3.8-3             
[15] BiocParallel_1.42.0         MultiAssayExperiment_1.34.0
[17] SummarizedExperiment_1.38.1 Biobase_2.68.0             
[19] GenomicRanges_1.60.0        GenomeInfoDb_1.44.0        
[21] IRanges_2.42.0              MatrixGenerics_1.20.0      
[23] matrixStats_1.5.0           S4Vectors_0.46.0           
[25] BiocGenerics_0.54.0         generics_0.1.4             

loaded via a namespace (and not attached):
  [1] GSVA_2.2.0                 ProtGenerics_1.40.0       
  [3] spatstat.sparse_3.1-0      bitops_1.0-9              
  [5] httr_1.4.7                 RColorBrewer_1.1-3        
  [7] numDeriv_2016.8-1.1        EnsDb.Hsapiens.v79_2.99.0 
  [9] tools_4.5.0                doRNG_1.8.6.2             
 [11] backports_1.5.0            R6_2.6.1                  
 [13] HDF5Array_1.36.0           lazyeval_0.2.2            
 [15] mgcv_1.9-3                 rhdf5filters_1.20.0       
 [17] withr_3.0.2                gridExtra_2.3             
 [19] coxme_2.2-22               cli_3.6.5                 
 [21] spatstat.explore_3.4-3     labeling_0.4.3            
 [23] spatstat.data_3.1-6        genefilter_1.90.0         
 [25] SingleCellSignalR_1.20.0   Rsamtools_2.24.0          
 [27] ggupset_0.4.1              R.utils_2.13.0            
 [29] dichromat_2.0-0.1          limma_3.64.1              
 [31] rstudioapi_0.17.1          RSQLite_2.3.11            
 [33] shape_1.4.6.1              BiocIO_1.18.0             
 [35] gtools_3.9.5               spatstat.random_3.4-1     
 [37] car_3.1-3                  dplyr_1.1.4               
 [39] scam_1.2-19                Matrix_1.7-3              
 [41] abind_1.4-8                R.methodsS3_1.8.2         
 [43] lifecycle_1.0.4            yaml_2.3.10               
 [45] edgeR_4.6.2                carData_3.0-5             
 [47] rhdf5_2.52.0               gplots_3.2.0              
 [49] SparseArray_1.8.0          Rtsne_0.17                
 [51] grid_4.5.0                 blob_1.2.4                
 [53] dqrng_0.4.1                crayon_1.5.3              
 [55] bdsmatrix_1.3-7            lattice_0.22-6            
 [57] msigdbr_24.1.0             beachmat_2.24.0           
 [59] GenomicFeatures_1.60.0     annotate_1.86.0           
 [61] KEGGREST_1.48.0            magick_2.8.5              
 [63] pillar_1.10.2              knitr_1.50                
 [65] dcanr_1.24.0               metapod_1.16.0            
 [67] rjson_0.2.23               boot_1.3-31               
 [69] codetools_0.2-20           glue_1.8.0                
 [71] spatstat.univar_3.1-3      data.table_1.17.4         
 [73] vctrs_0.6.5                png_0.1-8                 
 [75] Rdpack_2.6.4               gtable_0.3.6              
 [77] assertthat_0.2.1           cachem_1.1.0              
 [79] xfun_0.52                  mime_0.13                 
 [81] rbibutils_2.3              S4Arrays_1.8.0            
 [83] reformulas_0.4.1           pheatmap_1.0.12           
 [85] iterators_1.0.14           statmod_1.5.0             
 [87] bluster_1.18.0             nlme_3.1-168              
 [89] bit64_4.6.0-1              filelock_1.0.3            
 [91] irlba_2.3.5.1              KernSmooth_2.23-26        
 [93] colorspace_2.1-1           DBI_1.2.3                 
 [95] tidyselect_1.2.1           proxyC_0.5.2              
 [97] bit_4.6.0                  compiler_4.5.0            
 [99] curl_6.2.3                 AUCell_1.30.1             
[101] graph_1.86.0               BiocNeighbors_2.2.0       
[103] h5mread_1.0.1              DelayedArray_0.34.1       
[105] plotly_4.10.4              rtracklayer_1.68.0        
[107] scales_1.4.0               caTools_1.18.3            
[109] rappdirs_0.3.3             stringr_1.5.1             
[111] digest_0.6.37              goftest_1.2-3             
[113] spatstat.utils_3.1-4       minqa_1.2.8               
[115] rmarkdown_2.29             XVector_0.48.0            
[117] htmltools_0.5.8.1          pkgconfig_2.0.3           
[119] lme4_1.1-37                sparseMatrixStats_1.20.0  
[121] fastmap_1.2.0              ensembldb_2.32.0          
[123] rlang_1.1.6                GlobalOptions_0.1.2       
[125] htmlwidgets_1.6.4          ggthemes_5.1.0            
[127] UCSC.utils_1.4.0           DelayedMatrixStats_1.30.0 
[129] ggh4x_0.3.0                farver_2.1.2              
[131] jsonlite_2.0.0             R.oo_1.27.1               
[133] BiocSingular_1.24.0        RCurl_1.98-1.17           
[135] magrittr_2.0.3             Formula_1.2-5             
[137] scuttle_1.18.0             GenomeInfoDbData_1.2.14   
[139] Rhdf5lib_1.30.0            Rcpp_1.0.14               
[141] babelgene_22.9             ape_5.8-1                 
[143] ggnewscale_0.5.1           stringi_1.8.7             
[145] MASS_7.3-65                plyr_1.8.9                
[147] parallel_4.5.0             deldir_2.0-4              
[149] Biostrings_2.76.0          splines_4.5.0             
[151] tensor_1.5                 multtest_2.64.0           
[153] circlize_0.4.16            locfit_1.5-9.12           
[155] igraph_2.1.4               ggpubr_0.6.0              
[157] ranger_0.17.0              spatstat.geom_3.4-1       
[159] ggsignif_0.6.4             rngtools_1.5.2            
[161] reshape2_1.4.4             ScaledMatrix_1.16.0       
[163] BiocVersion_3.21.1         XML_3.99-0.18             
[165] evaluate_1.0.3             scran_1.36.0              
[167] BiocManager_1.30.25        nloptr_2.2.1              
[169] foreach_1.5.2              tweenr_2.0.3              
[171] EnsDb.Mmusculus.v79_2.99.0 tidyr_1.3.1               
[173] purrr_1.0.4                polyclip_1.10-7           
[175] BiocBaseUtils_1.10.0       ggforce_0.4.2             
[177] rsvd_1.0.5                 broom_1.0.8               
[179] xtable_1.8-4               restfulr_0.0.15           
[181] AnnotationFilter_1.32.0    rstatix_0.7.2             
[183] viridisLite_0.4.2          class_7.3-23              
[185] tibble_3.2.1               lmerTest_3.1-3            
[187] memoise_2.0.1              AnnotationDbi_1.70.0      
[189] GenomicAlignments_1.44.0   cluster_2.1.8.1           
[191] concaveman_1.1.0           GSEABase_1.70.0           </code></pre>
</div>
</div>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./06-changes_in_marker_expression.html" class="pagination-link" aria-label="Changes in marker expression">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Changes in marker expression</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./08-case_study1.html" class="pagination-link" aria-label="Case Study: Head and Neck Squamous Cell Carcinoma (Ferguson et al., 2022)">
        <span class="nav-page-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Case Study: Head and Neck Squamous Cell Carcinoma (Ferguson et al., 2022)</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>