<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>12&nbsp; Case Study: Head and Neck Squamous Cell Carcinoma (Ferguson et al., 2022) – Spatial analysis playbook</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./07-classification.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script><script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script><link rel="stylesheet" href="styles.css">
</head>
<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="./index.html" class="navbar-brand navbar-brand-logo">
    <img src="./images/USydLogo.svg" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Spatial analysis playbook</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./08-case_study1.html">Case studies</a></li><li class="breadcrumb-item"><a href="./08-case_study1.html"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Case Study: Head and Neck Squamous Cell Carcinoma (Ferguson et al., 2022)</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Overview</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Datasets</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./getting_started.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Getting Started</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Processing</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01-processing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Cell segmentation and pre-processing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02-quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Quality Control</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Cell annotation</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03a-cell_annotation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Unsupervised clustering for cell annotation</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Cell localisation</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04a-cell_localisation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Cell localisation between pairs of cell types</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04b-cell_localisation_parent.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Cell relationships relative to expected behaviour</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">Spatial domains</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05-cellular_niches.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Identifying spatial domains with unsupervised clustering</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true">
 <span class="menu-text">Marker expression</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06-changes_in_marker_expression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Changes in marker expression</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true">
 <span class="menu-text">Classification</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07-classification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Finding associations between clinical variables and spatial features</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" role="navigation" aria-expanded="true">
 <span class="menu-text">Case studies</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-8" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08-case_study1.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Case Study: Head and Neck Squamous Cell Carcinoma (Ferguson et al., 2022)</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#read-in-images" id="toc-read-in-images" class="nav-link active" data-scroll-target="#read-in-images"><span class="header-section-number">12.1</span> Read in images</a></li>
  <li>
<a href="#simpleseg-segment-the-cells-in-the-images" id="toc-simpleseg-segment-the-cells-in-the-images" class="nav-link" data-scroll-target="#simpleseg-segment-the-cells-in-the-images"><span class="header-section-number">12.2</span> SimpleSeg: Segment the cells in the images</a>
  <ul class="collapse">
<li><a href="#run-simpleseg" id="toc-run-simpleseg" class="nav-link" data-scroll-target="#run-simpleseg"><span class="header-section-number">12.2.1</span> Run simpleSeg</a></li>
  <li><a href="#visualise-separation" id="toc-visualise-separation" class="nav-link" data-scroll-target="#visualise-separation"><span class="header-section-number">12.2.2</span> Visualise separation</a></li>
  <li><a href="#visualise-outlines" id="toc-visualise-outlines" class="nav-link" data-scroll-target="#visualise-outlines"><span class="header-section-number">12.2.3</span> Visualise outlines</a></li>
  </ul>
</li>
  <li><a href="#summarise-cell-features" id="toc-summarise-cell-features" class="nav-link" data-scroll-target="#summarise-cell-features"><span class="header-section-number">12.3</span> Summarise cell features</a></li>
  <li><a href="#load-the-clinical-data" id="toc-load-the-clinical-data" class="nav-link" data-scroll-target="#load-the-clinical-data"><span class="header-section-number">12.4</span> Load the clinical data</a></li>
  <li><a href="#normalise-the-data" id="toc-normalise-the-data" class="nav-link" data-scroll-target="#normalise-the-data"><span class="header-section-number">12.5</span> Normalise the data</a></li>
  <li>
<a href="#fusesom-cluster-cells-into-cell-types" id="toc-fusesom-cluster-cells-into-cell-types" class="nav-link" data-scroll-target="#fusesom-cluster-cells-into-cell-types"><span class="header-section-number">12.6</span> FuseSOM: Cluster cells into cell types</a>
  <ul class="collapse">
<li><a href="#interpreting-cluster-phenotype" id="toc-interpreting-cluster-phenotype" class="nav-link" data-scroll-target="#interpreting-cluster-phenotype"><span class="header-section-number">12.6.1</span> Interpreting cluster phenotype</a></li>
  <li><a href="#testing-for-association-between-the-proportion-of-each-cell-type-and-progressor-status" id="toc-testing-for-association-between-the-proportion-of-each-cell-type-and-progressor-status" class="nav-link" data-scroll-target="#testing-for-association-between-the-proportion-of-each-cell-type-and-progressor-status"><span class="header-section-number">12.6.2</span> Testing for association between the proportion of each cell type and progressor status</a></li>
  </ul>
</li>
  <li><a href="#spicyr-test-spatial-relationships" id="toc-spicyr-test-spatial-relationships" class="nav-link" data-scroll-target="#spicyr-test-spatial-relationships"><span class="header-section-number">12.7</span> spicyR: Test spatial relationships</a></li>
  <li>
<a href="#lisaclust-find-cellular-neighbourhoods" id="toc-lisaclust-find-cellular-neighbourhoods" class="nav-link" data-scroll-target="#lisaclust-find-cellular-neighbourhoods"><span class="header-section-number">12.8</span> lisaClust: Find cellular neighbourhoods</a>
  <ul class="collapse">
<li><a href="#region-cell-type-enrichment-heatmap" id="toc-region-cell-type-enrichment-heatmap" class="nav-link" data-scroll-target="#region-cell-type-enrichment-heatmap"><span class="header-section-number">12.8.1</span> Region-cell type enrichment heatmap</a></li>
  <li><a href="#visualise-regions" id="toc-visualise-regions" class="nav-link" data-scroll-target="#visualise-regions"><span class="header-section-number">12.8.2</span> Visualise regions</a></li>
  <li><a href="#test-for-association-with-progression" id="toc-test-for-association-with-progression" class="nav-link" data-scroll-target="#test-for-association-with-progression"><span class="header-section-number">12.8.3</span> Test for association with progression</a></li>
  </ul>
</li>
  <li>
<a href="#statial-identify-changes-in-cell-state" id="toc-statial-identify-changes-in-cell-state" class="nav-link" data-scroll-target="#statial-identify-changes-in-cell-state"><span class="header-section-number">12.9</span> Statial: Identify changes in cell state</a>
  <ul class="collapse">
<li><a href="#spatiomark-continuous-changes-in-marker-expression-associated-with-varying-levels-of-localisation" id="toc-spatiomark-continuous-changes-in-marker-expression-associated-with-varying-levels-of-localisation" class="nav-link" data-scroll-target="#spatiomark-continuous-changes-in-marker-expression-associated-with-varying-levels-of-localisation"><span class="header-section-number">12.9.1</span> SpatioMark: Continuous changes in marker expression associated with varying levels of localisation</a></li>
  <li><a href="#kontextual-robust-quantification-of-cell-type-localisation-which-is-invariant-to-changes-in-tissue-structure" id="toc-kontextual-robust-quantification-of-cell-type-localisation-which-is-invariant-to-changes-in-tissue-structure" class="nav-link" data-scroll-target="#kontextual-robust-quantification-of-cell-type-localisation-which-is-invariant-to-changes-in-tissue-structure"><span class="header-section-number">12.9.2</span> Kontextual: Robust quantification of cell type localisation which is invariant to changes in tissue structure</a></li>
  </ul>
</li>
  <li>
<a href="#classifyr-classification" id="toc-classifyr-classification" class="nav-link" data-scroll-target="#classifyr-classification"><span class="header-section-number">12.10</span> ClassifyR: Classification</a>
  <ul class="collapse">
<li><a href="#visualise-cross-validated-prediction-performance" id="toc-visualise-cross-validated-prediction-performance" class="nav-link" data-scroll-target="#visualise-cross-validated-prediction-performance"><span class="header-section-number">12.10.1</span> Visualise cross-validated prediction performance</a></li>
  </ul>
</li>
  <li><a href="#sessioninfo" id="toc-sessioninfo" class="nav-link" data-scroll-target="#sessioninfo"><span class="header-section-number">12.11</span> sessionInfo</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./08-case_study1.html">Case studies</a></li><li class="breadcrumb-item"><a href="./08-case_study1.html"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Case Study: Head and Neck Squamous Cell Carcinoma (Ferguson et al., 2022)</span></a></li></ol></nav><div class="quarto-title">
<h1 class="title">
<span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Case Study: Head and Neck Squamous Cell Carcinoma (Ferguson et al., 2022)</span>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><p>In this section, we will demonstrate how the workflow can be applied to a single dataset, the <a href="./datasets.html">Ferguson 2022</a> dataset. The key conclusion of this manuscript (amongst others) is that spatial information about cells and the immune environment can be used to predict primary tumour progression or metastases in patients. We will use our workflow to reach a similar conclusion.</p>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BodenmillerGroup/cytomapper">cytomapper</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/simpleSeg/">simpleSeg</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">FuseSOM</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/">ggpubr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ellispatrick.github.io/spicyR/">spicyR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/ClassifyR/">ClassifyR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ellispatrick.github.io/lisaClust/">lisaClust</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/Statial">Statial</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stemangiola/tidySingleCellExperiment">tidySingleCellExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment">SpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/SydneyBioX/SpatialDatasets">SpatialDatasets</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># set parameters</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># whether to use multiple cores (recommended)</span></span>
<span><span class="va">use_mc</span> <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="va">is_windows</span> <span class="op">=</span> <span class="va">.Platform</span><span class="op">$</span><span class="va">OS.type</span> <span class="op">==</span> <span class="st">"windows"</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">use_mc</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">nCores</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html">ceiling</a></span><span class="op">(</span><span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span><span class="op">)</span>, <span class="fl">1</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">nCores</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">BPPARAM</span> <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html">SerialParam</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">is_windows</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">BPPARAM</span> <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html">SnowParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="va">nCores</span>, type <span class="op">=</span> <span class="st">"SOCK"</span><span class="op">)</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">BPPARAM</span> <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="va">nCores</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="va">BPPARAM</span> <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html">SerialParam</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="read-in-images" class="level2" data-number="12.1"><h2 data-number="12.1" class="anchored" data-anchor-id="read-in-images">
<span class="header-section-number">12.1</span> Read in images</h2>
<p>As before, we can load the Ferguson 2022 image data from the <code>SpatialDatasets</code> package into a <code>CytoImageList</code> and store them as h5 file-on-disk in a temporary directory. We will also assign the metadata columns of the <code>CytoImageList</code> object using the <code>mcols</code> function.</p>
<div class="cell" data-time_it="true">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pathToImages</span> <span class="op">&lt;-</span> <span class="fu">SpatialDatasets</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialDatasets/man/SpatialDatasets.html">Ferguson_Images</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html">unzip</a></span><span class="op">(</span><span class="va">pathToImages</span>, exdir <span class="op">=</span> <span class="va">tmp</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Store images in a CytoImageList on_disk as h5 files to save memory.</span></span>
<span><span class="va">images</span> <span class="op">&lt;-</span> <span class="fu">cytomapper</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/loadImages.html">loadImages</a></span><span class="op">(</span></span>
<span>  <span class="va">tmp</span>,</span>
<span>  single_channel <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  on_disk <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  h5FilesPath <span class="op">=</span> <span class="fu">HDF5Array</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/HDF5Array/man/dump-management.html">getHDF5DumpDir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># assign metadata columns</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">mcols</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">S4Vectors</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html">DataFrame</a></span><span class="op">(</span>imageID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div style="text-align: right;">
<em>Time for this code chunk to run with 6 cores: 146.41 s
</em>
</div>
<p></p>
</div>
<p>As we’re reading the image channels directly from the names of the TIFF image, we will first clean them for ease of downstream processing.</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/CytoImageList-naming.html">channelNames</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/CytoImageList-naming.html">channelNames</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>                          <span class="co"># remove preceding letters</span></span>
<span>                          <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">".*_"</span>, replacement <span class="op">=</span> <span class="st">""</span>, x <span class="op">=</span> <span class="va">_</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>                          <span class="co"># remove the .ome</span></span>
<span>                          <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">".ome"</span>, replacement <span class="op">=</span> <span class="st">""</span>, x <span class="op">=</span> <span class="va">_</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Similarly, the image names will be taken from the folder name containing the individual TIFF images for each channel. These will often also need to be cleaned.</p>
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># cleaning image names to obtain image IDs</span></span>
<span><span class="va">split_names</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"_"</span><span class="op">)</span>, <span class="va">`[`</span>, <span class="fl">3</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">split_names</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">mcols</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">S4Vectors</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html">DataFrame</a></span><span class="op">(</span>imageID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="simpleseg-segment-the-cells-in-the-images" class="level2" data-number="12.2"><h2 data-number="12.2" class="anchored" data-anchor-id="simpleseg-segment-the-cells-in-the-images">
<span class="header-section-number">12.2</span> SimpleSeg: Segment the cells in the images</h2>
<p>Our <a href="https://master.bioconductor.org/packages/release/bioc/html/simpleSeg.html">simpleSeg</a> R package Bioconductor provides a series of functions to generate simple segmentation masks of images. A key strength of the simpleSeg package is that we have included multiple ways to perform some simple segmentation operations and incorporated multiple automatic procedures to optimise some key parameters when these aren’t specified. These functions leverage the functionality of the <a href="https://bioconductor.org/packages/release/bioc/html/EBImage.html">EBImage</a> package on Bioconductor. For more flexibility when performing your segmentation in R, we recommend learning to use the EBImage package.</p>
<section id="run-simpleseg" class="level3" data-number="12.2.1"><h3 data-number="12.2.1" class="anchored" data-anchor-id="run-simpleseg">
<span class="header-section-number">12.2.1</span> Run simpleSeg</h3>
<p>If your images are stored in a <code>list</code> or <code>CytoImageList</code> they can be segmented with a simple call to <code>simpleSeg</code>. <code>simpleSeg</code> is an R implementation of a simple segmentation technique which traces out the nuclei using a specified channel (by setting <code>nucleus =</code>) and then dilates around the traced nuclei by a specified amount (controlled using <code>discSize</code>). The nucleus can be traced out using either one specified channel, or by using the principal components of all channels most correlated to the specified nuclear channel by setting <code>pca = TRUE</code>.</p>
<p>In the example below, we used simpleSeg to trace the nuclei signal in the images based on the HH3 channel, expanding outward from the nucleus by 3 pixels. A more detailed explanation of each of the key parameters is available in the <a href="./01-processing.html">Processing</a> section.</p>
<div class="cell" data-time_it="true">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># generate masks</span></span>
<span><span class="va">masks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/simpleSeg/reference/simpleSeg.html">simpleSeg</a></span><span class="op">(</span><span class="va">images</span>,</span>
<span>                   nucleus <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"HH3"</span><span class="op">)</span>,</span>
<span>                   cellBody <span class="op">=</span> <span class="st">"dilate"</span>,</span>
<span>                   discSize <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                   sizeSelection <span class="op">=</span> <span class="fl">20</span>,</span>
<span>                   transform <span class="op">=</span> <span class="st">"sqrt"</span>,</span>
<span>                   tissue <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"panCK"</span>, <span class="st">"CD45"</span>, <span class="st">"HH3"</span><span class="op">)</span>,</span>
<span>                   cores <span class="op">=</span> <span class="va">nCores</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div style="text-align: right;">
<em>Time for this code chunk to run with 6 cores: 65.99 s
</em>
</div>
<p></p>
</div>
</section><section id="visualise-separation" class="level3" data-number="12.2.2"><h3 data-number="12.2.2" class="anchored" data-anchor-id="visualise-separation">
<span class="header-section-number">12.2.2</span> Visualise separation</h3>
<p>We can then use the <code>display</code> and <code>colorLabels</code> functions from EBImage to examine the performance of the cell segmentation.</p>
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">EBImage</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html">display</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/colorLabels.html">colorLabels</a></span><span class="op">(</span><span class="va">masks</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-visualise segmentation-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="visualise-outlines" class="level3" data-number="12.2.3"><h3 data-number="12.2.3" class="anchored" data-anchor-id="visualise-outlines">
<span class="header-section-number">12.2.3</span> Visualise outlines</h3>
<p>The <code>plotPixels</code> function in <code>cytomapper</code> makes it easy to overlay the mask on top of the nucleus intensity marker to see how well our segmentation process has performed. Here, we can see that the segmentation appears to be performing reasonably. If you see over or under-segmentation of your images, <code>discSize</code> is a key parameter in <code>simpleSeg</code> for optimising the size of the dilation disc after segmenting out the nuclei.</p>
<div class="cell">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/plotPixels.html">plotPixels</a></span><span class="op">(</span>image <span class="op">=</span> <span class="va">images</span><span class="op">[</span><span class="st">"F3"</span><span class="op">]</span>, </span>
<span>           mask <span class="op">=</span> <span class="va">masks</span><span class="op">[</span><span class="st">"F3"</span><span class="op">]</span>,</span>
<span>           img_id <span class="op">=</span> <span class="st">"imageID"</span>, </span>
<span>           colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"HH3"</span><span class="op">)</span>, </span>
<span>           display <span class="op">=</span> <span class="st">"single"</span>,</span>
<span>           colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>HH3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>,<span class="st">"blue"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>           legend <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>           bcg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>             HH3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span></span>
<span>           <span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-plot pixels-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>If you wish to visualise multiple markers instead of just the HH3 marker and see how the segmentation mask performs, this can also be done. Here, we can see that our segmentation mask has done a good job of capturing the CD31 signal, but perhaps not such a good job of capturing the FXIIIA signal, which often lies outside of our dilated nuclear mask. This could suggest that we might need to increase the <code>discSize</code> of our dilation.</p>
<div class="cell">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/plotPixels.html">plotPixels</a></span><span class="op">(</span>image <span class="op">=</span> <span class="va">images</span><span class="op">[</span><span class="st">"F3"</span><span class="op">]</span>, </span>
<span>           mask <span class="op">=</span> <span class="va">masks</span><span class="op">[</span><span class="st">"F3"</span><span class="op">]</span>,</span>
<span>           img_id <span class="op">=</span> <span class="st">"imageID"</span>, </span>
<span>           colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"HH3"</span>, <span class="st">"CD31"</span>, <span class="st">"FX111A"</span><span class="op">)</span>, </span>
<span>           display <span class="op">=</span> <span class="st">"single"</span>,</span>
<span>           colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>HH3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>,<span class="st">"blue"</span><span class="op">)</span>,</span>
<span>                         CD31 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"red"</span><span class="op">)</span>,</span>
<span>                         FX111A <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"green"</span><span class="op">)</span> <span class="op">)</span>,</span>
<span>           legend <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>           bcg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>             HH3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>             CD31 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>             FX111A <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1.5</span><span class="op">)</span></span>
<span>           <span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-plot pixels multiple channels-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Here, we can see that our segmentation mask has done a good job of capturing the CD31 signal, but perhaps not such a good job of capturing the FXIIIA signal, which often lies outside of our dilated nuclear mask. This could suggest that we might need to increase the <code>discSize</code> of our dilation.</p>
</section></section><section id="summarise-cell-features" class="level2" data-number="12.3"><h2 data-number="12.3" class="anchored" data-anchor-id="summarise-cell-features">
<span class="header-section-number">12.3</span> Summarise cell features</h2>
<p>We can then use the <code>measureObjects</code> function from the <code>cytomapper</code> package to characterise the phenotypes of each of the segmented cells. <code>measureObjects</code> will calculate the average intensity of each channel within each cell as well as a few morphological features. By default, <code>measureObjects</code> will return a <code>SingleCellExperiment</code> object, where the channel intensities are stored in the <code>counts</code> assay and the spatial location of each cell is stored in <code>colData</code> in the <code>m.cx</code> and <code>m.cy</code> columns. We can ask <code>measureObjects</code> to return a <code>SpatialExperiment</code> object instead by specifying <code>return_as = "spe"</code>, and the spatial coordinates of each cell will be stored in the <code>spatialCoords</code> slot.</p>
<div class="cell" data-time_it="true">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Summarise the expression of each marker in each cell</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu">cytomapper</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/measureObjects.html">measureObjects</a></span><span class="op">(</span><span class="va">masks</span>,</span>
<span>                                    <span class="va">images</span>,</span>
<span>                                    img_id <span class="op">=</span> <span class="st">"imageID"</span>,</span>
<span>                                    return_as <span class="op">=</span> <span class="st">"spe"</span>,</span>
<span>                                    BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html">spatialCoordsNames</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div style="text-align: right;">
<em>Time for this code chunk to run with 6 cores: 223.48 s
</em>
</div>
<p></p>
</div>
</section><section id="load-the-clinical-data" class="level2" data-number="12.4"><h2 data-number="12.4" class="anchored" data-anchor-id="load-the-clinical-data">
<span class="header-section-number">12.4</span> Load the clinical data</h2>
<p>To associate features in our image with disease progression, it is important to read in information which links image identifiers to their progression status. The clinical data is available through the <code>SpatialDatasets</code> package.</p>
<div class="cell">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">clinical</span> <span class="op">&lt;-</span> <span class="fu">SpatialDatasets</span><span class="fu">::</span><span class="fu">fergusonClinical</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># ensure image IDs match</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">clinical</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">clinical</span><span class="op">$</span><span class="va">imageID</span></span>
<span><span class="va">clinical</span> <span class="op">&lt;-</span> <span class="va">clinical</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># put the clinical data into the colData slot</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">cells</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">cells</span><span class="op">)</span>, <span class="va">clinical</span><span class="op">[</span><span class="va">cells</span><span class="op">$</span><span class="va">imageID</span>, <span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If needed, the <code>SpatialExperiment</code> object can be stored as an R Data file.</p>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span><span class="op">(</span><span class="va">cells</span>, file <span class="op">=</span> <span class="st">"data/cells.rda"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In case you already have your <code>SpatialExperiment</code>/<code>SingleCellExperiment</code> object, you may only be interested in our downstream workflow. For the sake of convenience, we’ve provided capability to directly load in the <code>SpatialExperiment</code> object that we’ve generated.</p>
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"data/cells.rda"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="normalise-the-data" class="level2" data-number="12.5"><h2 data-number="12.5" class="anchored" data-anchor-id="normalise-the-data">
<span class="header-section-number">12.5</span> Normalise the data</h2>
<p>Next, we can check if the marker intensities of each cell require some form of transformation or normalisation. The reason we do this is two-fold:</p>
<ol type="1">
<li>The intensities of images are often highly skewed, preventing any meaningful downstream analysis.</li>
<li>The intensities across different images are often different, meaning that what is considered “positive” can be different across images.</li>
</ol>
<p>By transforming and normalising the data, we aim to reduce these two effects. Below, we extract the marker intensities from the <code>counts</code> assay and take a closer look at the CD3 marker, which should be expressed in the majority of T cells.</p>
<div class="cell">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plot densities of CD3 for each image.</span></span>
<span><span class="va">cells</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ttservice/man/join_features.html">join_features</a></span><span class="op">(</span>features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span>, shape <span class="op">=</span> <span class="st">"wide"</span>, assay <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">CD3</span>, colour <span class="op">=</span> <span class="va">imageID</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-density plots-1.png" class="img-fluid figure-img" width="480"></p>
</figure>
</div>
</div>
</div>
<p>Here, we can see that the intensities are very clearly skewed, and it is difficult to distinguish a CD3- cell from a CD3+ cell. Further, we can clearly see some image-level batch effect, where across images, the intensity peaks differ drastically.</p>
<p>Another method of visualising batch effects is using a dimensionality reduction technique and visualising how the images separate out on a 2D plot. If no batch effect is expected, we should see the images largely overlap with each other.</p>
<div class="cell">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Usually we specify a subset of the original markers which are informative to separating out distinct cell types for the UMAP and clustering.</span></span>
<span><span class="va">ct_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"podoplanin"</span>, <span class="st">"CD13"</span>, <span class="st">"CD31"</span>,</span>
<span>                <span class="st">"panCK"</span>, <span class="st">"CD3"</span>, <span class="st">"CD4"</span>, <span class="st">"CD8a"</span>,</span>
<span>                <span class="st">"CD20"</span>, <span class="st">"CD68"</span>, <span class="st">"CD16"</span>, <span class="st">"CD14"</span>, <span class="st">"HLADR"</span>, <span class="st">"CD66a"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span><span class="co"># Perform dimension reduction using UMAP.</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html">runUMAP</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  subset_row <span class="op">=</span> <span class="va">ct_markers</span>,</span>
<span>  exprs_values <span class="op">=</span> <span class="st">"counts"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Select a subset of images to plot.</span></span>
<span><span class="va">someImages</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">cells</span><span class="op">$</span><span class="va">imageID</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">10</span>, <span class="fl">20</span>, <span class="fl">30</span>, <span class="fl">40</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># UMAP by imageID.</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html">plotReducedDim</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span><span class="op">[</span>, <span class="va">cells</span><span class="op">$</span><span class="va">imageID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">someImages</span><span class="op">]</span>,</span>
<span>  dimred <span class="op">=</span> <span class="st">"UMAP"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"imageID"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-umap-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The UMAP also indicates that some level of batch effect exists in our dataset.</p>
<p>To mitigate this, we can use the <code>normalizeCells</code> function from <code>simpleSeg</code>. Below, we perform normalisation (specified by <code>method =</code>) by 1) trimming the 99th percentile, 2) dividing by the mean and 3) removing the 1st principal component. This modified data is then stored in the <code>norm</code> assay by default.</p>
<div class="cell">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Leave out the nuclei markers from our normalisation process. </span></span>
<span><span class="va">useMarkers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"DNA1"</span>, <span class="st">"DNA2"</span>, <span class="st">"HH3"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Transform and normalise the marker expression of each cell type.</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/simpleSeg/reference/normalizeCells.html">normalizeCells</a></span><span class="op">(</span><span class="va">cells</span>,</span>
<span>                        markers <span class="op">=</span> <span class="va">useMarkers</span>,</span>
<span>                        transformation <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                        method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"trim99"</span>, <span class="st">"mean"</span>, <span class="st">"PC1"</span><span class="op">)</span>,</span>
<span>                        assayIn <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>                        cores <span class="op">=</span> <span class="va">nCores</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot densities of CD3 for each image</span></span>
<span><span class="va">cells</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ttservice/man/join_features.html">join_features</a></span><span class="op">(</span>features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span>, shape <span class="op">=</span> <span class="st">"wide"</span>, assay <span class="op">=</span> <span class="st">"norm"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">CD3</span>, colour <span class="op">=</span> <span class="va">imageID</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-normalisation-1.png" class="img-fluid figure-img" width="480"></p>
</figure>
</div>
</div>
</div>
<p>We can see that this normalised data appears more bimodal, not perfectly, but likely to a sufficient degree for clustering, as we can at least observe a clear CD3+ peak at 1.00, and a CD3- peak at around 0.3. For more information on available normalisation and transformation methods and a detailed explanation of the parameters, refer to our <a href="./02-quality_control.html">Quality Control</a> section.</p>
<p>We can also appreciate through the UMAP a reduction of the batch effect we initially saw.</p>
<div class="cell">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span><span class="co"># Perform dimension reduction using UMAP.</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html">runUMAP</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  subset_row <span class="op">=</span> <span class="va">ct_markers</span>,</span>
<span>  exprs_values <span class="op">=</span> <span class="st">"norm"</span>,</span>
<span>  name <span class="op">=</span> <span class="st">"normUMAP"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">someImages</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">cells</span><span class="op">$</span><span class="va">imageID</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">10</span>, <span class="fl">20</span>, <span class="fl">30</span>, <span class="fl">40</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># UMAP by imageID.</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html">plotReducedDim</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span><span class="op">[</span>, <span class="va">cells</span><span class="op">$</span><span class="va">imageID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">someImages</span><span class="op">]</span>,</span>
<span>  dimred <span class="op">=</span> <span class="st">"normUMAP"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"imageID"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-normalised UMAP-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="fusesom-cluster-cells-into-cell-types" class="level2" data-number="12.6"><h2 data-number="12.6" class="anchored" data-anchor-id="fusesom-cluster-cells-into-cell-types">
<span class="header-section-number">12.6</span> FuseSOM: Cluster cells into cell types</h2>
<p>We can also appreciate from the UMAP above that there is a division of clusters, most likely representing different cell types. We next aim to empirically distinguish each cluster using our <a href="https://www.bioconductor.org/packages/release/bioc/html/FuseSOM.html">FuseSOM</a> package for clustering. FuseSOM provides a pipeline for the clustering of highly multiplexed in situ imaging cytometry assays. This pipeline uses the Self Organising Map architecture coupled with Multiview hierarchical clustering and provides functions for the estimation of the number of clusters.</p>
<p>Here, we cluster using the <code>runFuseSOM</code> function. We specify the number of clusters to identify to be <code>numClusters = 10</code>. We also specify a set of cell-type specific markers to use (<code>ct_markers</code>), as we want our clusters to be distinct based off cell type markers, rather than markers which might pick up a transitioning cell state.</p>
<div class="cell">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Generate SOM and cluster cells into 10 groups</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/runFuseSOM.html">runFuseSOM</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  markers <span class="op">=</span> <span class="va">ct_markers</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"norm"</span>,</span>
<span>  numClusters <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can also observe how reasonable our choice of <code>k = 10</code> was, using the <code>estimateNumCluster</code> and <code>optiPlot</code> functions. Here we examine the Gap method, but others, such as Silhouette and Within Cluster Distance are also available. We can see that there are elbow points in the gap statistic at <code>k = 7</code>, <code>k = 10</code>, and <code>k = 11</code>. We’ve specified <code>k = 10</code>, striking a good balance between the number of clusters and the gap statistic. For more discussion on how to select an appropriate value for <code>k</code>, refer to our <a href="./03a-cell_annotation.html">Unsupervised Clustering</a> section.</p>
<div class="cell">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/estimateNumCluster.html">estimateNumCluster</a></span><span class="op">(</span><span class="va">cells</span>, kSeq <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/optiPlot.html">optiPlot</a></span><span class="op">(</span><span class="va">cells</span>, method <span class="op">=</span> <span class="st">"gap"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-elbow plot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<section id="interpreting-cluster-phenotype" class="level3" data-number="12.6.1"><h3 data-number="12.6.1" class="anchored" data-anchor-id="interpreting-cluster-phenotype">
<span class="header-section-number">12.6.1</span> Interpreting cluster phenotype</h3>
<p>We can begin the process of understanding what each of these cell clusters are by using the <code>plotGroupedHeatmap</code> function from <code>scater</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Visualise marker expression in each cluster.</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotGroupedHeatmap.html">plotGroupedHeatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  features <span class="op">=</span> <span class="va">ct_markers</span>,</span>
<span>  group <span class="op">=</span> <span class="st">"clusters"</span>,</span>
<span>  exprs_values <span class="op">=</span> <span class="st">"norm"</span>,</span>
<span>  center <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  zlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  block <span class="op">=</span> <span class="st">"clusters"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-heatmap-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>At the least, here we can see we capture all the major immune populations that we expect to see, including the CD4 and CD8 T cells, the CD20+ B cells, the CD68+ myeloid populations, the CD66+ granulocytes, the podoplanin+ epithelial cells, and the panCK+ tumour cells.</p>
<p>Given domain-specific knowledge of the tumour-immune landscape, we can go ahead and annotate these clusters as cell types given their expression profiles.</p>
<div class="cell">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="va">cells</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>cellType <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html">case_when</a></span><span class="op">(</span></span>
<span>      <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_1"</span> <span class="op">~</span> <span class="st">"epithelial"</span>,</span>
<span>      <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_2"</span> <span class="op">~</span> <span class="st">"squamous"</span>, <span class="co"># tumour cell type</span></span>
<span>      <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_3"</span> <span class="op">~</span> <span class="st">"monocyte"</span>,</span>
<span>      <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_4"</span> <span class="op">~</span> <span class="st">"myeloid"</span>,</span>
<span>      <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_5"</span> <span class="op">~</span> <span class="st">"granulocyte"</span>,</span>
<span>      <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_6"</span> <span class="op">~</span> <span class="st">"CD8_T"</span>,</span>
<span>      <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_7"</span> <span class="op">~</span> <span class="st">"CD4_T"</span>,</span>
<span>      <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_8"</span> <span class="op">~</span> <span class="st">"B"</span>,</span>
<span>      <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_9"</span> <span class="op">~</span> <span class="st">"endothelial"</span>,</span>
<span>      <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_10"</span> <span class="op">~</span> <span class="st">"dendritic"</span>,</span>
<span>      </span>
<span>  <span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We might also be interested in how these cell types are distributed on the images themselves. Here we examine the distribution of clusters on image F3, noting the healthy epithelial and endothelial structures surrounded by tumour cells.</p>
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDim</a></span><span class="op">(</span><span class="va">cells</span>, <span class="st">"spatialCoords"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html">spatialCoords</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cells</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">imageID</span> <span class="op">==</span> <span class="st">"F3"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html">plotReducedDim</a></span><span class="op">(</span><span class="st">"spatialCoords"</span>, colour_by <span class="op">=</span> <span class="st">"cellType"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-image plot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can also use the UMAP we computed earlier to visualise our data in a lower dimension and see how well our annotated cell types cluster out.</p>
<div class="cell">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># UMAP by cell type</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html">plotReducedDim</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span><span class="op">[</span>, <span class="va">cells</span><span class="op">$</span><span class="va">imageID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">someImages</span><span class="op">]</span>,</span>
<span>  dimred <span class="op">=</span> <span class="st">"normUMAP"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"cellType"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-umap by celltype-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="testing-for-association-between-the-proportion-of-each-cell-type-and-progressor-status" class="level3" data-number="12.6.2"><h3 data-number="12.6.2" class="anchored" data-anchor-id="testing-for-association-between-the-proportion-of-each-cell-type-and-progressor-status">
<span class="header-section-number">12.6.2</span> Testing for association between the proportion of each cell type and progressor status</h3>
<p>We recommend using a package such as <code>diffcyt</code> for testing for changes in abundance of cell types. However, the <code>colTest</code> function from the spicyR package allows us to quickly test for associations between the proportions of the cell types and progression status using either Wilcoxon rank sum tests or t-tests.</p>
<div class="cell">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Test for changes in cell type proportion across progression status with student's t-test</span></span>
<span><span class="va">testProp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/colTest.html">colTest</a></span><span class="op">(</span><span class="va">cells</span>, </span>
<span>                    condition <span class="op">=</span> <span class="st">"group"</span>, </span>
<span>                    feature <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>                    type <span class="op">=</span> <span class="st">"ttest"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">testProp</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            mean in group NP mean in group P tval.t   pval adjPval     cluster
CD8_T                 0.0580          0.0330   2.70 0.0088   0.088       CD8_T
endothelial           0.0820          0.0650   2.00 0.0580   0.290 endothelial
squamous              0.5800          0.6300  -1.50 0.1400   0.470    squamous
dendritic             0.0350          0.0290   1.20 0.2600   0.530   dendritic
myeloid               0.0061          0.0044   1.10 0.2800   0.530     myeloid
B                     0.0210          0.0130   0.99 0.3300   0.530           B</code></pre>
</div>
</div>
<p>Here, we can see that both CD4 T and granulocytes cells appear to be present in different proportions in non-progressors compared to progressors.</p>
<p>Let’s examine one of these clusters using our <code>getProp</code> function from the spicyR package, which conveniently transforms our proportions into a feature matrix of images by cell type, enabling convenient downstream classification or analysis.</p>
<div class="cell">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">prop</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/getProp.html">getProp</a></span><span class="op">(</span><span class="va">cells</span>, feature <span class="op">=</span> <span class="st">"cellType"</span><span class="op">)</span></span>
<span><span class="va">prop</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             B      CD4_T      CD8_T  dendritic endothelial
A2 0.004477278 0.02529662 0.09805238 0.01880457  0.05842848
A3 0.005432814 0.04183267 0.13998551 0.02173126  0.07424846
A4 0.005383814 0.03977075 0.13042723 0.01128864  0.06304272
A5 0.018041999 0.09316770 0.06980183 0.04761905  0.06773144
A6 0.014440433 0.08561114 0.04228984 0.02836514  0.07684373</code></pre>
</div>
</div>
<p>Next, let’s visualise how different the proportions are for the most significant cell type (CD4 T cell) across progression statuses using a boxplot.</p>
<div class="cell">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Obtain most significant cell type - CD4 T cells</span></span>
<span><span class="va">clusterToUse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">testProp</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="va">prop</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html">all_of</a></span><span class="op">(</span><span class="va">clusterToUse</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span><span class="st">"imageID"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">clinical</span>, by <span class="op">=</span> <span class="st">"imageID"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">group</span>, y <span class="op">=</span> <span class="va">.data</span><span class="op">[[</span><span class="va">clusterToUse</span><span class="op">]</span><span class="op">]</span>, fill <span class="op">=</span> <span class="va">group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08a-boxplot CD4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The boxplot visualisation of CD4 T cell proportion clearly shows that progressors have a lower proportion of CD4 T cells in the tumour core. We can repeat the above for granulocytes as well. In contrast to CD4 T cells, progressors have a higher proportion of granulocytes compared to non-progressors.</p>
<div class="cell">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Obtain second most significant cell type - granulocytes</span></span>
<span><span class="va">clusterToUse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">testProp</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span></span>
<span><span class="va">prop</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html">all_of</a></span><span class="op">(</span><span class="va">clusterToUse</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span><span class="st">"imageID"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">clinical</span>, by <span class="op">=</span> <span class="st">"imageID"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">group</span>, y <span class="op">=</span> <span class="va">.data</span><span class="op">[[</span><span class="va">clusterToUse</span><span class="op">]</span><span class="op">]</span>, fill <span class="op">=</span> <span class="va">group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/unnamed-chunk-1-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>If you have already clustered and annotated your cells, you may only be interested in our downstream analysis capabilities, looking into identifying localisation (spicyR), cell regions (lisaClust), and cell-cell interactions (SpatioMark &amp; Kontextual). Therefore, for the sake of convenience, we’ve provided capability to directly load in the SpatialExperiment (SPE) object that we’ve generated up to this point, complete with clusters, cell type annotations, and normalised intensities.</p>
<div class="cell">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"data/computed_cells.rda"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="spicyr-test-spatial-relationships" class="level2" data-number="12.7"><h2 data-number="12.7" class="anchored" data-anchor-id="spicyr-test-spatial-relationships">
<span class="header-section-number">12.7</span> spicyR: Test spatial relationships</h2>
<p>Our <a href="https://www.bioconductor.org/packages/devel/bioc/html/spicyR.html">spicyR</a> package offers a range of functions to support the analysis of immunofluorescence, imaging mass cytometry data, and other assays that provide deep phenotyping of individual cells and their spatial distribution. In this example, we use the <code>spicy</code> function to assess changes in spatial relationships between pairwise combinations of cells.</p>
<p>In simple terms, spicyR utilises the L-function to evaluate whether cell types are localized or dispersed. The L-function quantifies “closeness” between points, where higher values indicate increased localisation and lower values suggest dispersion.</p>
<p>Here, we quantify spatial relationships and mildly account for some global tissue structure using <code>sigma = 50</code>. For a more detailed explanation of the key parameters, refer to the <a href="./04a-cell_localisation.html">Cell Localisation</a> section. Additional information on optimising the parameters can be found in the <a href="https://doi.org/10.1093/bioinformatics/btac268">spicyR paper</a>.</p>
<div class="cell">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">spicyTest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/spicy.html">spicy</a></span><span class="op">(</span><span class="va">cells</span>,</span>
<span>                   condition <span class="op">=</span> <span class="st">"group"</span>,</span>
<span>                   cellTypeCol <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>                   imageIDCol <span class="op">=</span> <span class="st">"imageID"</span>,</span>
<span>                   sigma <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                   BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/topPairs.html">topPairs</a></span><span class="op">(</span><span class="va">spicyTest</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                         intercept coefficient    p.value adj.pvalue
granulocyte__dendritic   -8.544472   13.166206 0.01984412  0.8140202
dendritic__granulocyte   -7.913578   12.986740 0.02296098  0.8140202
epithelial__myeloid       4.599553  -13.632031 0.02840491  0.8140202
myeloid__epithelial       4.593292  -13.017068 0.03256081  0.8140202
dendritic__dendritic     41.164389  -11.645611 0.06060092  0.9366501
epithelial__squamous      2.143769   -3.616736 0.07664431  0.9366501
CD8_T__granulocyte       -9.196658   11.719884 0.08850050  0.9366501
granulocyte__granulocyte 60.308964  -26.776805 0.09868144  0.9366501
CD4_T__dendritic         10.414405    8.173354 0.10228860  0.9366501
dendritic__CD4_T         10.696917    8.124193 0.10273571  0.9366501
                                from          to
granulocyte__dendritic   granulocyte   dendritic
dendritic__granulocyte     dendritic granulocyte
epithelial__myeloid       epithelial     myeloid
myeloid__epithelial          myeloid  epithelial
dendritic__dendritic       dendritic   dendritic
epithelial__squamous      epithelial    squamous
CD8_T__granulocyte             CD8_T granulocyte
granulocyte__granulocyte granulocyte granulocyte
CD4_T__dendritic               CD4_T   dendritic
dendritic__CD4_T           dendritic       CD4_T</code></pre>
</div>
</div>
<p>The most significant interaction appears to occur between granulocytes and dendritic cells with a p-value of 0.019. We can visualise these relationships using the <code>signifPlot</code> function.</p>
<div class="cell">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Visualise which relationships are changing the most.</span></span>
<span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/signifPlot.html">signifPlot</a></span><span class="op">(</span><span class="va">spicyTest</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-signifPlot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We observe that cell type pairs appear to become less attractive (or avoid more) in the progression sample, except for epithelial cells, which appear to become localised. We can also see that granulocytes and myeloid cells are more dispersed in progressors compared to non-progressors.</p>
<p>As an example, we will use the <code>spicyBoxPlot</code> function to investigate whether the squamous tumour cell type localises with the granulocytes and assess whether this localisation influences tumour progression versus non-progression.</p>
<div class="cell">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/spicyBoxPlot.html">spicyBoxPlot</a></span><span class="op">(</span><span class="va">spicyTest</span>, </span>
<span>             from <span class="op">=</span> <span class="st">"squamous"</span>, </span>
<span>             to <span class="op">=</span> <span class="st">"granulocyte"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 1 row containing non-finite outside the scale range
(`stat_boxplot()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-spicyBoxPlot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Alternatively, we can look at the most differentially localised relationship between progressors and non-progressors by specifying <code>rank = 1</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/spicyBoxPlot.html">spicyBoxPlot</a></span><span class="op">(</span><span class="va">spicyTest</span>, </span>
<span>             rank <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 1 row containing non-finite outside the scale range
(`stat_boxplot()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-spicyBoxPlot with rank-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Now that we have examined cellular localisation relationships, we will move on to identifying spatial domains, or regions of co-localisation.</p>
</section><section id="lisaclust-find-cellular-neighbourhoods" class="level2" data-number="12.8"><h2 data-number="12.8" class="anchored" data-anchor-id="lisaclust-find-cellular-neighbourhoods">
<span class="header-section-number">12.8</span> lisaClust: Find cellular neighbourhoods</h2>
<p>Our <a href="https://bioconductor.org/packages/release/bioc/html/lisaClust.html">lisaClust</a> package on Bioconductor provides a series of functions to identify and visualise regions of tissue where spatial associations between cell types is similar. This package can be used to provide a high-level summary of cell-type co-localisation in multiplexed imaging data that has been segmented at a single-cell resolution. Here we use the <code>lisaClust</code> function to clusters cells into 4 regions with distinct spatial ordering. By default, these identified regions are stored in the <code>regions</code> column in the <code>colData</code> of our <code>SpatialExperiment</code> object.</p>
<div class="cell">
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Cluster cells into spatial regions with similar composition.</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/lisaClust.html">lisaClust</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  Rs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">50</span>, <span class="fl">100</span><span class="op">)</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  sigma <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The `BPPARAM` argument of `lisaClust()` is deprecated as of lisaClust 1.14.4.
ℹ Please use the `cores` argument instead.
ℹ The deprecated feature was likely used in the lisaClust package.
  Please report the issue at
  &lt;https://github.com/ellispatrick/lisaClust/issues&gt;.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Generating local L-curves.</code></pre>
</div>
</div>
<p>For more information on how to choose an appropriate value for <code>k</code> , refer to the section on <a href="./05-cellular_niches.html">Identifying spatial domains</a>.</p>
<section id="region-cell-type-enrichment-heatmap" class="level3" data-number="12.8.1"><h3 data-number="12.8.1" class="anchored" data-anchor-id="region-cell-type-enrichment-heatmap">
<span class="header-section-number">12.8.1</span> Region-cell type enrichment heatmap</h3>
<p>We can try to interpret which spatial orderings the regions are quantifying using the <code>regionMap</code> function. This plots the frequency of each cell type in a region relative to what you would expect by chance.</p>
<div class="cell">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Visualise the enrichment of each cell type in each region</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/regionMap.html">regionMap</a></span><span class="op">(</span><span class="va">cells</span>, cellType <span class="op">=</span> <span class="st">"cellType"</span>, limit <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-regionMap-1.png" class="img-fluid figure-img" width="480"></p>
</figure>
</div>
</div>
</div>
<p>We can see here that our regions have neatly separated according to biological milieu, with region 1 and region 3 containing our immune cell types, region 4 containing our healthy epithelial and endothelial cells, and region 2 containing our tumour cells.</p>
</section><section id="visualise-regions" class="level3" data-number="12.8.2"><h3 data-number="12.8.2" class="anchored" data-anchor-id="visualise-regions">
<span class="header-section-number">12.8.2</span> Visualise regions</h3>
<p>We can quickly examine the spatial arrangement of these regions using <code>plotReducedDim</code> on image F3, where we can see the same division of immune, healthy, and tumour tissue that we identified in our <code>regionMap</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cells</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">imageID</span> <span class="op">==</span> <span class="st">"F3"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html">plotReducedDim</a></span><span class="op">(</span><span class="st">"spatialCoords"</span>, colour_by <span class="op">=</span> <span class="st">"region"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-regions on image-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Although slower, we have also implemented a function called <code>hatchingPlot</code> that overlays region information using a hatching pattern, allowing it to be viewed alongside cell type classifications. The granularity of the region boundaries can be controlled using the <code>nbp</code> argument.</p>
<div class="cell">
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Use hatching to visualise regions and cell types.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/hatchingPlot.html">hatchingPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  useImages <span class="op">=</span> <span class="st">"F3"</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>  nbp <span class="op">=</span> <span class="fl">300</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Concave windows are temperamental. Try choosing values of window.length &gt; and &lt; 1 if you have problems.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-hatchingPlot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="test-for-association-with-progression" class="level3" data-number="12.8.3"><h3 data-number="12.8.3" class="anchored" data-anchor-id="test-for-association-with-progression">
<span class="header-section-number">12.8.3</span> Test for association with progression</h3>
<p>Similar to cell type proportions, we can quickly use the <code>colTest</code> function to test for associations between the proportions of cells in each region and progression status by specifying <code>feature = "region"</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Test if the proportion of each region is associated</span></span>
<span><span class="co"># with progression status.</span></span>
<span><span class="va">testRegion</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/colTest.html">colTest</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  feature <span class="op">=</span> <span class="st">"region"</span>,</span>
<span>  condition <span class="op">=</span> <span class="st">"group"</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"ttest"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">testRegion</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         mean in group NP mean in group P tval.t  pval adjPval  cluster
region_4            0.580            0.64  -1.70 0.094    0.20 region_4
region_1            0.220            0.19   1.70 0.099    0.20 region_1
region_2            0.150            0.12   0.98 0.330    0.44 region_2
region_3            0.058            0.05   0.55 0.590    0.59 region_3</code></pre>
</div>
</div>
<p>From the results above, it appears that none of the regions above are significantly associated with changes in progression status.</p>
</section></section><section id="statial-identify-changes-in-cell-state" class="level2" data-number="12.9"><h2 data-number="12.9" class="anchored" data-anchor-id="statial-identify-changes-in-cell-state">
<span class="header-section-number">12.9</span> Statial: Identify changes in cell state</h2>
<p>Our <a href="https://www.bioconductor.org/packages/release/bioc/html/Statial.html">Statial</a> package provides a suite of functions (<code>Kontextual</code>) for robust quantification of cell type localisation which are invariant to changes in tissue structure. In addition, we provide a suite of functions (SpatioMark) for uncovering continuous changes in marker expression associated with varying levels of localisation.</p>
<section id="spatiomark-continuous-changes-in-marker-expression-associated-with-varying-levels-of-localisation" class="level3" data-number="12.9.1"><h3 data-number="12.9.1" class="anchored" data-anchor-id="spatiomark-continuous-changes-in-marker-expression-associated-with-varying-levels-of-localisation">
<span class="header-section-number">12.9.1</span> SpatioMark: Continuous changes in marker expression associated with varying levels of localisation</h3>
<p>The first step in analysing these changes is to calculate the spatial proximity (<code>getDistances</code>) of each cell to every cell type. These values will then be stored in the <code>reducedDims</code> slot of the <code>SingleCellExperiment</code> object under the names <code>distances</code>. SpatioMark also provides functionality to look into proximal cell abundance using the <code>getAbundance</code> function, which is further explored in the <a href="./06-changes_in_marker_expression.html">Changes in marker expression</a> section.</p>
<div class="cell">
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cells</span><span class="op">$</span><span class="va">m.cx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html">spatialCoords</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span><span class="op">[</span>,<span class="st">"x"</span><span class="op">]</span></span>
<span><span class="va">cells</span><span class="op">$</span><span class="va">m.cy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html">spatialCoords</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span><span class="op">[</span>,<span class="st">"y"</span><span class="op">]</span></span>
<span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/getDistances.html">getDistances</a></span><span class="op">(</span><span class="va">cells</span>,</span>
<span>  maxDist <span class="op">=</span> <span class="fl">200</span>,</span>
<span>  nCores <span class="op">=</span> <span class="va">nCores</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>  spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"m.cx"</span>, <span class="st">"m.cy"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can then visualise an example image, specified with <code>image = "F3"</code> and a particular marker interaction with cell type localisation. To visualise these changes, we specify two cell types with the <code>from</code> and <code>to</code> parameters, and a marker with the <code>marker</code> parameter (cell-cell-marker interactions). Here, we specify the changes in the marker podoplanin in squamous tumour cells as its localisation to epithelial cells increases or decreases, where we can observe that podoplanin decreases in tumour cells as its distance to the central cluster of epithelial cells increases.</p>
<div class="cell">
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/plotStateChanges.html">plotStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">cells</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>  spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"m.cx"</span>, <span class="st">"m.cy"</span><span class="op">)</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  image <span class="op">=</span> <span class="st">"F3"</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"squamous"</span>,</span>
<span>  to <span class="op">=</span> <span class="st">"epithelial"</span>,</span>
<span>  marker <span class="op">=</span> <span class="st">"podoplanin"</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  shape <span class="op">=</span> <span class="fl">19</span>,</span>
<span>  interactive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plotModelFit <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot image</span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">image</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-plotStateChanges-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># plot the scatter plot</span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">scatter</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-scatter plot podoplanin-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>SpatioMark aims to holistically uncover all such significant relationships by looking at all interactions across all images. The <code>calcStateChanges</code> function provided by Statial can be expanded for this exact purpose - by not specifying cell types, a marker, or an image, <code>calcStateChanges</code> will examine the most significant correlations between distance and marker expression across the entire dataset.</p>
<div class="cell">
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">state_dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/calcStateChanges.html">calcStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">cells</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  assay <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  nCores <span class="op">=</span> <span class="va">nCores</span>,</span>
<span>  minCells <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">state_dist</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      imageID primaryCellType otherCellType     marker         coef      tval
83730      D6        squamous             B       PDL2  0.002257419  46.97671
81937      D2        squamous       myeloid      panCK -0.001636384 -40.22898
81970      D2        squamous       myeloid       TIM3 -0.002043811 -36.71967
83624      D6        squamous     dendritic      HLADR -0.001209405 -36.70992
84990      F5        squamous   granulocyte       PDL2  0.001800396  37.14811
78711      B1        squamous         CD8_T      CXCR3 -0.002693811 -38.68014
83701      D6        squamous             B      panCK  0.001874891  35.81925
76947      A2        squamous             B      CXCR3  0.001531085  35.99011
88021      H5        squamous             B      panCK  0.001951588  36.59020
91086      J5        squamous    epithelial podoplanin -0.001581229 -34.44390
               pval           fdr
83730  0.000000e+00  0.000000e+00
81937 1.236575e-294 5.709268e-290
81970 1.520922e-252 4.681398e-248
83624 2.057602e-249 4.749973e-245
84990 4.651527e-249 8.590440e-245
78711 6.097199e-240 9.383589e-236
83701 3.032438e-239 4.000218e-235
76947 1.186479e-236 1.369494e-232
88021 1.408016e-229 1.444625e-225
91086 2.192554e-222 2.024605e-218</code></pre>
</div>
</div>
<p>Here, we can see that one of the most significant relationships is between myeloid cells and squamous cells (tumour cells) for the marker TIM3 in image D2. The negative coefficient associated with this relationship tells us that as myeloid cells and squamous cells become more localised, TIM3 expression on the squamous cells decreases.</p>
<p>We can confirm this by examining image D2 using the <code>plotStateChanges</code> function.</p>
<div class="cell">
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/plotStateChanges.html">plotStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">cells</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>  spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"m.cx"</span>, <span class="st">"m.cy"</span><span class="op">)</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  image <span class="op">=</span> <span class="st">"D2"</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"squamous"</span>,</span>
<span>  to <span class="op">=</span> <span class="st">"myeloid"</span>,</span>
<span>  marker <span class="op">=</span> <span class="st">"TIM3"</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  shape <span class="op">=</span> <span class="fl">19</span>,</span>
<span>  interactive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plotModelFit <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot image</span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">image</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-plotStateChanges panCK-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># plot the scatter plot</span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">scatter</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-scatter plot panCK-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The results from our SpatioMark outputs can be converted from a <code>data.frame</code> to a <code>matrix</code>, using the <code>prepMatrix</code> function. The choice of extracting either the t-statistic or the coefficient of the linear regression can be specified using the <code>column = "tval"</code> parameter, with the coefficient being the default extracted parameter. We can see that with SpatioMark, we get some features which are significant after adjusting for FDR.</p>
<div class="cell">
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Preparing outcome vector</span></span>
<span><span class="va">outcome</span> <span class="op">&lt;-</span> <span class="va">cells</span><span class="op">$</span><span class="va">group</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">cells</span><span class="op">$</span><span class="va">imageID</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">outcome</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">cells</span><span class="op">$</span><span class="va">imageID</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">cells</span><span class="op">$</span><span class="va">imageID</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Preparing features for Statial</span></span>
<span><span class="va">distMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/prepMatrix.html">prepMatrix</a></span><span class="op">(</span><span class="va">state_dist</span><span class="op">)</span></span>
<span></span>
<span><span class="va">distMat</span> <span class="op">&lt;-</span> <span class="va">distMat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">outcome</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Remove some very small values</span></span>
<span><span class="va">distMat</span> <span class="op">&lt;-</span> <span class="va">distMat</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">distMat</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0.0001</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">.8</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># test for changes in marker expression across outcomes</span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/colTest.html">colTest</a></span><span class="op">(</span><span class="va">distMat</span>, <span class="va">outcome</span>, type <span class="op">=</span> <span class="st">"ttest"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                               mean in group NP mean in group P tval.t    pval
squamous__squamous__VISTA              -0.00590        -0.00230   -3.7 0.00066
endothelial__endothelial__CD13         -0.00190        -0.00068   -3.5 0.00110
endothelial__endothelial__DNA1         -0.42000        -0.03800   -3.6 0.00110
endothelial__endothelial__DNA2         -0.76000        -0.07200   -3.5 0.00120
endothelial__dendritic__ICOS           -0.00026         0.00006   -3.3 0.00280
squamous__squamous__CADM1              -0.00520        -0.00250   -3.1 0.00360
                               adjPval                        cluster
squamous__squamous__VISTA        0.072      squamous__squamous__VISTA
endothelial__endothelial__CD13   0.072 endothelial__endothelial__CD13
endothelial__endothelial__DNA1   0.072 endothelial__endothelial__DNA1
endothelial__endothelial__DNA2   0.072 endothelial__endothelial__DNA2
endothelial__dendritic__ICOS     0.130   endothelial__dendritic__ICOS
squamous__squamous__CADM1        0.140      squamous__squamous__CADM1</code></pre>
</div>
</div>
<p>When we compare changes in marker expression across outcomes, we can see that one of the most significant relationships is <code>squamous__CD4_T__OX40</code> , or the expression of marker OX40 in squamous cells and their localisation with CD4 T cells. Expression of the marker OX40 in MC is lower in progressors vs non-progressors. OX40 is a co-stimulatory receptor found on activated T cells, primarily CD4+ and CD8+ T cells. It belongs to the tumor necrosis factor receptor (TNFR) superfamily and plays a crucial role in regulating the immune response.</p>
<p>Understanding the spatial relationships of cells within tissue is crucial for interpreting their roles within tissues. However, without defining appropriate contexts, these relationships can be misinterpreted due to confounding factors such as tissue stricture and the choice of region to image. <code>Kontextual</code> provides a framework to generate contexts to measure spatial relationships between cells.</p>
</section><section id="kontextual-robust-quantification-of-cell-type-localisation-which-is-invariant-to-changes-in-tissue-structure" class="level3" data-number="12.9.2"><h3 data-number="12.9.2" class="anchored" data-anchor-id="kontextual-robust-quantification-of-cell-type-localisation-which-is-invariant-to-changes-in-tissue-structure">
<span class="header-section-number">12.9.2</span> Kontextual: Robust quantification of cell type localisation which is invariant to changes in tissue structure</h3>
<p><code>Kontextual</code> is a method to evaluate the localisation relationship between two cell types in an image. <code>Kontextual</code> builds on the L-function by contextualising the relationship between two cell types in reference to the typical spatial behaviour of a <span class="math inline">\(3^{rd}\)</span> cell type/population. By taking this approach, <code>Kontextual</code> is invariant to changes in the window of the image as well as tissue structures which may be present.</p>
<p>The definitions of cell types and cell states are somewhat ambiguous, cell types imply well defined groups of cells that serve different roles from one another, on the other hand cell states imply that cells are a dynamic entity which cannot be discretised, and thus exist in a continuum. For the purposes of using <code>Kontextual</code> we treat cell states as identified clusters of cells, where larger clusters represent a “parent” cell population, and finer sub-clusters representing a “child” cell population. For example a CD4 T cell may be considered a child to a larger parent population of Immune cells. <code>Kontextual</code> thus aims to see how a child population of cells deviate from the spatial behaviour of their parent population, and how that influences the localisation between the child cell state and another cell state.</p>
<section id="cell-type-hierarchy" class="level4" data-number="12.9.2.1"><h4 data-number="12.9.2.1" class="anchored" data-anchor-id="cell-type-hierarchy">
<span class="header-section-number">12.9.2.1</span> Cell type hierarchy</h4>
<p>A key input for Kontextual is an annotation of cell type hierarchies. We will need these to organise all the cells present into cell state populations or clusters, e.g.&nbsp;all the different B cell types are put in a vector called bcells.</p>
<p>Here, we use the <a href="http://www.bioconductor.org/packages/release/bioc/html/treekoR.html">treekoR</a> package on Bioconductor to define these hierarchies in a data driven way.</p>
<div class="cell">
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">exprs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">cells</span>, <span class="st">"norm"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">fergusonTree</span> <span class="op">&lt;-</span> <span class="fu">treekoR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/treekoR/man/getClusterTree.html">getClusterTree</a></span><span class="op">(</span><span class="va">exprs</span>,</span>
<span>                                        <span class="va">cells</span><span class="op">$</span><span class="va">cellType</span>,</span>
<span>                                        hierarchy_method <span class="op">=</span> <span class="st">"hopach"</span>,</span>
<span>                                        scale_exprs <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">parent1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD8_T"</span>, <span class="st">"CD4_T"</span>, <span class="st">"dendritic"</span><span class="op">)</span></span>
<span><span class="va">parent2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"B"</span>, <span class="st">"granulocyte"</span><span class="op">)</span></span>
<span><span class="va">parent3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">parent1</span>, <span class="va">parent2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">parent4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"myeloid"</span>, <span class="st">"epithelial"</span>, <span class="st">"squamous"</span><span class="op">)</span></span>
<span><span class="va">parent5</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">parent4</span>, <span class="st">"endothelial"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">all</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">parent1</span>, <span class="va">parent2</span>, <span class="va">parent3</span>, <span class="va">parent4</span>, <span class="va">parent5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">treeDf</span> <span class="op">=</span> <span class="fu">Statial</span><span class="fu">::</span><span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/parentCombinations.html">parentCombinations</a></span><span class="op">(</span><span class="va">all</span>, <span class="va">parent1</span>, <span class="va">parent2</span>, <span class="va">parent3</span>, <span class="va">parent4</span>, <span class="va">parent5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fergusonTree</span><span class="op">$</span><span class="va">clust_tree</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-parentCombinations-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><code>Kontextual</code> accepts a <code>SingleCellExperiment</code> object, a single image, or list of images from a <code>SingleCellExperiment</code> object, which gets passed into the <code>cells</code> argument. Here, we’ve specified Kontextual to perform calculations on all pairwise combinations for every cluster using the <code>parentCombinations</code> function to create the <code>treeDf</code> dataframe which we’ve specified in the <code>parentDf</code> parameter. The argument <code>r</code> will specify the radius which the cell relationship will be evaluated on. <code>Kontextual</code> supports parallel processing, the number of cores can be specified using the <code>cores</code> argument. <code>Kontextual</code> can take a single value or multiple values for each argument and will test all combinations of the arguments specified.</p>
<p>We can calculate all pairwise relationships across all images for a single radius.</p>
<div class="cell">
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">kontext</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/Kontextual.html">Kontextual</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">cells</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>  spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"m.cx"</span>, <span class="st">"m.cy"</span><span class="op">)</span>,</span>
<span>  parentDf <span class="op">=</span> <span class="va">treeDf</span>,</span>
<span>  r <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  cores <span class="op">=</span> <span class="va">nCores</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Again, we can use the same <code>colTest</code> to quickly test for associations between the Kontextual values and survival probability using either Wilcoxon rank sum tests or t-tests. Similar to SpatioMark, we can specify using either the original L-function by specifying <code>column = "original"</code> in our <code>prepMatrix</code> function.</p>
<div class="cell">
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Converting Kontextual result into data matrix</span></span>
<span><span class="va">kontextMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/prepMatrix.html">prepMatrix</a></span><span class="op">(</span><span class="va">kontext</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Replace NAs with 0</span></span>
<span><span class="va">kontextMat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">kontextMat</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu">spicyR</span><span class="fu">::</span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/colTest.html">colTest</a></span><span class="op">(</span><span class="va">kontextMat</span>, <span class="va">outcome</span>, type <span class="op">=</span> <span class="st">"ttest"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                  mean in group NP mean in group P tval.t
CD4_T__endothelial__parent5                    7.5           14.00   -3.0
squamous__granulocyte__parent3                -4.2            1.70   -2.5
endothelial__CD4_T__parent3                    3.2            7.50   -2.4
granulocyte__squamous__parent5                -1.6            0.71   -2.0
dendritic__CD4_T__parent3                     -6.9           -0.72   -2.0
granulocyte__endothelial__parent5             11.0           -0.76    2.0
                                    pval adjPval
CD4_T__endothelial__parent5       0.0057    0.65
squamous__granulocyte__parent3    0.0190    0.65
endothelial__CD4_T__parent3       0.0250    0.65
granulocyte__squamous__parent5    0.0530    0.65
dendritic__CD4_T__parent3         0.0580    0.65
granulocyte__endothelial__parent5 0.0590    0.65
                                                            cluster
CD4_T__endothelial__parent5             CD4_T__endothelial__parent5
squamous__granulocyte__parent3       squamous__granulocyte__parent3
endothelial__CD4_T__parent3             endothelial__CD4_T__parent3
granulocyte__squamous__parent5       granulocyte__squamous__parent5
dendritic__CD4_T__parent3                 dendritic__CD4_T__parent3
granulocyte__endothelial__parent5 granulocyte__endothelial__parent5</code></pre>
</div>
</div>
<p>One of the most significant relationships is <code>squamous__granulocyte__parent3</code> . When we compare the mean values for this relationship between progressors and non-progressors, we can see that squamous cells and granulocytes show increased co-localisation in progressors vs non-progressors with respect to the parent 3 broader population. When we refer to the cell type hierarchy we defined, we can see that parent 3 refers to the parent population of all immune cells.</p>
<p>Now that we have identified and calculated several features that can be used to measure spatial associations across cell types and regions of localisation, we will see how these metrics can be used to predict clinical outcomes for patients.</p>
</section></section></section><section id="classifyr-classification" class="level2" data-number="12.10"><h2 data-number="12.10" class="anchored" data-anchor-id="classifyr-classification">
<span class="header-section-number">12.10</span> ClassifyR: Classification</h2>
<p>Our <a href="https://www.bioconductor.org/packages/release/bioc/html/ClassifyR.html">ClassifyR</a> package on Bioconductor provides a convenient framework for evaluating classification in R. We provide functionality to easily include four key modelling stages: data transformation, feature selection, classifier training and prediction, into a cross-validation loop. Here we use the <code>crossValidate</code> function to perform 100 repeats of 5-fold cross-validation to evaluate the performance of a random forest applied to five quantifications of our IMC data:</p>
<ol type="1">
<li><p>Cell type proportions (FuseSOM)</p></li>
<li><p>Cell type localisation from spicyR using the L-function</p></li>
<li><p>Region proportions from lisaClust</p></li>
<li><p>Cell type localisation reference to a parent cell type from <code>Kontextual</code></p></li>
<li><p>Cell changes in response to proximal changes from <code>SpatioMark</code></p></li>
</ol>
<p>For a more thorough explanation of ClassifyR, refer to our <a href="./07-classification.html">Classification</a> section.</p>
<div class="cell">
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Create list to store data.frames</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add proportions of each cell type in each image</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"proportions"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/getProp.html">getProp</a></span><span class="op">(</span><span class="va">cells</span>, <span class="st">"cellType"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add pair-wise associations</span></span>
<span><span class="va">spicyMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/bind.html">bind</a></span><span class="op">(</span><span class="va">spicyTest</span><span class="op">)</span></span>
<span><span class="va">spicyMat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">spicyMat</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="va">spicyMat</span> <span class="op">&lt;-</span> <span class="va">spicyMat</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">!</span><span class="va">condition</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span><span class="op">(</span><span class="st">"imageID"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"spicyR"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">spicyMat</span></span>
<span></span>
<span><span class="co"># Add proportions of each region in each image</span></span>
<span><span class="co"># to the list of dataframes.</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"lisaClust"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/getProp.html">getProp</a></span><span class="op">(</span><span class="va">cells</span>, <span class="st">"region"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Add SpatioMark features</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"SpatioMark"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">distMat</span></span>
<span></span>
<span><span class="co"># Add Kontextual features</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"Kontextual"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">kontextMat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Set seed</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform cross-validation of a random forest model</span></span>
<span><span class="co"># with 100 repeats of 5-fold cross-validation.</span></span>
<span><span class="va">cv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/crossValidate.html">crossValidate</a></span><span class="op">(</span></span>
<span>  measurements <span class="op">=</span> <span class="va">data</span>,</span>
<span>  outcome <span class="op">=</span> <span class="va">outcome</span>,</span>
<span>  classifier <span class="op">=</span> <span class="st">"randomForest"</span>,</span>
<span>  nFolds <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  nRepeats <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  nCores <span class="op">=</span> <span class="va">nCores</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="visualise-cross-validated-prediction-performance" class="level3" data-number="12.10.1"><h3 data-number="12.10.1" class="anchored" data-anchor-id="visualise-cross-validated-prediction-performance">
<span class="header-section-number">12.10.1</span> Visualise cross-validated prediction performance</h3>
<p>Here we use the <code>performancePlot</code> function to assess the AUC from each repeat of the 5-fold cross-validation.</p>
<div class="cell">
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Calculate AUC for each cross-validation repeat and plot.</span></span>
<span><span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/performancePlot.html">performancePlot</a></span><span class="op">(</span></span>
<span>  <span class="va">cv</span>,</span>
<span>  metric <span class="op">=</span> <span class="st">"AUC"</span>,</span>
<span>  characteristicsList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Assay Name"</span><span class="op">)</span>,</span>
<span>  orderingList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"Assay Name"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"proportions"</span>, <span class="st">"spicyR"</span>, <span class="st">"lisaClust"</span>, <span class="st">"Kontextual"</span>, <span class="st">"SpatioMark"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-performancePlot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Both cell type proportions and SpatioMark appear to capture information that is predictive of patient outcome.</p>
<p>We can also visualise which features were good at classifying which patients using the <code>sampleMetricMap</code> function from ClassifyR.</p>
<div class="cell">
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/samplesMetricMap.html">samplesMetricMap</a></span><span class="op">(</span><span class="va">cv</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-samplesMetricMap-1.png" class="img-fluid figure-img" width="1152"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>TableGrob (2 x 1) "arrange": 2 grobs
  z     cells    name                 grob
1 1 (2-2,1-1) arrange       gtable[layout]
2 2 (1-1,1-1) arrange text[GRID.text.2130]</code></pre>
</div>
</div>
<p>Overall, it appears that we were more easily able to identify non-progressors compared to progressors, and patients G6 and H2 within the progressor cohort were particularly difficult to classify.</p>
<p>Here we have used a pipeline of our spatial analysis R packages to demonstrate an easy way to segment, cluster, normalise, quantify and classify high dimensional in situ cytometry data all within R.</p>
</section></section><section id="sessioninfo" class="level2" data-number="12.11"><h2 data-number="12.11" class="anchored" data-anchor-id="sessioninfo">
<span class="header-section-number">12.11</span> sessionInfo</h2>
<div class="cell">
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.5.0 (2025-04-11)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.4.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.1

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Australia/Sydney
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] SpatialDatasets_1.6.3           ExperimentHub_2.16.0           
 [3] AnnotationHub_3.16.0            BiocFileCache_2.16.0           
 [5] dbplyr_2.5.0                    SpatialExperiment_1.18.1       
 [7] ttservice_0.4.1                 tidyr_1.3.1                    
 [9] tidySingleCellExperiment_1.18.1 Statial_1.10.0                 
[11] lisaClust_1.16.0                ClassifyR_3.12.0               
[13] survival_3.8-3                  BiocParallel_1.42.0            
[15] MultiAssayExperiment_1.34.0     spicyR_1.20.1                  
[17] scater_1.36.0                   scuttle_1.18.0                 
[19] ggpubr_0.6.0                    FuseSOM_1.10.0                 
[21] simpleSeg_1.9.2                 ggplot2_3.5.2                  
[23] dplyr_1.1.4                     cytomapper_1.20.0              
[25] SingleCellExperiment_1.30.1     SummarizedExperiment_1.38.1    
[27] Biobase_2.68.0                  GenomicRanges_1.60.0           
[29] GenomeInfoDb_1.44.0             IRanges_2.42.0                 
[31] S4Vectors_0.46.0                BiocGenerics_0.54.0            
[33] generics_0.1.4                  MatrixGenerics_1.20.0          
[35] matrixStats_1.5.0               EBImage_4.50.0                 

loaded via a namespace (and not attached):
  [1] dichromat_2.0-0.1           tiff_0.1-12                
  [3] dcanr_1.24.0                FCPS_1.3.4                 
  [5] nnet_7.3-20                 goftest_1.2-3              
  [7] Biostrings_2.76.0           HDF5Array_1.36.0           
  [9] TH.data_1.1-3               vctrs_0.6.5                
 [11] spatstat.random_3.4-1       shape_1.4.6.1              
 [13] digest_0.6.37               png_0.1-8                  
 [15] proxy_0.4-27                BiocBaseUtils_1.10.0       
 [17] ggrepel_0.9.6               deldir_2.0-4               
 [19] permute_0.9-7               magick_2.8.5               
 [21] MASS_7.3-65                 reshape2_1.4.4             
 [23] httpuv_1.6.16               foreach_1.5.2              
 [25] withr_3.0.2                 ggfun_0.1.8                
 [27] psych_2.5.3                 xfun_0.52                  
 [29] ellipsis_0.3.2              doRNG_1.8.6.2              
 [31] memoise_2.0.1               ggbeeswarm_0.7.2           
 [33] RProtoBufLib_2.20.0         diptest_0.77-1             
 [35] princurve_2.1.6             systemfonts_1.2.3          
 [37] tidytree_0.4.6              zoo_1.8-14                 
 [39] GlobalOptions_0.1.2         V8_6.0.3                   
 [41] DEoptimR_1.1-3-1            Formula_1.2-5              
 [43] prabclus_2.3-4              KEGGREST_1.48.0            
 [45] promises_1.3.2              httr_1.4.7                 
 [47] rstatix_0.7.2               rhdf5filters_1.20.0        
 [49] fpc_2.2-13                  rhdf5_2.52.0               
 [51] rstudioapi_0.17.1           UCSC.utils_1.4.0           
 [53] concaveman_1.1.0            curl_6.2.3                 
 [55] ScaledMatrix_1.16.0         h5mread_1.0.1              
 [57] analogue_0.18.0             polyclip_1.10-7            
 [59] GenomeInfoDbData_1.2.14     SparseArray_1.8.0          
 [61] fftwtools_0.9-11            doParallel_1.0.17          
 [63] xtable_1.8-4                stringr_1.5.1              
 [65] evaluate_1.0.3              S4Arrays_1.8.0             
 [67] irlba_2.3.5.1               colorspace_2.1-1           
 [69] filelock_1.0.3              spatstat.data_3.1-6        
 [71] flexmix_2.3-20              magrittr_2.0.3             
 [73] ggtree_3.16.0               later_1.4.2                
 [75] viridis_0.6.5               modeltools_0.2-24          
 [77] lattice_0.22-6              genefilter_1.90.0          
 [79] spatstat.geom_3.4-1         robustbase_0.99-4-1        
 [81] XML_3.99-0.18               cowplot_1.1.3              
 [83] RcppAnnoy_0.0.22            ggupset_0.4.1              
 [85] class_7.3-23                svgPanZoom_0.3.4           
 [87] pillar_1.10.2               nlme_3.1-168               
 [89] iterators_1.0.14            compiler_4.5.0             
 [91] beachmat_2.24.0             stringi_1.8.7              
 [93] tensor_1.5                  minqa_1.2.8                
 [95] plyr_1.8.9                  treekoR_1.16.0             
 [97] crayon_1.5.3                abind_1.4-8                
 [99] gridGraphics_0.5-1          locfit_1.5-9.12            
[101] sp_2.2-0                    bit_4.6.0                  
[103] terra_1.8-50                sandwich_3.1-1             
[105] multcomp_1.4-28             fastcluster_1.3.0          
[107] codetools_0.2-20            textshaping_1.0.1          
[109] BiocSingular_1.24.0         coop_0.6-3                 
[111] GetoptLong_1.0.5            plotly_4.10.4              
[113] mime_0.13                   splines_4.5.0              
[115] circlize_0.4.16             Rcpp_1.0.14                
[117] profileModel_0.6.1          knitr_1.50                 
[119] blob_1.2.4                  clue_0.3-66                
[121] BiocVersion_3.21.1          lme4_1.1-37                
[123] fs_1.6.6                    nnls_1.6                   
[125] Rdpack_2.6.4                ggsignif_0.6.4             
[127] ggplotify_0.1.2             tibble_3.2.1               
[129] Matrix_1.7-3                scam_1.2-19                
[131] statmod_1.5.0               svglite_2.2.1              
[133] tweenr_2.0.3                pkgconfig_2.0.3            
[135] pheatmap_1.0.12             tools_4.5.0                
[137] cachem_1.1.0                rbibutils_2.3              
[139] RSQLite_2.3.11              viridisLite_0.4.2          
[141] DBI_1.2.3                   numDeriv_2016.8-1.1        
[143] fastmap_1.2.0               rmarkdown_2.29             
[145] scales_1.4.0                grid_4.5.0                 
[147] shinydashboard_0.7.3        broom_1.0.8                
[149] patchwork_1.3.0             brglm_0.7.2                
[151] BiocManager_1.30.25         carData_3.0-5              
[153] farver_2.1.2                reformulas_0.4.1           
[155] mgcv_1.9-3                  yaml_2.3.10                
[157] ggthemes_5.1.0              cli_3.6.5                  
[159] purrr_1.0.4                 hopach_2.68.0              
[161] lifecycle_1.0.4             uwot_0.2.3                 
[163] mvtnorm_1.3-3               kernlab_0.9-33             
[165] backports_1.5.0             annotate_1.86.0            
[167] cytolib_2.20.0              gtable_0.3.6               
[169] rjson_0.2.23                parallel_4.5.0             
[171] ape_5.8-1                   limma_3.64.1               
[173] edgeR_4.6.2                 jsonlite_2.0.0             
[175] bitops_1.0-9                bit64_4.6.0-1              
[177] Rtsne_0.17                  FlowSOM_2.16.0             
[179] yulab.utils_0.2.0           vegan_2.6-10               
[181] spatstat.utils_3.1-4        BiocNeighbors_2.2.0        
[183] ranger_0.17.0               flowCore_2.20.0            
[185] bdsmatrix_1.3-7             spatstat.univar_3.1-3      
[187] lazyeval_0.2.2              ConsensusClusterPlus_1.72.0
[189] shiny_1.10.0                htmltools_0.5.8.1          
[191] diffcyt_1.28.0              rappdirs_0.3.3             
[193] glue_1.8.0                  XVector_0.48.0             
[195] RCurl_1.98-1.17             treeio_1.32.0              
[197] mclust_6.1.1                mnormt_2.1.1               
[199] coxme_2.2-22                jpeg_0.1-11                
[201] gridExtra_2.3               boot_1.3-31                
[203] igraph_2.1.4                R6_2.6.1                   
[205] ggiraph_0.8.13              labeling_0.4.3             
[207] ggh4x_0.3.0                 cluster_2.1.8.1            
[209] rngtools_1.5.2              Rhdf5lib_1.30.0            
[211] aplot_0.2.5                 nloptr_2.2.1               
[213] DelayedArray_0.34.1         tidyselect_1.2.1           
[215] vipor_0.4.7                 ggforce_0.4.2              
[217] raster_3.6-32               car_3.1-3                  
[219] AnnotationDbi_1.70.0        rsvd_1.0.5                 
[221] DataVisualizations_1.3.3    data.table_1.17.4          
[223] ComplexHeatmap_2.24.0       htmlwidgets_1.6.4          
[225] RColorBrewer_1.1-3          rlang_1.1.6                
[227] spatstat.sparse_3.1-0       spatstat.explore_3.4-3     
[229] colorRamps_2.3.4            lmerTest_3.1-3             
[231] uuid_1.2-1                  ggnewscale_0.5.1           
[233] fansi_1.0.6                 beeswarm_0.4.0             </code></pre>
</div>
</div>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./07-classification.html" class="pagination-link" aria-label="Finding associations between clinical variables and spatial features">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Finding associations between clinical variables and spatial features</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav>
</div> <!-- /content -->



</body></html>