<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>13&nbsp; Case Study: Head and Neck Squamous Cell Carcinoma (Ferguson et al., 2022) – Spatial analysis playbook</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./09-references.html" rel="next">
<link href="./07-classification.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script><script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script><link rel="stylesheet" href="styles.css">
</head>
<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="./index.html" class="navbar-brand navbar-brand-logo">
    <img src="./images/USydLogo.svg" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Spatial analysis playbook</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./08-case_study1.html">Case studies</a></li><li class="breadcrumb-item"><a href="./08-case_study1.html"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Case Study: Head and Neck Squamous Cell Carcinoma (Ferguson et al., 2022)</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Overview</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Datasets</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./getting_started.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Getting Started</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Processing</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01-processing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Cell segmentation and pre-processing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02-quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Quality Control</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Cell annotation</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03a-cell_annotation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Unsupervised clustering for cell annotation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03b-cell_annotation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Supervised cell annotation with a reference</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Cell localisation</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04a-cell_localisation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Cell localisation between pairs of cell types</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04b-cell_localisation_parent.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Cell relationships relative to expected behaviour</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">Spatial domains</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05-cellular_niches.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Identifying spatial domains with unsupervised clustering</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true">
 <span class="menu-text">Marker expression</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06-changes_in_marker_expression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Changes in marker expression</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true">
 <span class="menu-text">Classification</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07-classification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Finding associations between clinical variables and spatial features</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" role="navigation" aria-expanded="true">
 <span class="menu-text">Case studies</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-8" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08-case_study1.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Case Study: Head and Neck Squamous Cell Carcinoma (Ferguson et al., 2022)</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-9" role="navigation" aria-expanded="true">
 <span class="menu-text">References</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-9" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-9" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./09-references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#load-libraries" id="toc-load-libraries" class="nav-link active" data-scroll-target="#load-libraries"><span class="header-section-number">13.1</span> Load libraries</a></li>
  <li>
<a href="#read-in-images" id="toc-read-in-images" class="nav-link" data-scroll-target="#read-in-images"><span class="header-section-number">13.2</span> Read in images</a>
  <ul class="collapse">
<li><a href="#clean-channel-names" id="toc-clean-channel-names" class="nav-link" data-scroll-target="#clean-channel-names"><span class="header-section-number">13.2.1</span> Clean channel names</a></li>
  <li><a href="#clean-image-names" id="toc-clean-image-names" class="nav-link" data-scroll-target="#clean-image-names"><span class="header-section-number">13.2.2</span> Clean image names</a></li>
  </ul>
</li>
  <li>
<a href="#simpleseg-segment-the-cells-in-the-images" id="toc-simpleseg-segment-the-cells-in-the-images" class="nav-link" data-scroll-target="#simpleseg-segment-the-cells-in-the-images"><span class="header-section-number">13.3</span> SimpleSeg: Segment the cells in the images</a>
  <ul class="collapse">
<li><a href="#run-simpleseg" id="toc-run-simpleseg" class="nav-link" data-scroll-target="#run-simpleseg"><span class="header-section-number">13.3.1</span> Run simpleSeg</a></li>
  <li><a href="#visualise-separation" id="toc-visualise-separation" class="nav-link" data-scroll-target="#visualise-separation"><span class="header-section-number">13.3.2</span> Visualise separation</a></li>
  <li><a href="#visualise-outlines" id="toc-visualise-outlines" class="nav-link" data-scroll-target="#visualise-outlines"><span class="header-section-number">13.3.3</span> Visualise outlines</a></li>
  </ul>
</li>
  <li><a href="#summarise-cell-features" id="toc-summarise-cell-features" class="nav-link" data-scroll-target="#summarise-cell-features"><span class="header-section-number">13.4</span> Summarise cell features</a></li>
  <li>
<a href="#load-the-clinical-data" id="toc-load-the-clinical-data" class="nav-link" data-scroll-target="#load-the-clinical-data"><span class="header-section-number">13.5</span> Load the clinical data</a>
  <ul class="collapse">
<li><a href="#read-the-clinical-data" id="toc-read-the-clinical-data" class="nav-link" data-scroll-target="#read-the-clinical-data"><span class="header-section-number">13.5.1</span> Read the clinical data</a></li>
  </ul>
</li>
  <li><a href="#normalise-the-data" id="toc-normalise-the-data" class="nav-link" data-scroll-target="#normalise-the-data"><span class="header-section-number">13.6</span> Normalise the data</a></li>
  <li>
<a href="#fusesom-cluster-cells-into-cell-types" id="toc-fusesom-cluster-cells-into-cell-types" class="nav-link" data-scroll-target="#fusesom-cluster-cells-into-cell-types"><span class="header-section-number">13.7</span> FuseSOM: Cluster cells into cell types</a>
  <ul class="collapse">
<li><a href="#interpreting-cluster-phenotype" id="toc-interpreting-cluster-phenotype" class="nav-link" data-scroll-target="#interpreting-cluster-phenotype"><span class="header-section-number">13.7.1</span> Interpreting cluster phenotype</a></li>
  <li><a href="#check-cell-type-frequencies" id="toc-check-cell-type-frequencies" class="nav-link" data-scroll-target="#check-cell-type-frequencies"><span class="header-section-number">13.7.2</span> Check cell type frequencies</a></li>
  <li><a href="#testing-for-association-between-the-proportion-of-each-cell-type-and-progressor-status" id="toc-testing-for-association-between-the-proportion-of-each-cell-type-and-progressor-status" class="nav-link" data-scroll-target="#testing-for-association-between-the-proportion-of-each-cell-type-and-progressor-status"><span class="header-section-number">13.7.3</span> Testing for association between the proportion of each cell type and progressor status</a></li>
  </ul>
</li>
  <li><a href="#spicyr-test-spatial-relationships" id="toc-spicyr-test-spatial-relationships" class="nav-link" data-scroll-target="#spicyr-test-spatial-relationships"><span class="header-section-number">13.8</span> spicyR: Test spatial relationships</a></li>
  <li>
<a href="#lisaclust-find-cellular-neighbourhoods" id="toc-lisaclust-find-cellular-neighbourhoods" class="nav-link" data-scroll-target="#lisaclust-find-cellular-neighbourhoods"><span class="header-section-number">13.9</span> lisaClust: Find cellular neighbourhoods</a>
  <ul class="collapse">
<li><a href="#region-cell-type-enrichment-heatmap" id="toc-region-cell-type-enrichment-heatmap" class="nav-link" data-scroll-target="#region-cell-type-enrichment-heatmap"><span class="header-section-number">13.9.1</span> Region-cell type enrichment heatmap</a></li>
  <li><a href="#visualise-regions" id="toc-visualise-regions" class="nav-link" data-scroll-target="#visualise-regions"><span class="header-section-number">13.9.2</span> Visualise regions</a></li>
  <li><a href="#test-for-association-with-progression" id="toc-test-for-association-with-progression" class="nav-link" data-scroll-target="#test-for-association-with-progression"><span class="header-section-number">13.9.3</span> Test for association with progression</a></li>
  </ul>
</li>
  <li>
<a href="#statial-identify-changes-in-cell-state." id="toc-statial-identify-changes-in-cell-state." class="nav-link" data-scroll-target="#statial-identify-changes-in-cell-state."><span class="header-section-number">13.10</span> Statial: Identify changes in cell state.</a>
  <ul class="collapse">
<li><a href="#spatiomark-continuous-changes-in-marker-expression-associated-with-varying-levels-of-localisation." id="toc-spatiomark-continuous-changes-in-marker-expression-associated-with-varying-levels-of-localisation." class="nav-link" data-scroll-target="#spatiomark-continuous-changes-in-marker-expression-associated-with-varying-levels-of-localisation."><span class="header-section-number">13.10.1</span> SpatioMark: Continuous changes in marker expression associated with varying levels of localisation.</a></li>
  <li><a href="#kontextual-robust-quantification-of-cell-type-localisation-which-is-invariant-to-changes-in-tissue-structure" id="toc-kontextual-robust-quantification-of-cell-type-localisation-which-is-invariant-to-changes-in-tissue-structure" class="nav-link" data-scroll-target="#kontextual-robust-quantification-of-cell-type-localisation-which-is-invariant-to-changes-in-tissue-structure"><span class="header-section-number">13.10.2</span> Kontextual: Robust quantification of cell type localisation which is invariant to changes in tissue structure</a></li>
  </ul>
</li>
  <li>
<a href="#classifyr-classification" id="toc-classifyr-classification" class="nav-link" data-scroll-target="#classifyr-classification"><span class="header-section-number">13.11</span> ClassifyR: Classification</a>
  <ul class="collapse">
<li><a href="#visualise-cross-validated-prediction-performance" id="toc-visualise-cross-validated-prediction-performance" class="nav-link" data-scroll-target="#visualise-cross-validated-prediction-performance"><span class="header-section-number">13.11.1</span> Visualise cross-validated prediction performance</a></li>
  </ul>
</li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary"><span class="header-section-number">13.12</span> Summary</a></li>
  <li><a href="#sessioninfo" id="toc-sessioninfo" class="nav-link" data-scroll-target="#sessioninfo"><span class="header-section-number">13.13</span> sessionInfo</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./08-case_study1.html">Case studies</a></li><li class="breadcrumb-item"><a href="./08-case_study1.html"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Case Study: Head and Neck Squamous Cell Carcinoma (Ferguson et al., 2022)</span></a></li></ol></nav><div class="quarto-title">
<h1 class="title">
<span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Case Study: Head and Neck Squamous Cell Carcinoma (Ferguson et al., 2022)</span>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><p>In this section, we will demonstrate how the workflow can be applied to a single dataset, the Ferguson 2022 dataset. The key conclusion of this manuscript (amongst others) is that spatial information about cells and the immune environment can be used to predict primary tumour progression or metastases in patients. We will use our workflow to reach a similar conclusion.</p>
<section id="load-libraries" class="level2" data-number="13.1"><h2 data-number="13.1" class="anchored" data-anchor-id="load-libraries">
<span class="header-section-number">13.1</span> Load libraries</h2>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BodenmillerGroup/cytomapper">cytomapper</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/simpleSeg/">simpleSeg</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">FuseSOM</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/">ggpubr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ellispatrick.github.io/spicyR/">spicyR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/ClassifyR/">ClassifyR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ellispatrick.github.io/lisaClust/">lisaClust</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/Statial">Statial</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stemangiola/tidySingleCellExperiment">tidySingleCellExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment">SpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/SydneyBioX/SpatialDatasets">SpatialDatasets</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">use_mc</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">use_mc</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">nCores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html">detectCores</a></span><span class="op">(</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="va">nCores</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="op">}</span></span>
<span><span class="va">BPPARAM</span> <span class="op">&lt;-</span> <span class="fu">simpleSeg</span><span class="fu">:::</span><span class="fu"><a href="https://sydneybiox.github.io/simpleSeg/reference/generateBPParam.html">generateBPParam</a></span><span class="op">(</span><span class="va">nCores</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="read-in-images" class="level2" data-number="13.2"><h2 data-number="13.2" class="anchored" data-anchor-id="read-in-images">
<span class="header-section-number">13.2</span> Read in images</h2>
<p>As before, we can load the Ferguson 2022 image data from the <code>SpatialDatasets</code> package into a <code>CytoImageList</code> and store them as h5 file-on-disk in a temporary directory. We will also assign the metadata columns of the <code>CytoImageList</code> object using the <code>mcols</code> function.</p>
<div class="cell" data-time_it="true">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pathToImages</span> <span class="op">&lt;-</span> <span class="fu">SpatialDatasets</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialDatasets/man/SpatialDatasets.html">Ferguson_Images</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>see ?SpatialDatasets and browseVignettes('SpatialDatasets') for documentation</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>loading from cache</code></pre>
</div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html">unzip</a></span><span class="op">(</span><span class="va">pathToImages</span>, exdir <span class="op">=</span> <span class="va">tmp</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Store images in a CytoImageList on_disk as h5 files to save memory.</span></span>
<span><span class="va">images</span> <span class="op">&lt;-</span> <span class="fu">cytomapper</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/loadImages.html">loadImages</a></span><span class="op">(</span></span>
<span>  <span class="va">tmp</span>,</span>
<span>  single_channel <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  on_disk <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  h5FilesPath <span class="op">=</span> <span class="fu">HDF5Array</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/HDF5Array/man/dump-management.html">getHDF5DumpDir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># assign metadata columns</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">mcols</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">S4Vectors</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html">DataFrame</a></span><span class="op">(</span>imageID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Time for this code chunk to run with 5.5 cores: 139.49 seconds</p>
</div>
<section id="clean-channel-names" class="level3" data-number="13.2.1"><h3 data-number="13.2.1" class="anchored" data-anchor-id="clean-channel-names">
<span class="header-section-number">13.2.1</span> Clean channel names</h3>
<p>As we’re reading the image channels directly from the names of the TIFF image, we will first clean them for ease of downstream processing.</p>
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/CytoImageList-naming.html">channelNames</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/CytoImageList-naming.html">channelNames</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>                          <span class="co"># remove preceding letters</span></span>
<span>                          <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">".*_"</span>, replacement <span class="op">=</span> <span class="st">""</span>, x <span class="op">=</span> <span class="va">_</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>                          <span class="co"># remove the .ome</span></span>
<span>                          <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">".ome"</span>, replacement <span class="op">=</span> <span class="st">""</span>, x <span class="op">=</span> <span class="va">_</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="clean-image-names" class="level3" data-number="13.2.2"><h3 data-number="13.2.2" class="anchored" data-anchor-id="clean-image-names">
<span class="header-section-number">13.2.2</span> Clean image names</h3>
<p>Similarly, the image names will be taken from the folder name containing the individual TIFF images for each channel. These will often also need to be cleaned.</p>
<div class="cell">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># cleaning image names to obtain image IDs</span></span>
<span><span class="va">split_names</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"_"</span><span class="op">)</span>, <span class="va">`[`</span>, <span class="fl">3</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">split_names</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">mcols</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">S4Vectors</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html">DataFrame</a></span><span class="op">(</span>imageID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="simpleseg-segment-the-cells-in-the-images" class="level2" data-number="13.3"><h2 data-number="13.3" class="anchored" data-anchor-id="simpleseg-segment-the-cells-in-the-images">
<span class="header-section-number">13.3</span> SimpleSeg: Segment the cells in the images</h2>
<p>Next, we will use our <a href="https://master.bioconductor.org/packages/release/bioc/html/simpleSeg.html">simpleSeg</a> package to generate simple segmentation masks of images.</p>
<p>Our simpleSeg R package on <a href="https://github.com/SydneyBioX/simpleSeg" class="uri">https://github.com/SydneyBioX/simpleSeg</a> provides a series of functions to generate simple segmentation masks of images. A key strength of the simpleSeg package is that we have included multiple ways to perform some simple segmentation operations and incorporated multiple automatic procedures to optimise some key parameters when these aren’t specified. These functions leverage the functionality of the <a href="https://bioconductor.org/packages/release/bioc/html/EBImage.html">EBImage</a> package on Bioconductor. For more flexibility when performing your segmentation in R, we recommend learning to use the EBImage package.</p>
<section id="run-simpleseg" class="level3" data-number="13.3.1"><h3 data-number="13.3.1" class="anchored" data-anchor-id="run-simpleseg">
<span class="header-section-number">13.3.1</span> Run simpleSeg</h3>
<p>If your images are stored in a <code>list</code> or <code>CytoImageList</code> they can be segmented with a simple call to <code>simpleSeg</code>. <code>simpleSeg</code> is an R implementation of a simple segmentation technique which traces out the nuclei using a specified channel (by setting <code>nucleus =</code>) then dilates around the traced nuclei by a specified amount (controlled using <code>discSize</code>). The nucleus can be traced out using either one specified channel, or by using the principal components of all channels most correlated to the specified nuclear channel by setting <code>pca = TRUE</code>.</p>
<p>In the example below, we used simpleSeg to trace the nuclei signal in the images based on the HH3 channel, expanding outward from the nucleus by 3 pixels. A more detailed explanation of each of the key parameters is available in the <a href="./01-processing.html">Processing</a> section.</p>
<div class="cell" data-time_it="true">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># generate masks</span></span>
<span><span class="va">masks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/simpleSeg/reference/simpleSeg.html">simpleSeg</a></span><span class="op">(</span><span class="va">images</span>,</span>
<span>                   nucleus <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"HH3"</span><span class="op">)</span>,</span>
<span>                   cellBody <span class="op">=</span> <span class="st">"dilate"</span>,</span>
<span>                   discSize <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                   sizeSelection <span class="op">=</span> <span class="fl">20</span>,</span>
<span>                   transform <span class="op">=</span> <span class="st">"sqrt"</span>,</span>
<span>                   tissue <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"panCK"</span>, <span class="st">"CD45"</span>, <span class="st">"HH3"</span><span class="op">)</span>,</span>
<span>                   cores <span class="op">=</span> <span class="va">nCores</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Time for this code chunk to run with 5.5 cores: 41.81 seconds</p>
</div>
</section><section id="visualise-separation" class="level3" data-number="13.3.2"><h3 data-number="13.3.2" class="anchored" data-anchor-id="visualise-separation">
<span class="header-section-number">13.3.2</span> Visualise separation</h3>
<p>We can then use the <code>display</code> and <code>colorLabels</code> functions from EBImage to examine the performance of the cell segmentation.</p>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">EBImage</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html">display</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/colorLabels.html">colorLabels</a></span><span class="op">(</span><span class="va">masks</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-visualise segmentation-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="visualise-outlines" class="level3" data-number="13.3.3"><h3 data-number="13.3.3" class="anchored" data-anchor-id="visualise-outlines">
<span class="header-section-number">13.3.3</span> Visualise outlines</h3>
<p>The <code>plotPixels</code> function in <code>cytomapper</code> makes it easy to overlay the mask on top of the nucleus intensity marker to see how well our segmentation process has performed. Here, we can see that the segmentation appears to be performing reasonably. If you see over or under-segmentation of your images, <code>discSize</code> is a key parameter in <code>simpleSeg</code> for optimising the size of the dilation disc after segmenting out the nuclei.</p>
<div class="cell">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/plotPixels.html">plotPixels</a></span><span class="op">(</span>image <span class="op">=</span> <span class="va">images</span><span class="op">[</span><span class="st">"F3"</span><span class="op">]</span>, </span>
<span>           mask <span class="op">=</span> <span class="va">masks</span><span class="op">[</span><span class="st">"F3"</span><span class="op">]</span>,</span>
<span>           img_id <span class="op">=</span> <span class="st">"imageID"</span>, </span>
<span>           colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"HH3"</span><span class="op">)</span>, </span>
<span>           display <span class="op">=</span> <span class="st">"single"</span>,</span>
<span>           colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>HH3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>,<span class="st">"blue"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>           legend <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>           bcg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>             HH3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span></span>
<span>           <span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-plot pixels-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>If you wish to visualise multiple markers instead of just the HH3 marker and see how the segmentation mask performs, this can also be done. Here, we can see that our segmentation mask has done a good job of capturing the CD31 signal, but perhaps not such a good job of capturing the FXIIIA signal, which often lies outside of our dilated nuclear mask. This could suggest that we might need to increase the <code>discSize</code> of our dilation.</p>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/plotPixels.html">plotPixels</a></span><span class="op">(</span>image <span class="op">=</span> <span class="va">images</span><span class="op">[</span><span class="st">"F3"</span><span class="op">]</span>, </span>
<span>           mask <span class="op">=</span> <span class="va">masks</span><span class="op">[</span><span class="st">"F3"</span><span class="op">]</span>,</span>
<span>           img_id <span class="op">=</span> <span class="st">"imageID"</span>, </span>
<span>           colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"HH3"</span>, <span class="st">"CD31"</span>, <span class="st">"FX111A"</span><span class="op">)</span>, </span>
<span>           display <span class="op">=</span> <span class="st">"single"</span>,</span>
<span>           colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>HH3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>,<span class="st">"blue"</span><span class="op">)</span>,</span>
<span>                         CD31 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"red"</span><span class="op">)</span>,</span>
<span>                         FX111A <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"green"</span><span class="op">)</span> <span class="op">)</span>,</span>
<span>           legend <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>           bcg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>             HH3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>             CD31 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>             FX111A <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1.5</span><span class="op">)</span></span>
<span>           <span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-plot pixels multiple channels-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section></section><section id="summarise-cell-features" class="level2" data-number="13.4"><h2 data-number="13.4" class="anchored" data-anchor-id="summarise-cell-features">
<span class="header-section-number">13.4</span> Summarise cell features</h2>
<p>We can then use the <code>measureObjects</code> function from the <code>cytomapper</code> package to characterise the phenotypes of each of the segmented cells. <code>measureObjects</code> will calculate the average intensity of each channel within each cell as well as a few morphological features. By default, <code>measureObjects</code> will return a <code>SingleCellExperiment</code> object, where the channel intensities are stored in the <code>counts</code> assay and the spatial location of each cell is stored in <code>colData</code> in the <code>m.cx</code> and <code>m.cy</code> columns. We can ask <code>measureObjects</code> to return a <code>SpatialExperiment</code> object instead by specifying <code>return_as = "spe</code>, and the spatial coordinates of each cell will be stored in the <code>spatialCoords</code> slot.</p>
<div class="cell" data-time_it="true">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Summarise the expression of each marker in each cell</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu">cytomapper</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/measureObjects.html">measureObjects</a></span><span class="op">(</span><span class="va">masks</span>,</span>
<span>                                    <span class="va">images</span>,</span>
<span>                                    img_id <span class="op">=</span> <span class="st">"imageID"</span>,</span>
<span>                                    return_as <span class="op">=</span> <span class="st">"spe"</span>,</span>
<span>                                    BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html">spatialCoordsNames</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Time for this code chunk to run with 5.5 cores: 344.15 seconds</p>
</div>
</section><section id="load-the-clinical-data" class="level2" data-number="13.5"><h2 data-number="13.5" class="anchored" data-anchor-id="load-the-clinical-data">
<span class="header-section-number">13.5</span> Load the clinical data</h2>
<p>To associate features in our image with disease progression, it is important to read in information which links image identifiers to their progression status. The clinical data is available through the <code>SpatialDatasets</code> package.</p>
<section id="read-the-clinical-data" class="level3" data-number="13.5.1"><h3 data-number="13.5.1" class="anchored" data-anchor-id="read-the-clinical-data">
<span class="header-section-number">13.5.1</span> Read the clinical data</h3>
<div class="cell">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">clinical</span> <span class="op">&lt;-</span> <span class="fu">SpatialDatasets</span><span class="fu">::</span><span class="fu">fergusonClinical</span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>see ?SpatialDatasets and browseVignettes('SpatialDatasets') for documentation</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>loading from cache</code></pre>
</div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># ensure image IDs match</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">clinical</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">clinical</span><span class="op">$</span><span class="va">imageID</span></span>
<span><span class="va">clinical</span> <span class="op">&lt;-</span> <span class="va">clinical</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># put the clinical data into the colData slot</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">cells</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">cells</span><span class="op">)</span>, <span class="va">clinical</span><span class="op">[</span><span class="va">cells</span><span class="op">$</span><span class="va">imageID</span>, <span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If needed, the <code>SpatialExperiment</code> object can be stored as an R Data file.</p>
<div class="cell">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span><span class="op">(</span><span class="va">cells</span>, file <span class="op">=</span> <span class="st">"spe_Ferguson_2022.rda"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In case you already have your <code>SpatialExperiment</code>/<code>SingleCellExperiment</code> object, you may only be interested in our downstream workflow. For the sake of convenience, we’ve provided capability to directly load in the <code>SpatialExperiment</code> object that we’ve generated.</p>
<div class="cell">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"data/cells.rda"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="normalise-the-data" class="level2" data-number="13.6"><h2 data-number="13.6" class="anchored" data-anchor-id="normalise-the-data">
<span class="header-section-number">13.6</span> Normalise the data</h2>
<p>Next, we can check if the marker intensities of each cell require some form of transformation or normalisation. The reason we do this is two-fold:<br></p>
<ol type="1">
<li>The intensities of images are often highly skewed, preventing any meaningful downstream analysis.<br>
</li>
<li>The intensities across different images are often different, meaning that what is considered “positive” can be different across images.</li>
</ol>
<p>By transforming and normalising the data, we aim to reduce these two effects. Below, we extract the marker intensities from the <code>counts</code> assay and take a closer look at the CD3 marker, which should be expressed in the majority of T cells.</p>
<div class="cell">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plot densities of CD3 for each image.</span></span>
<span><span class="va">cells</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ttservice/man/join_features.html">join_features</a></span><span class="op">(</span>features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span>, shape <span class="op">=</span> <span class="st">"wide"</span>, assay <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">CD3</span>, colour <span class="op">=</span> <span class="va">imageID</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-density plots-1.png" class="img-fluid figure-img" width="480"></p>
</figure>
</div>
</div>
</div>
<p>Here, we can see that the intensities are very clearly skewed, and it is difficult to distinguish a CD3- cell from a CD3+ cell. Further, we can clearly see some image-level batch effect, where across images, the intensity peaks differ drastically.</p>
<p>Another method of visualising batch effects is using a dimensionality reduction technique and visualising how the images separate out on a 2D plot. If no batch effect is expected, we should see the images largely overlap with each other.</p>
<div class="cell">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Usually we specify a subset of the original markers which are informative to separating out distinct cell types for the UMAP and clustering.</span></span>
<span><span class="va">ct_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"podoplanin"</span>, <span class="st">"CD13"</span>, <span class="st">"CD31"</span>,</span>
<span>                <span class="st">"panCK"</span>, <span class="st">"CD3"</span>, <span class="st">"CD4"</span>, <span class="st">"CD8a"</span>,</span>
<span>                <span class="st">"CD20"</span>, <span class="st">"CD68"</span>, <span class="st">"CD16"</span>, <span class="st">"CD14"</span>, <span class="st">"HLADR"</span>, <span class="st">"CD66a"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span><span class="co"># Perform dimension reduction using UMAP.</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html">runUMAP</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  subset_row <span class="op">=</span> <span class="va">ct_markers</span>,</span>
<span>  exprs_values <span class="op">=</span> <span class="st">"counts"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Select a subset of images to plot.</span></span>
<span><span class="va">someImages</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">cells</span><span class="op">$</span><span class="va">imageID</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">10</span>, <span class="fl">20</span>, <span class="fl">30</span>, <span class="fl">40</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># UMAP by imageID.</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html">plotReducedDim</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span><span class="op">[</span>, <span class="va">cells</span><span class="op">$</span><span class="va">imageID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">someImages</span><span class="op">]</span>,</span>
<span>  dimred <span class="op">=</span> <span class="st">"UMAP"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"imageID"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-umap-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The UMAP also indicates that some level of batch effect exists in our dataset.</p>
<p>To mitigate this, we can use the <code>normalizeCells</code> function from <code>simpleSeg</code>. Below, we perform normalisation (specified by <code>method =</code>) by 1) trimming the 99th percentile, 2) dividing by the mean and 3) removing the 1st principal component. This modified data is then stored in the <code>norm</code> assay by default.</p>
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Leave out the nuclei markers from our normalisation process. </span></span>
<span><span class="va">useMarkers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"DNA1"</span>, <span class="st">"DNA2"</span>, <span class="st">"HH3"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Transform and normalise the marker expression of each cell type.</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/simpleSeg/reference/normalizeCells.html">normalizeCells</a></span><span class="op">(</span><span class="va">cells</span>,</span>
<span>                        markers <span class="op">=</span> <span class="va">useMarkers</span>,</span>
<span>                        transformation <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                        method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"trim99"</span>, <span class="st">"mean"</span>, <span class="st">"PC1"</span><span class="op">)</span>,</span>
<span>                        assayIn <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>                        cores <span class="op">=</span> <span class="va">nCores</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot densities of CD3 for each image</span></span>
<span><span class="va">cells</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ttservice/man/join_features.html">join_features</a></span><span class="op">(</span>features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span>, shape <span class="op">=</span> <span class="st">"wide"</span>, assay <span class="op">=</span> <span class="st">"norm"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">CD3</span>, colour <span class="op">=</span> <span class="va">imageID</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-normalisation-1.png" class="img-fluid figure-img" width="480"></p>
</figure>
</div>
</div>
</div>
<p>We can see that this normalised data appears more bimodal, not perfectly, but likely to a sufficient degree for clustering, as we can at least observe a clear CD3+ peak at 1.00, and a CD3- peak at around 0.3. For more information on available normalisation and transformation methods and a detailed explanation of the parameters, refer to our <a href="./02-quality_control.html">Quality Control</a> section.</p>
<p>We can also appreciate through the UMAP a reduction of the batch effect we initially saw.</p>
<div class="cell">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span><span class="co"># Perform dimension reduction using UMAP.</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html">runUMAP</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  subset_row <span class="op">=</span> <span class="va">ct_markers</span>,</span>
<span>  exprs_values <span class="op">=</span> <span class="st">"norm"</span>,</span>
<span>  name <span class="op">=</span> <span class="st">"normUMAP"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">someImages</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">cells</span><span class="op">$</span><span class="va">imageID</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">10</span>, <span class="fl">20</span>, <span class="fl">30</span>, <span class="fl">40</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># UMAP by imageID.</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html">plotReducedDim</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span><span class="op">[</span>, <span class="va">cells</span><span class="op">$</span><span class="va">imageID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">someImages</span><span class="op">]</span>,</span>
<span>  dimred <span class="op">=</span> <span class="st">"normUMAP"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"imageID"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-normalised UMAP-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="fusesom-cluster-cells-into-cell-types" class="level2" data-number="13.7"><h2 data-number="13.7" class="anchored" data-anchor-id="fusesom-cluster-cells-into-cell-types">
<span class="header-section-number">13.7</span> FuseSOM: Cluster cells into cell types</h2>
<p>We can also appreciate from the UMAP above that there is a division of clusters, most likely representing different cell types. We next aim to empirically distinguish each cluster using our FuseSOM package for clustering.</p>
<p>Our FuseSOM R package can be found on bioconductor at <a href="https://www.bioconductor.org/packages/release/bioc/html/FuseSOM.html" class="uri">https://www.bioconductor.org/packages/release/bioc/html/FuseSOM.html</a>, and provides a pipeline for the clustering of highly multiplexed in situ imaging cytometry assays. This pipeline uses the Self Organising Map architecture coupled with Multiview hierarchical clustering and provides functions for the estimation of the number of clusters.</p>
<p>Here we cluster using the <code>runFuseSOM</code> function. We specify the number of clusters to identify to be <code>numClusters = 10</code>. We also specify a set of cell-type specific markers to use, as we want our clusters to be distinct based off cell type markers, rather than markers which might pick up a transitioning cell state.</p>
<div class="cell">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Generate SOM and cluster cells into 10 groups</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/runFuseSOM.html">runFuseSOM</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  markers <span class="op">=</span> <span class="va">ct_markers</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"norm"</span>,</span>
<span>  numClusters <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>You have provided a dataset of class SpatialExperiment</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Everything looks good. Now running the FuseSOM algorithm</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Now Generating the Self Organizing Map Grid</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Optimal Grid Size is: 8</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Now Running the Self Organizing Map Model</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Now Clustering the Prototypes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required namespace: fastcluster</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Now Mapping Clusters to the Original Data</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The Prototypes have been Clustered and Mapped Successfully</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The FuseSOM algorithm has completed successfully</code></pre>
</div>
</div>
<p>We can also observe how reasonable our choice of <code>k = 10</code> was, using the <code>estimateNumCluster</code> and <code>optiPlot</code> functions. Here we examine the Gap method, but others, such as Silhouette and Within Cluster Distance are also available. We can see that there are elbow points in the gap statistic at <code>k = 7</code>, <code>k = 10</code>, and <code>k = 11</code>. We’ve specified <code>k = 10</code>, striking a good balance between the number of clusters and the gap statistic. For more discussion on how to select an appropriate value for <code>k</code>, refer to our <a href="./03a-cell_annotation.html">Unsupervised Clustering</a> section.</p>
<div class="cell">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/estimateNumCluster.html">estimateNumCluster</a></span><span class="op">(</span><span class="va">cells</span>, kSeq <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>You have provided a dataset of class: SpatialExperiment</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Now Computing the Number of Clusters using Discriminant Analysis</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Now Computing The Number Of Clusters Using Distance Analysis</code></pre>
</div>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/optiPlot.html">optiPlot</a></span><span class="op">(</span><span class="va">cells</span>, method <span class="op">=</span> <span class="st">"gap"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>You have provided a dataset of class: SpatialExperiment</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-elbow plot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<section id="interpreting-cluster-phenotype" class="level3" data-number="13.7.1"><h3 data-number="13.7.1" class="anchored" data-anchor-id="interpreting-cluster-phenotype">
<span class="header-section-number">13.7.1</span> Interpreting cluster phenotype</h3>
<p>We can begin the process of understanding what each of these cell clusters are by using the <code>plotGroupedHeatmap</code> function from <code>scater</code>. At the least, here we can see we capture all the major immune populations that we expect to see, including the CD4 and CD8 T cells, the CD20+ B cells, the CD68+ myeloid populations, the CD66+ granulocytes, the podoplanin+ epithelial cells, and the panCK+ tumour cells.</p>
<div class="cell">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Visualise marker expression in each cluster.</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotGroupedHeatmap.html">plotGroupedHeatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  features <span class="op">=</span> <span class="va">ct_markers</span>,</span>
<span>  group <span class="op">=</span> <span class="st">"clusters"</span>,</span>
<span>  exprs_values <span class="op">=</span> <span class="st">"norm"</span>,</span>
<span>  center <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  zlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  block <span class="op">=</span> <span class="st">"clusters"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-heatmap-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Given domain-specific knowledge of the tumour-immune landscape, we can go ahead and annotate these clusters as cell types given their expression profiles.</p>
<div class="cell">
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="va">cells</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>cellType <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html">case_when</a></span><span class="op">(</span></span>
<span>    <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_1"</span> <span class="op">~</span> <span class="st">"GC"</span>, <span class="co"># Granulocytes</span></span>
<span>    <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_2"</span> <span class="op">~</span> <span class="st">"MC"</span>, <span class="co"># Myeloid cells</span></span>
<span>    <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_3"</span> <span class="op">~</span> <span class="st">"SC"</span>, <span class="co"># Squamous cells</span></span>
<span>    <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_4"</span> <span class="op">~</span> <span class="st">"EP"</span>, <span class="co"># Epithelial cells</span></span>
<span>    <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_5"</span> <span class="op">~</span> <span class="st">"SC"</span>, <span class="co"># Squamous cells (tumour cells)</span></span>
<span>    <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_6"</span> <span class="op">~</span> <span class="st">"TC_CD4"</span>, <span class="co"># CD4 T cells</span></span>
<span>    <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_7"</span> <span class="op">~</span> <span class="st">"BC"</span>, <span class="co"># B cells</span></span>
<span>    <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_8"</span> <span class="op">~</span> <span class="st">"EC"</span>, <span class="co"># Endothelial cells</span></span>
<span>    <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_9"</span> <span class="op">~</span> <span class="st">"TC_CD8"</span>, <span class="co"># CD8 T cells</span></span>
<span>    <span class="va">clusters</span> <span class="op">==</span> <span class="st">"cluster_10"</span> <span class="op">~</span> <span class="st">"DC"</span> <span class="co"># Dendritic cells</span></span>
<span>  <span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>New names:
New names:
• `UMAP1` -&gt; `UMAP1...1`
• `UMAP2` -&gt; `UMAP2...2`
• `UMAP1` -&gt; `UMAP1...3`
• `UMAP2` -&gt; `UMAP2...4`</code></pre>
</div>
</div>
<p>We might also be interested in how these cell types are distributed on the images themselves. Here we examine the distribution of clusters on image F3, noting the healthy epithelial and endothelial structures surrounded by tumour cells.</p>
<div class="cell">
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDim</a></span><span class="op">(</span><span class="va">cells</span>, <span class="st">"spatialCoords"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html">spatialCoords</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cells</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">imageID</span> <span class="op">==</span> <span class="st">"F3"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html">plotReducedDim</a></span><span class="op">(</span><span class="st">"spatialCoords"</span>, colour_by <span class="op">=</span> <span class="st">"cellType"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>New names:
• `UMAP1` -&gt; `UMAP1...1`
• `UMAP2` -&gt; `UMAP2...2`
• `UMAP1` -&gt; `UMAP1...3`
• `UMAP2` -&gt; `UMAP2...4`</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-image plot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="check-cell-type-frequencies" class="level3" data-number="13.7.2"><h3 data-number="13.7.2" class="anchored" data-anchor-id="check-cell-type-frequencies">
<span class="header-section-number">13.7.2</span> Check cell type frequencies</h3>
<p>It is always helpful to check the number of cells in each cluster. Here, we observe that Cluster 10 contains a large number of cells, likely tumour cells, characterized by high panCK expression and inconsistent expression of other markers. In contrast, Cluster 4 has very few cells.</p>
<div class="cell">
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Check cell type frequencies.</span></span>
<span><span class="va">cells</span><span class="op">$</span><span class="va">cellType</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    EP     EC     DC     GC TC_CD4     BC     SC TC_CD8     MC 
   708   3290   4769   5262   8896  11788  11842  13587  92903 </code></pre>
</div>
</div>
<p>We can also use the UMAP we computed earlier to visualise our data in a lower dimension and see how well our annotated cell types cluster out.</p>
<div class="cell">
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># UMAP by cell type</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html">plotReducedDim</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span><span class="op">[</span>, <span class="va">cells</span><span class="op">$</span><span class="va">imageID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">someImages</span><span class="op">]</span>,</span>
<span>  dimred <span class="op">=</span> <span class="st">"normUMAP"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"cellType"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-umap by celltype-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="testing-for-association-between-the-proportion-of-each-cell-type-and-progressor-status" class="level3" data-number="13.7.3"><h3 data-number="13.7.3" class="anchored" data-anchor-id="testing-for-association-between-the-proportion-of-each-cell-type-and-progressor-status">
<span class="header-section-number">13.7.3</span> Testing for association between the proportion of each cell type and progressor status</h3>
<p>We recommend using a package such as <code>diffcyt</code> for testing for changes in abundance of cell types. However, the <code>colTest</code> function from the spicyR package allows us to quickly test for associations between the proportions of the cell types and progression status using either Wilcoxon rank sum tests or t-tests. Here we see a p-value less than 0.05 for our CD8 T cells, but this does not equate to a small FDR.</p>
<div class="cell">
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Perform simple student's t-tests on the columns of the proportion matrix.</span></span>
<span><span class="va">testProp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/colTest.html">colTest</a></span><span class="op">(</span><span class="va">cells</span>, </span>
<span>                    condition <span class="op">=</span> <span class="st">"group"</span>, </span>
<span>                    feature <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>                    type <span class="op">=</span> <span class="st">"ttest"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">testProp</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       mean in group NP mean in group P tval.t   pval adjPval cluster
TC_CD4           0.0590          0.0330   2.70 0.0092   0.049  TC_CD4
GC               0.0250          0.0510  -2.80 0.0110   0.049      GC
TC_CD8           0.1000          0.0840   1.70 0.1000   0.300  TC_CD8
EP               0.0062          0.0046   1.00 0.3200   0.500      EP
SC               0.0700          0.0840  -1.00 0.3200   0.500      SC
DC               0.0340          0.0300   0.99 0.3300   0.500      DC</code></pre>
</div>
</div>
<p>Let’s examine one of these clusters using our <code>getProp</code> function from the spicyR package, which conveniently transforms our proportions into a feature matrix of images by cell type, enabling convenient downstream classification or analysis.</p>
<div class="cell">
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">prop</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/getProp.html">getProp</a></span><span class="op">(</span><span class="va">cells</span>, feature <span class="op">=</span> <span class="st">"cellType"</span><span class="op">)</span></span>
<span><span class="va">prop</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           BC         DC          EC           EP         GC
A2 0.04290358 0.01690141 0.003250271 0.0010834236 0.02881907
A3 0.04834329 0.01991671 0.002715915 0.0005431831 0.01575231
A4 0.04116013 0.01163598 0.002952414 0.0053838138 0.01823550
A5 0.09108911 0.04780764 0.017821782 0.0008486563 0.01838755
A6 0.08572844 0.02824579 0.013875124 0.0019821606 0.01189296</code></pre>
</div>
</div>
<p>Next, let’s visualise how different the proportions are across progression statuses using a boxplot.</p>
<div class="cell">
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Obtain most significant cell type </span></span>
<span><span class="va">clusterToUse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">testProp</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="va">prop</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html">all_of</a></span><span class="op">(</span><span class="va">clusterToUse</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span><span class="st">"imageID"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">clinical</span>, by <span class="op">=</span> <span class="st">"imageID"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">group</span>, y <span class="op">=</span> <span class="va">.data</span><span class="op">[[</span><span class="va">clusterToUse</span><span class="op">]</span><span class="op">]</span>, fill <span class="op">=</span> <span class="va">group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08a-boxplot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>It appears that the CD8 T cells are the most differentially abundant cell type across our progressors and non-progressors. The boxplot visualisation of CD8 T cell proportion clearly shows that progressors have a lower proportion of CD8 T cells in the tumour core.</p>
<p><strong>NB</strong>: If you have already clustered and annotated your cells, you may only be interested in our downstream analysis capabilities, looking into identifying localisation (spicyR), cell regions (lisaClust), and cell-cell interactions (SpatioMark &amp; Kontextual). Therefore, for the sake of convenience, we’ve provided capability to directly load in the SpatialExperiment (SPE) object that we’ve generated up to this point, complete with clusters, cell type annotations, and normalised intensities.</p>
<div class="cell">
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"data/computed_cells.rda"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="spicyr-test-spatial-relationships" class="level2" data-number="13.8"><h2 data-number="13.8" class="anchored" data-anchor-id="spicyr-test-spatial-relationships">
<span class="header-section-number">13.8</span> spicyR: Test spatial relationships</h2>
<p>Our <a href="https://www.bioconductor.org/packages/devel/bioc/html/spicyR.html">spicyR</a> package offers a range of functions to support the analysis of immunofluorescence, imaging mass cytometry data, and other assays that provide deep phenotyping of individual cells and their spatial distribution. In this example, we use the <code>spicy</code> function to assess changes in spatial relationships between pairwise combinations of cells.</p>
<p>In simple terms, spicyR utilises the L-function to evaluate whether cell types are localized or dispersed. The L-function quantifies “closeness” between points, where higher values indicate increased localisation and lower values suggest dispersion.</p>
<p>Here, we quantify spatial relationships using a combination of 10 radii from 10 to 100 by specifying <code>Rs = 1:10*10</code> and mildly account for some global tissue structure using <code>sigma = 50</code>. For a more detailed explanation of the key parameters, refer to the <a href="./04a-cell_localisation.html">Cell Localisation</a> section. Additional information on optimising the parameters can be found in the <a href="https://doi.org/10.1093/bioinformatics/btac268">spicyR paper</a>.</p>
<div class="cell">
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">spicyTest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/spicy.html">spicy</a></span><span class="op">(</span><span class="va">cells</span>,</span>
<span>                   condition <span class="op">=</span> <span class="st">"group"</span>,</span>
<span>                   cellTypeCol <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>                   imageIDCol <span class="op">=</span> <span class="st">"imageID"</span>,</span>
<span>                   Rs <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">*</span><span class="fl">10</span>,</span>
<span>                   sigma <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                   BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Coercing condition into factor. Using group = NP as base comparison group. If
this is not the desired base group, please convert cells$group into a factor
and change the order of levels(cells$group) so that the base group is at index
1.</code></pre>
</div>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/topPairs.html">topPairs</a></span><span class="op">(</span><span class="va">spicyTest</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        intercept coefficient    p.value adj.pvalue from to
GC__MC  2.5225478   -4.772656 0.02970197  0.5084069   GC MC
EP__BC -1.6808757   11.358866 0.03072870  0.5084069   EP BC
BC__EP -1.7355224   10.878726 0.04145476  0.5084069   BC EP
GC__EP  0.4338434  -15.446652 0.04286558  0.5084069   GC EP
SC__SC 36.2057721  -12.991741 0.05423307  0.5084069   SC SC
EP__GC  0.2370792  -14.226026 0.06087964  0.5084069   EP GC
SC__DC -0.7454582    6.168406 0.06162965  0.5084069   SC DC
MC__GC  2.2961515   -4.183908 0.06210399  0.5084069   MC GC
DC__SC -0.3706126    6.001380 0.06561924  0.5084069   DC SC
GC__GC 43.6023084  -15.697248 0.08022246  0.5084069   GC GC</code></pre>
</div>
</div>
<p>The most significant interaction appears to occur between B cells (BC) and endothelial cells (EC) with a p-value of 0.03. We can visualise these relatiionships using the <code>signifPlot</code> function, where we observe that cell type pairs appear to become less attractive (or avoid more) in the progression sample.</p>
<div class="cell">
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Visualise which relationships are changing the most.</span></span>
<span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/signifPlot.html">signifPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">spicyTest</span>,</span>
<span>  breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1.5</span>, <span class="fl">1.5</span>, <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-signifPlot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>spicyR also includes functionality for visualising individual pairwise relationships. As an example, we can investigate whether the SC tumour cell type localizes with the GC granular cell type and assess whether this localisation influences tumour progression versus non-progression.</p>
<div class="cell">
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/spicyBoxPlot.html">spicyBoxPlot</a></span><span class="op">(</span><span class="va">spicyTest</span>, </span>
<span>             from <span class="op">=</span> <span class="st">"SC"</span>, </span>
<span>             to <span class="op">=</span> <span class="st">"GC"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-spicyBoxPlot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Alternatively, we can look at the most differentially localised relationship between progressors and non-progressors by specifying <code>rank = 1</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/spicyBoxPlot.html">spicyBoxPlot</a></span><span class="op">(</span><span class="va">spicyTest</span>, </span>
<span>             rank <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-spicyBoxPlot with rank-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="lisaclust-find-cellular-neighbourhoods" class="level2" data-number="13.9"><h2 data-number="13.9" class="anchored" data-anchor-id="lisaclust-find-cellular-neighbourhoods">
<span class="header-section-number">13.9</span> lisaClust: Find cellular neighbourhoods</h2>
<p>Our <a href="https://bioconductor.org/packages/release/bioc/html/lisaClust.html">lisaClust</a> package on Bioconductor provides a series of functions to identify and visualise regions of tissue where spatial associations between cell types is similar. This package can be used to provide a high-level summary of cell-type co-localisation in multiplexed imaging data that has been segmented at a single-cell resolution. Here we use the <code>lisaClust</code> function to clusters cells into 5 regions with distinct spatial ordering. By default, these identified regions are stored in the <code>regions</code> column in the <code>colData</code> of our <code>SpatialExperiment</code> object.</p>
<div class="cell">
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Cluster cells into spatial regions with similar composition.</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/lisaClust.html">lisaClust</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  sigma <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Generating local L-curves.</code></pre>
</div>
</div>
<section id="region-cell-type-enrichment-heatmap" class="level3" data-number="13.9.1"><h3 data-number="13.9.1" class="anchored" data-anchor-id="region-cell-type-enrichment-heatmap">
<span class="header-section-number">13.9.1</span> Region-cell type enrichment heatmap</h3>
<p>We can try to interpret which spatial orderings the regions are quantifying using the <code>regionMap</code> function. This plots the frequency of each cell type in a region relative to what you would expect by chance.</p>
<div class="cell">
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Visualise the enrichment of each cell type in each region</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/regionMap.html">regionMap</a></span><span class="op">(</span><span class="va">cells</span>, cellType <span class="op">=</span> <span class="st">"cellType"</span>, limit <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-regionMap-1.png" class="img-fluid figure-img" width="480"></p>
</figure>
</div>
</div>
</div>
<p>We can see here that our regions have neatly separated according to biological milieu, with region 2 containing our immune cell types, region 1 containing our tumour cells, and regions 3 and 4 representing our healthy epithelial and endothelial cells.</p>
</section><section id="visualise-regions" class="level3" data-number="13.9.2"><h3 data-number="13.9.2" class="anchored" data-anchor-id="visualise-regions">
<span class="header-section-number">13.9.2</span> Visualise regions</h3>
<p>We can quickly examine the spatial arrangement of these regions using <code>plotReducedDim</code> on image F3, where we can see the same division of immune, healthy, and tumour tissue that we identified in our <code>regionMap</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cells</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">imageID</span> <span class="op">==</span> <span class="st">"F3"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html">plotReducedDim</a></span><span class="op">(</span><span class="st">"spatialCoords"</span>, colour_by <span class="op">=</span> <span class="st">"region"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-regions on image-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Although slower, we have also implemented a function called <code>hatchingPlot</code> that overlays region information using a hatching pattern, allowing it to be viewed alongside cell type classifications. The granularity of the region boundaries can be controlled using the <code>nbp</code> argument.</p>
<div class="cell">
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Use hatching to visualise regions and cell types.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/hatchingPlot.html">hatchingPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  useImages <span class="op">=</span> <span class="st">"F3"</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>  nbp <span class="op">=</span> <span class="fl">300</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Concave windows are temperamental. Try choosing values of window.length &gt; and &lt; 1 if you have problems.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-hatchingPlot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="test-for-association-with-progression" class="level3" data-number="13.9.3"><h3 data-number="13.9.3" class="anchored" data-anchor-id="test-for-association-with-progression">
<span class="header-section-number">13.9.3</span> Test for association with progression</h3>
<p>Similar to cell type proportions, we can quickly use the <code>colTest</code> function to test for associations between the proportions of cells in each region and progression status by specifying <code>feature = "region"</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Test if the proportion of each region is associated</span></span>
<span><span class="co"># with progression status.</span></span>
<span><span class="va">testRegion</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/colTest.html">colTest</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  feature <span class="op">=</span> <span class="st">"region"</span>,</span>
<span>  condition <span class="op">=</span> <span class="st">"group"</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"ttest"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">testRegion</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         mean in group NP mean in group P tval.t  pval adjPval  cluster
region_3            0.260           0.280  -1.90 0.062    0.25 region_3
region_1            0.120           0.092   1.10 0.280    0.43 region_1
region_2            0.067           0.052   1.00 0.320    0.43 region_2
region_4            0.560           0.570  -0.61 0.550    0.55 region_4</code></pre>
</div>
</div>
<p>From the results above, it appears that none of the regions above are significantly associated with changes in progression status.</p>
</section></section><section id="statial-identify-changes-in-cell-state." class="level2" data-number="13.10"><h2 data-number="13.10" class="anchored" data-anchor-id="statial-identify-changes-in-cell-state.">
<span class="header-section-number">13.10</span> Statial: Identify changes in cell state.</h2>
<p>Our Statial package (https://www.bioconductor.org/packages/release/bioc/html/Statial.html) provides a suite of functions (Kontextual) for robust quantification of cell type localisation which are invariant to changes in tissue structure. In addition, we provide a suite of functions (SpatioMark) for uncovering continuous changes in marker expression associated with varying levels of localisation.</p>
<section id="spatiomark-continuous-changes-in-marker-expression-associated-with-varying-levels-of-localisation." class="level3" data-number="13.10.1"><h3 data-number="13.10.1" class="anchored" data-anchor-id="spatiomark-continuous-changes-in-marker-expression-associated-with-varying-levels-of-localisation.">
<span class="header-section-number">13.10.1</span> SpatioMark: Continuous changes in marker expression associated with varying levels of localisation.</h3>
<p>The first step in analysing these changes is to calculate the spatial proximity (<code>getDistances</code>) of each cell to every cell type. These values will then be stored in the <code>reducedDims</code> slot of the <code>SingleCellExperiment</code> object under the names <code>distances</code>. SpatioMark also provides functionality to look into proximal cell abundance using the <code>getAbundance()</code> function, which is further explored within the <code>Statial</code> package vignette.</p>
<div class="cell">
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cells</span><span class="op">$</span><span class="va">m.cx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html">spatialCoords</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span><span class="op">[</span>,<span class="st">"x"</span><span class="op">]</span></span>
<span><span class="va">cells</span><span class="op">$</span><span class="va">m.cy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html">spatialCoords</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span><span class="op">[</span>,<span class="st">"y"</span><span class="op">]</span></span>
<span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/getDistances.html">getDistances</a></span><span class="op">(</span><span class="va">cells</span>,</span>
<span>  maxDist <span class="op">=</span> <span class="fl">200</span>,</span>
<span>  nCores <span class="op">=</span> <span class="va">nCores</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>  spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"m.cx"</span>, <span class="st">"m.cy"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can then visualise an example image, specified with <code>image = "F3"</code> and a particular marker interaction with cell type localisation. To visualise these changes, we specify two cell types with the <code>from</code> and <code>to</code> parameters, and a marker with the <code>marker</code> parameter (cell-cell-marker interactions). Here, we specify the changes in the marker podoplanin in <code>SC</code> tumour cells as its localisation to <code>EP</code> epithelial cells increases or decreases, where we can observe that podoplanin decreases in tumour cells as its distance to the central cluster of epithelial cells increases.</p>
<div class="cell">
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/plotStateChanges.html">plotStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">cells</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>  spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"m.cx"</span>, <span class="st">"m.cy"</span><span class="op">)</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  image <span class="op">=</span> <span class="st">"F3"</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"SC"</span>,</span>
<span>  to <span class="op">=</span> <span class="st">"EP"</span>,</span>
<span>  marker <span class="op">=</span> <span class="st">"podoplanin"</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  shape <span class="op">=</span> <span class="fl">19</span>,</span>
<span>  interactive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plotModelFit <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"lm"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$image</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-plotStateChanges-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
$scatter</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-plotStateChanges-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>SpatioMark aims to holistically uncover all such significant relationships by looking at all interactions across all images. The <code>calcStateChanges</code> function provided by Statial can be expanded for this exact purpose - by not specifying cell types, a marker, or an image, <code>calcStateChanges</code> will examine the most significant correlations between distance and marker expression across the entire dataset.</p>
<div class="cell">
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">state_dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/calcStateChanges.html">calcStateChanges</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">cells</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"distances"</span>,</span>
<span>  assay <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  nCores <span class="op">=</span> <span class="va">nCores</span>,</span>
<span>  minCells <span class="op">=</span> <span class="fl">100</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">state_dist</span><span class="op">[</span><span class="va">state_dist</span><span class="op">$</span><span class="va">imageID</span> <span class="op">==</span> <span class="st">"F3"</span>,<span class="op">]</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      imageID primaryCellType otherCellType     marker          coef      tval
35862      F3              MC            GC podoplanin -0.0017138476 -22.69802
35813      F3              MC        TC_CD8     granzB  0.0018015902  17.29237
35970      F3              MC            DC podoplanin -0.0023483328 -15.73779
35785      F3              MC        TC_CD8      panCK  0.0008986139  15.31246
35790      F3              MC        TC_CD8 podoplanin -0.0016517601 -15.25008
35934      F3              MC            EP podoplanin -0.0015441776 -15.24769
35885      F3              MC            GC     granzB  0.0011636708  14.72402
35922      F3              MC            BC       PDL2 -0.0030450146 -14.68869
35857      F3              MC            GC      panCK  0.0006283033  14.28967
35957      F3              MC            EP     granzB  0.0012276163  12.14207
               pval          fdr
35862 1.319907e-100 6.073493e-98
35813  3.877840e-62 6.428409e-60
35970  2.065433e-52 2.481265e-50
35785  7.193652e-50 8.020130e-48
35790  1.679878e-49 1.843108e-47
35934  1.735197e-49 1.898315e-47
35885  1.930051e-46 1.900612e-44
35922  3.076834e-46 3.014265e-44
35857  5.613956e-44 5.104599e-42
35957  1.125969e-32 6.360720e-31</code></pre>
</div>
</div>
<p>The results from our SpatioMark outputs can be converted from a <code>data.frame</code> to a <code>matrix</code>, using the <code><a href="https://sydneybiox.github.io/Statial/reference/prepMatrix.html">prepMatrix()</a></code> function. Note, the choice of extracting either the t-statistic or the coefficient of the linear regression can be specified using the <code>column = "tval"</code> parameter, with the coefficient being the default extracted parameter. We can see that with SpatioMark, we get some features which are significant after adjusting for FDR.</p>
<div class="cell">
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Preparing outcome vector</span></span>
<span><span class="va">outcome</span> <span class="op">&lt;-</span> <span class="va">cells</span><span class="op">$</span><span class="va">group</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">cells</span><span class="op">$</span><span class="va">imageID</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">outcome</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">cells</span><span class="op">$</span><span class="va">imageID</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">cells</span><span class="op">$</span><span class="va">imageID</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Preparing features for Statial</span></span>
<span><span class="va">distMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/prepMatrix.html">prepMatrix</a></span><span class="op">(</span><span class="va">state_dist</span><span class="op">)</span></span>
<span></span>
<span><span class="va">distMat</span> <span class="op">&lt;-</span> <span class="va">distMat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">outcome</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Remove some very small values</span></span>
<span><span class="va">distMat</span> <span class="op">&lt;-</span> <span class="va">distMat</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">distMat</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0.0001</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">.8</span><span class="op">]</span></span>
<span></span>
<span><span class="va">survivalResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/colTest.html">colTest</a></span><span class="op">(</span><span class="va">distMat</span>, <span class="va">outcome</span>, type <span class="op">=</span> <span class="st">"ttest"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">survivalResults</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                    mean in group NP mean in group P tval.t    pval adjPval
MC__MC__VISTA               -0.00610        -2.1e-03   -3.8 0.00051    0.16
MC__MC__CADM1               -0.00580        -2.2e-03   -3.6 0.00110    0.16
MC__BC__OX40                -0.00043         5.9e-05   -3.6 0.00110    0.16
MC__BC__VISTA               -0.00085        -2.2e-04   -3.4 0.00170    0.17
TC_CD8__TC_CD4__CD3         -0.00053        -1.3e-04   -3.3 0.00190    0.17
MC__MC__NFKBp65             -0.00520        -2.1e-03   -3.2 0.00250    0.18
                                cluster
MC__MC__VISTA             MC__MC__VISTA
MC__MC__CADM1             MC__MC__CADM1
MC__BC__OX40               MC__BC__OX40
MC__BC__VISTA             MC__BC__VISTA
TC_CD8__TC_CD4__CD3 TC_CD8__TC_CD4__CD3
MC__MC__NFKBp65         MC__MC__NFKBp65</code></pre>
</div>
</div>
</section><section id="kontextual-robust-quantification-of-cell-type-localisation-which-is-invariant-to-changes-in-tissue-structure" class="level3" data-number="13.10.2"><h3 data-number="13.10.2" class="anchored" data-anchor-id="kontextual-robust-quantification-of-cell-type-localisation-which-is-invariant-to-changes-in-tissue-structure">
<span class="header-section-number">13.10.2</span> Kontextual: Robust quantification of cell type localisation which is invariant to changes in tissue structure</h3>
<p><code>Kontextual</code> is a method to evaluate the localisation relationship between two cell types in an image. <code>Kontextual</code> builds on the L-function by contextualising the relationship between two cell types in reference to the typical spatial behaviour of a <span class="math inline">\(3^{rd}\)</span> cell type/population. By taking this approach, <code>Kontextual</code> is invariant to changes in the window of the image as well as tissue structures which may be present.</p>
<p>The definitions of cell types and cell states are somewhat ambiguous, cell types imply well defined groups of cells that serve different roles from one another, on the other hand cell states imply that cells are a dynamic entity which cannot be discretised, and thus exist in a continuum. For the purposes of using <code>Kontextual</code> we treat cell states as identified clusters of cells, where larger clusters represent a “parent” cell population, and finer sub-clusters representing a “child” cell population. For example a CD4 T cell may be considered a child to a larger parent population of Immune cells. <code>Kontextual</code> thus aims to see how a child population of cells deviate from the spatial behaviour of their parent population, and how that influences the localisation between the child cell state and another cell state.</p>
<section id="cell-type-hierarchy" class="level4" data-number="13.10.2.1"><h4 data-number="13.10.2.1" class="anchored" data-anchor-id="cell-type-hierarchy">
<span class="header-section-number">13.10.2.1</span> Cell type hierarchy</h4>
<p>A key input for Kontextual is an annotation of cell type hierarchies. We will need these to organise all the cells present into cell state populations or clusters, e.g.&nbsp;all the different B cell types are put in a vector called bcells.</p>
<p>Here, we use the <code>treeKor</code> bioconductor package <a href="http://www.bioconductor.org/packages/release/bioc/html/treekoR.html">treekoR</a> to define these hierarchies in a data driven way.</p>
<div class="cell">
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">exprs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">cells</span>, <span class="st">"norm"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">fergusonTree</span> <span class="op">&lt;-</span> <span class="fu">treekoR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/treekoR/man/getClusterTree.html">getClusterTree</a></span><span class="op">(</span><span class="va">exprs</span>,</span>
<span>                                        <span class="va">cells</span><span class="op">$</span><span class="va">cellType</span>,</span>
<span>                                        hierarchy_method <span class="op">=</span> <span class="st">"hopach"</span>,</span>
<span>                                        scale_exprs <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">parent1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"TC_CD8"</span>, <span class="st">"TC_CD4"</span>, <span class="st">"DC"</span><span class="op">)</span></span>
<span><span class="va">parent2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"BC"</span>, <span class="st">"GC"</span><span class="op">)</span></span>
<span><span class="va">parent3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">parent1</span>, <span class="va">parent2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">parent4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MC"</span>, <span class="st">"EP"</span>, <span class="st">"SC"</span><span class="op">)</span></span>
<span><span class="va">parent5</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">parent4</span>, <span class="st">"EC"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">all</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">parent1</span>, <span class="va">parent2</span>, <span class="va">parent3</span>, <span class="va">parent4</span>, <span class="va">parent5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">treeDf</span> <span class="op">=</span> <span class="fu">Statial</span><span class="fu">::</span><span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/parentCombinations.html">parentCombinations</a></span><span class="op">(</span><span class="va">all</span>, <span class="va">parent1</span>, <span class="va">parent2</span>, <span class="va">parent3</span>, <span class="va">parent4</span>, <span class="va">parent5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fergusonTree</span><span class="op">$</span><span class="va">clust_tree</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-parentCombinations-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><code>Kontextual</code> accepts a <code>SingleCellExperiment</code> object, a single image, or list of images from a <code>SingleCellExperiment</code> object, which gets passed into the <code>cells</code> argument. Here, we’ve specified Kontextual to perform calculations on all pairwise combinations for every cluster using the <code><a href="https://sydneybiox.github.io/Statial/reference/parentCombinations.html">parentCombinations()</a></code> function to create the <code>treeDf</code> dataframe which we’ve specified in the <code>parentDf</code> parameter. The argument <code>r</code> will specify the radius which the cell relationship will be evaluated on. <code>Kontextual</code> supports parallel processing, the number of cores can be specified using the <code>cores</code> argument. <code>Kontextual</code> can take a single value or multiple values for each argument and will test all combinations of the arguments specified.</p>
<p>We can calculate all pairwise relationships across all images for a single radius.</p>
<div class="cell">
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">kontext</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/Kontextual.html">Kontextual</a></span><span class="op">(</span></span>
<span>  cells <span class="op">=</span> <span class="va">cells</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>  spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"m.cx"</span>, <span class="st">"m.cy"</span><span class="op">)</span>,</span>
<span>  parentDf <span class="op">=</span> <span class="va">treeDf</span>,</span>
<span>  r <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  cores <span class="op">=</span> <span class="va">nCores</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Again, we can use the same <code><a href="https://sydneybiox.github.io/spicyR/reference/colTest.html">colTest()</a></code> to quickly test for associations between the Kontextual values and progression status using either Wilcoxon rank sum tests or t-tests. Similar to SpatioMark, we can specify using either the original L-function by specifying <code>column = "original"</code> in our <code><a href="https://sydneybiox.github.io/Statial/reference/prepMatrix.html">prepMatrix()</a></code> function.</p>
<div class="cell">
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Converting Kontextual result into data matrix</span></span>
<span><span class="va">kontextMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/Statial/reference/prepMatrix.html">prepMatrix</a></span><span class="op">(</span><span class="va">kontext</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Replace NAs with 0</span></span>
<span><span class="va">kontextMat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">kontextMat</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span></span>
<span><span class="va">survivalResults</span> <span class="op">&lt;-</span> <span class="fu">spicyR</span><span class="fu">::</span><span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/colTest.html">colTest</a></span><span class="op">(</span><span class="va">kontextMat</span>, <span class="va">outcome</span>, type <span class="op">=</span> <span class="st">"ttest"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">survivalResults</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                    mean in group NP mean in group P tval.t    pval adjPval
MC__TC_CD4__parent3            -1.10             3.4   -3.7 0.00086    0.12
EP__BC__parent3                -4.90             6.7   -3.3 0.00200    0.13
EP__BC__parent2                 0.73             9.8   -3.1 0.00500    0.13
DC__BC__parent3                -4.00             2.3   -2.9 0.00580    0.13
DC__SC__parent5                 0.24             7.9   -2.9 0.00590    0.13
MC__TC_CD4__parent1             0.13             3.7   -2.9 0.00600    0.13
                                cluster
MC__TC_CD4__parent3 MC__TC_CD4__parent3
EP__BC__parent3         EP__BC__parent3
EP__BC__parent2         EP__BC__parent2
DC__BC__parent3         DC__BC__parent3
DC__SC__parent5         DC__SC__parent5
MC__TC_CD4__parent1 MC__TC_CD4__parent1</code></pre>
</div>
</div>
</section></section></section><section id="classifyr-classification" class="level2" data-number="13.11"><h2 data-number="13.11" class="anchored" data-anchor-id="classifyr-classification">
<span class="header-section-number">13.11</span> ClassifyR: Classification</h2>
<p>Our ClassifyR package, <a href="https://github.com/SydneyBioX/ClassifyR" class="uri">https://github.com/SydneyBioX/ClassifyR</a>, formalises a convenient framework for evaluating classification in R. We provide functionality to easily include four key modelling stages; Data transformation, feature selection, classifier training and prediction; into a cross-validation loop. Here we use the <code>crossValidate</code> function to perform 100 repeats of 5-fold cross-validation to evaluate the performance of a random forest applied to five quantifications of our IMC data; 1) Cell type proportions 2) Cell type localisation from <code>spicyR</code> 3) Region proportions from <code>lisaClust</code> 4) Cell type localisation in reference to a parent cell type from <code>Kontextual</code> 5) Cell changes in response to proximal changes from <code>SpatioMark</code></p>
<div class="cell">
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Create list to store data.frames</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add proportions of each cell type in each image</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"Proportions"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/getProp.html">getProp</a></span><span class="op">(</span><span class="va">cells</span>, <span class="st">"cellType"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add pair-wise associations</span></span>
<span><span class="va">spicyMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/bind.html">bind</a></span><span class="op">(</span><span class="va">spicyTest</span><span class="op">)</span></span>
<span><span class="va">spicyMat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">spicyMat</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="va">spicyMat</span> <span class="op">&lt;-</span> <span class="va">spicyMat</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">!</span><span class="va">condition</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span><span class="op">(</span><span class="st">"imageID"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"SpicyR"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">spicyMat</span></span>
<span></span>
<span><span class="co"># Add proportions of each region in each image</span></span>
<span><span class="co"># to the list of dataframes.</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"LisaClust"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/spicyR/reference/getProp.html">getProp</a></span><span class="op">(</span><span class="va">cells</span>, <span class="st">"region"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Add SpatioMark features</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"SpatioMark"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">distMat</span></span>
<span></span>
<span><span class="co"># Add Kontextual features</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"Kontextual"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">kontextMat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Set seed</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform cross-validation of a random forest model</span></span>
<span><span class="co"># with 100 repeats of 5-fold cross-validation.</span></span>
<span><span class="va">cv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/crossValidate.html">crossValidate</a></span><span class="op">(</span></span>
<span>  measurements <span class="op">=</span> <span class="va">data</span>,</span>
<span>  outcome <span class="op">=</span> <span class="va">outcome</span>,</span>
<span>  classifier <span class="op">=</span> <span class="st">"randomForest"</span>,</span>
<span>  nFolds <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  nRepeats <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  nCores <span class="op">=</span> <span class="va">nCores</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="visualise-cross-validated-prediction-performance" class="level3" data-number="13.11.1"><h3 data-number="13.11.1" class="anchored" data-anchor-id="visualise-cross-validated-prediction-performance">
<span class="header-section-number">13.11.1</span> Visualise cross-validated prediction performance</h3>
<p>Here we use the <code>performancePlot</code> function to assess the AUC from each repeat of the 5-fold cross-validation. We see that the lisaClust regions appear to capture information which is predictive of progression status of the patients.</p>
<div class="cell">
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Calculate AUC for each cross-validation repeat and plot.</span></span>
<span><span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/performancePlot.html">performancePlot</a></span><span class="op">(</span></span>
<span>  <span class="va">cv</span>,</span>
<span>  metric <span class="op">=</span> <span class="st">"AUC"</span>,</span>
<span>  characteristicsList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Assay Name"</span><span class="op">)</span>,</span>
<span>  orderingList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"Assay Name"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Proportions"</span>, <span class="st">"SpicyR"</span>, <span class="st">"LisaClust"</span>, <span class="st">"Kontextual"</span>, <span class="st">"SpatioMark"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in .local(results, ...): AUC not found in all elements of results.
Calculating it now.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-performancePlot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can also visualise which features were good at classifying which patients using the <code>sampleMetricMap()</code> function from <code>ClassifyR</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/samplesMetricMap.html">samplesMetricMap</a></span><span class="op">(</span><span class="va">cv</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="08-case_study1_files/figure-html/08-samplesMetricMap-1.png" class="img-fluid figure-img" width="1152"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>TableGrob (2 x 1) "arrange": 2 grobs
  z     cells    name                 grob
1 1 (2-2,1-1) arrange       gtable[layout]
2 2 (1-1,1-1) arrange text[GRID.text.1955]</code></pre>
</div>
</div>
</section></section><section id="summary" class="level2" data-number="13.12"><h2 data-number="13.12" class="anchored" data-anchor-id="summary">
<span class="header-section-number">13.12</span> Summary</h2>
<p>Here we have used a pipeline of our spatial analysis R packages to demonstrate an easy way to segment, cluster, normalise, quantify and classify high dimensional in situ cytometry data all within R.</p>
</section><section id="sessioninfo" class="level2" data-number="13.13"><h2 data-number="13.13" class="anchored" data-anchor-id="sessioninfo">
<span class="header-section-number">13.13</span> sessionInfo</h2>
<div class="cell">
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.4.1 (2024-06-14)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.4.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Australia/Sydney
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] SpatialDatasets_1.4.0           ExperimentHub_2.14.0           
 [3] AnnotationHub_3.14.0            BiocFileCache_2.14.0           
 [5] dbplyr_2.5.0                    SpatialExperiment_1.16.0       
 [7] ttservice_0.4.1                 tidyr_1.3.1                    
 [9] tidySingleCellExperiment_1.16.0 Statial_1.8.0                  
[11] lisaClust_1.14.4                ClassifyR_3.10.5               
[13] survival_3.7-0                  BiocParallel_1.40.0            
[15] MultiAssayExperiment_1.32.0     generics_0.1.3                 
[17] spicyR_1.18.0                   scater_1.34.0                  
[19] scuttle_1.16.0                  ggpubr_0.6.0                   
[21] FuseSOM_1.8.0                   simpleSeg_1.8.0                
[23] ggplot2_3.5.1                   dplyr_1.1.4                    
[25] cytomapper_1.18.0               SingleCellExperiment_1.28.1    
[27] SummarizedExperiment_1.36.0     Biobase_2.66.0                 
[29] GenomicRanges_1.58.0            GenomeInfoDb_1.42.0            
[31] IRanges_2.40.0                  S4Vectors_0.44.0               
[33] BiocGenerics_0.52.0             MatrixGenerics_1.18.0          
[35] matrixStats_1.4.1               EBImage_4.48.0                 

loaded via a namespace (and not attached):
  [1] tiff_0.1-12                 dcanr_1.22.0               
  [3] FCPS_1.3.4                  nnet_7.3-19                
  [5] goftest_1.2-3               Biostrings_2.74.0          
  [7] HDF5Array_1.34.0            TH.data_1.1-2              
  [9] vctrs_0.6.5                 spatstat.random_3.3-2      
 [11] shape_1.4.6.1               digest_0.6.37              
 [13] png_0.1-8                   proxy_0.4-27               
 [15] ggrepel_0.9.6               deldir_2.0-4               
 [17] permute_0.9-7               magick_2.8.5               
 [19] MASS_7.3-61                 reshape2_1.4.4             
 [21] httpuv_1.6.15               foreach_1.5.2              
 [23] withr_3.0.2                 ggfun_0.1.7                
 [25] psych_2.4.6.26              xfun_0.49                  
 [27] ellipsis_0.3.2              doRNG_1.8.6.1              
 [29] memoise_2.0.1               ggbeeswarm_0.7.2           
 [31] RProtoBufLib_2.18.0         diptest_0.77-1             
 [33] princurve_2.1.6             systemfonts_1.1.0          
 [35] tidytree_0.4.6              zoo_1.8-12                 
 [37] GlobalOptions_0.1.2         V8_6.0.0                   
 [39] DEoptimR_1.1-3-1            Formula_1.2-5              
 [41] prabclus_2.3-4              KEGGREST_1.46.0            
 [43] promises_1.3.2              httr_1.4.7                 
 [45] rstatix_0.7.2               rhdf5filters_1.18.0        
 [47] fpc_2.2-13                  rhdf5_2.50.0               
 [49] rstudioapi_0.17.1           UCSC.utils_1.2.0           
 [51] concaveman_1.1.0            curl_6.0.1                 
 [53] zlibbioc_1.52.0             ScaledMatrix_1.14.0        
 [55] analogue_0.17-7             polyclip_1.10-7            
 [57] GenomeInfoDbData_1.2.13     SparseArray_1.6.0          
 [59] fftwtools_0.9-11            xtable_1.8-4               
 [61] stringr_1.5.1               doParallel_1.0.17          
 [63] evaluate_1.0.1              S4Arrays_1.6.0             
 [65] irlba_2.3.5.1               colorspace_2.1-1           
 [67] filelock_1.0.3              spatstat.data_3.1-4        
 [69] flexmix_2.3-19              magrittr_2.0.3             
 [71] ggtree_3.14.0               later_1.4.1                
 [73] viridis_0.6.5               modeltools_0.2-23          
 [75] lattice_0.22-6              genefilter_1.88.0          
 [77] spatstat.geom_3.3-4         robustbase_0.99-4-1        
 [79] XML_3.99-0.17               cowplot_1.1.3              
 [81] RcppAnnoy_0.0.22            ggupset_0.4.0              
 [83] class_7.3-22                svgPanZoom_0.3.4           
 [85] pillar_1.9.0                nlme_3.1-166               
 [87] iterators_1.0.14            compiler_4.4.1             
 [89] beachmat_2.22.0             stringi_1.8.4              
 [91] tensor_1.5                  minqa_1.2.8                
 [93] plyr_1.8.9                  treekoR_1.14.0             
 [95] crayon_1.5.3                abind_1.4-8                
 [97] gridGraphics_0.5-1          locfit_1.5-9.10            
 [99] sp_2.1-4                    bit_4.5.0                  
[101] terra_1.7-78                sandwich_3.1-1             
[103] multcomp_1.4-26             fastcluster_1.2.6          
[105] codetools_0.2-20            BiocSingular_1.22.0        
[107] coop_0.6-3                  GetoptLong_1.0.5           
[109] plotly_4.10.4               mime_0.12                  
[111] splines_4.4.1               circlize_0.4.16            
[113] Rcpp_1.0.13-1               profileModel_0.6.1         
[115] knitr_1.49                  blob_1.2.4                 
[117] utf8_1.2.4                  clue_0.3-66                
[119] BiocVersion_3.20.0          lme4_1.1-35.5              
[121] fs_1.6.5                    nnls_1.6                   
[123] ggsignif_0.6.4              ggplotify_0.1.2            
[125] tibble_3.2.1                Matrix_1.7-1               
[127] scam_1.2-17                 statmod_1.5.0              
[129] svglite_2.1.3               tweenr_2.0.3               
[131] pkgconfig_2.0.3             pheatmap_1.0.12            
[133] tools_4.4.1                 cachem_1.1.0               
[135] RSQLite_2.3.8               viridisLite_0.4.2          
[137] DBI_1.2.3                   numDeriv_2016.8-1.1        
[139] fastmap_1.2.0               rmarkdown_2.29             
[141] scales_1.3.0                grid_4.4.1                 
[143] shinydashboard_0.7.2        broom_1.0.7                
[145] patchwork_1.3.0             brglm_0.7.2                
[147] BiocManager_1.30.25         carData_3.0-5              
[149] farver_2.1.2                mgcv_1.9-1                 
[151] yaml_2.3.10                 ggthemes_5.1.0             
[153] cli_3.6.3                   purrr_1.0.2                
[155] hopach_2.66.0               lifecycle_1.0.4            
[157] uwot_0.2.2                  mvtnorm_1.3-2              
[159] kernlab_0.9-33              backports_1.5.0            
[161] annotate_1.84.0             cytolib_2.18.0             
[163] gtable_0.3.6                rjson_0.2.23               
[165] parallel_4.4.1              ape_5.8                    
[167] limma_3.62.1                edgeR_4.4.0                
[169] jsonlite_1.8.9              bitops_1.0-9               
[171] bit64_4.5.2                 Rtsne_0.17                 
[173] FlowSOM_2.14.0              yulab.utils_0.1.8          
[175] vegan_2.6-8                 spatstat.utils_3.1-1       
[177] BiocNeighbors_2.0.0         ranger_0.17.0              
[179] flowCore_2.18.0             bdsmatrix_1.3-7            
[181] spatstat.univar_3.1-1       lazyeval_0.2.2             
[183] ConsensusClusterPlus_1.70.0 shiny_1.9.1                
[185] htmltools_0.5.8.1           diffcyt_1.26.0             
[187] rappdirs_0.3.3              glue_1.8.0                 
[189] XVector_0.46.0              RCurl_1.98-1.16            
[191] treeio_1.30.0               mclust_6.1.1               
[193] mnormt_2.1.1                coxme_2.2-22               
[195] jpeg_0.1-10                 gridExtra_2.3              
[197] boot_1.3-31                 igraph_2.1.1               
[199] R6_2.5.1                    ggiraph_0.8.11             
[201] labeling_0.4.3              ggh4x_0.2.8                
[203] cluster_2.1.6               rngtools_1.5.2             
[205] Rhdf5lib_1.28.0             aplot_0.2.3                
[207] nloptr_2.1.1                DelayedArray_0.32.0        
[209] tidyselect_1.2.1            vipor_0.4.7                
[211] ggforce_0.4.2               raster_3.6-30              
[213] car_3.1-3                   AnnotationDbi_1.68.0       
[215] rsvd_1.0.5                  munsell_0.5.1              
[217] DataVisualizations_1.3.2    data.table_1.16.2          
[219] htmlwidgets_1.6.4           ComplexHeatmap_2.22.0      
[221] RColorBrewer_1.1-3          rlang_1.1.4                
[223] spatstat.sparse_3.1-0       spatstat.explore_3.3-3     
[225] colorRamps_2.3.4            lmerTest_3.1-3             
[227] uuid_1.2-1                  ggnewscale_0.5.0           
[229] fansi_1.0.6                 beeswarm_0.4.0             </code></pre>
</div>
</div>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./07-classification.html" class="pagination-link" aria-label="Finding associations between clinical variables and spatial features">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Finding associations between clinical variables and spatial features</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./09-references.html" class="pagination-link" aria-label="References">
        <span class="nav-page-text">References</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>