<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.555">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>spicyPlayBook - 4&nbsp; Cell localisation</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./05-cellular_niches.html" rel="next">
<link href="./03-cell_annotation.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="./index.html" class="navbar-brand navbar-brand-logo">
    <img src="./images/USydLogo.svg" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">spicyPlayBook</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./04-cell_localisation.html"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Cell localisation</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01-processing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Processing</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02-quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Quality Control</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03-cell_annotation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Cell Annotation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04-cell_localisation.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Cell localisation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05-cellular_niches.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Spatial domains</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06-changes_in_marker_expression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Marker expression</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07-classification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Classification</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08-case_study1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Case Study: Head and Neck Squamous Cell Carcinoma (Ferguson et al., 2022)</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./09-references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#spicyr-cell-localisation" id="toc-spicyr-cell-localisation" class="nav-link active" data-scroll-target="#spicyr-cell-localisation"><span class="header-section-number">4.1</span> spicyR: Cell localisation</a>
  <ul class="collapse">
  <li><a href="#linear-modelling" id="toc-linear-modelling" class="nav-link" data-scroll-target="#linear-modelling"><span class="header-section-number">4.1.1</span> Linear modelling</a></li>
  <li><a href="#test-for-changes-in-localisation-for-a-specific-pair-of-cells" id="toc-test-for-changes-in-localisation-for-a-specific-pair-of-cells" class="nav-link" data-scroll-target="#test-for-changes-in-localisation-for-a-specific-pair-of-cells"><span class="header-section-number">4.1.2</span> Test for changes in localisation for a specific pair of cells</a></li>
  <li><a href="#test-for-changes-in-localisation-for-all-pairwise-cell-combinations" id="toc-test-for-changes-in-localisation-for-all-pairwise-cell-combinations" class="nav-link" data-scroll-target="#test-for-changes-in-localisation-for-all-pairwise-cell-combinations"><span class="header-section-number">4.1.3</span> Test for changes in localisation for all pairwise cell combinations</a></li>
  <li><a href="#linear-modelling-for-custom-metrics" id="toc-linear-modelling-for-custom-metrics" class="nav-link" data-scroll-target="#linear-modelling-for-custom-metrics"><span class="header-section-number">4.1.4</span> Linear modelling for custom metrics</a></li>
  <li><a href="#performing-survival-analysis" id="toc-performing-survival-analysis" class="nav-link" data-scroll-target="#performing-survival-analysis"><span class="header-section-number">4.1.5</span> Performing survival analysis</a></li>
  <li><a href="#accounting-for-tissue-inhomogeneity" id="toc-accounting-for-tissue-inhomogeneity" class="nav-link" data-scroll-target="#accounting-for-tissue-inhomogeneity"><span class="header-section-number">4.1.6</span> Accounting for tissue inhomogeneity</a></li>
  <li><a href="#mixed-effects-modelling" id="toc-mixed-effects-modelling" class="nav-link" data-scroll-target="#mixed-effects-modelling"><span class="header-section-number">4.1.7</span> Mixed effects modelling</a></li>
  </ul></li>
  <li><a href="#kontextual-context-aware-cell-localisation" id="toc-kontextual-context-aware-cell-localisation" class="nav-link" data-scroll-target="#kontextual-context-aware-cell-localisation"><span class="header-section-number">4.2</span> Kontextual: Context aware cell localisation</a>
  <ul class="collapse">
  <li><a href="#using-cell-type-hierarchies-to-define-a-context" id="toc-using-cell-type-hierarchies-to-define-a-context" class="nav-link" data-scroll-target="#using-cell-type-hierarchies-to-define-a-context"><span class="header-section-number">4.2.1</span> Using cell type hierarchies to define a “context”</a></li>
  <li><a href="#application-on-triple-negative-breast-cancer-image." id="toc-application-on-triple-negative-breast-cancer-image." class="nav-link" data-scroll-target="#application-on-triple-negative-breast-cancer-image."><span class="header-section-number">4.2.2</span> Application on triple negative breast cancer image.</a></li>
  <li><a href="#calculating-all-pairwise-relationships" id="toc-calculating-all-pairwise-relationships" class="nav-link" data-scroll-target="#calculating-all-pairwise-relationships"><span class="header-section-number">4.2.3</span> Calculating all pairwise relationships</a></li>
  <li><a href="#associating-the-relationships-with-survival-outcomes." id="toc-associating-the-relationships-with-survival-outcomes." class="nav-link" data-scroll-target="#associating-the-relationships-with-survival-outcomes."><span class="header-section-number">4.2.4</span> Associating the relationships with survival outcomes.</a></li>
  </ul></li>
  <li><a href="#sessioninfo" id="toc-sessioninfo" class="nav-link" data-scroll-target="#sessioninfo"><span class="header-section-number">4.3</span> sessionInfo</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Cell localisation</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>Now that we’ve finished performing all of our upstream preprocessing (segmentation, quality control, and annotation), we can now begin dissecting out interesting findings for our datasets. One of the primary motivations behind pursuing a spatial technology as opposed to a space-agnostic technology such as single-cell RNA sequencing, is that we are more able to tease out whether changes are occurring spatially, i.e.&nbsp;are two cell types closer together in a disease state vs a non-disease state. Whilst these changes are usually visually obvious, to quantify localisation and dispersion relationships, more advanced statistical modelling is required.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load required packages</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(spicyR)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Statial)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SpatialExperiment)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SpatialDatasets)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(imcRtools)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(survival)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tibble)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(treekoR)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggsurvfit)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>nCores <span class="ot">&lt;-</span> <span class="dv">2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="spicyr-cell-localisation" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="spicyr-cell-localisation"><span class="header-section-number">4.1</span> spicyR: Cell localisation</h2>
<p>This section provides step-by-step instructions on assessing how the localisation of different cell types changes across different disease conditions.</p>
<p>We use the <span class="citation" data-cites="keren2018">(<a href="09-references.html#ref-keren2018" role="doc-biblioref"><strong>keren2018?</strong></a>)</span> breast cancer dataset to compare the spatial distribution of immune cells in individuals with different levels of tumour infiltration (cold and compartmentalised).</p>
<p>The data is stored as a <code>SpatialExperiment</code> object and contains single-cell spatial data from 41 images.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>kerenSPE <span class="ot">&lt;-</span> SpatialDatasets<span class="sc">::</span><span class="fu">spe_Keren_2018</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The cell types in this dataset includes 11 immune cell types (double negative CD3 T cells, CD4 T cells, B cells, monocytes, macrophages, CD8 T cells, neutrophils, natural killer cells, dendritic cells, regulatory T cells), 2 structural cell types (endothelial, mesenchymal), 2 tumour cell types (keratin+ tumour, tumour) and one unidentified category. Below we’ve provided detailed information on the statistical backend of <code>spicyR</code>.</p>
<section id="linear-modelling" class="level3" data-number="4.1.1">
<h3 data-number="4.1.1" class="anchored" data-anchor-id="linear-modelling"><span class="header-section-number">4.1.1</span> Linear modelling</h3>
<p>To investigate changes in localisation between two different cell types, we measure the level of localisation between two cell types by modelling with the L-function. The L-function is a variance-stabilised K-function given by the equation</p>
<p><span class="math display">\[
\widehat{L_{ij}} (r) = \sqrt{\frac{\widehat{K_{ij}}(r)}{\pi}}
\]</span></p>
<p>with <span class="math inline">\(\widehat{K_{ij}}\)</span> defined as</p>
<p><span class="math display">\[
\widehat{K_{ij}} (r) = \frac{|W|}{n_i n_j} \sum_{n_i} \sum_{n_j} 1 \{d_{ij} \leq r \} e_{ij} (r)
\]</span></p>
<p>where <span class="math inline">\(\widehat{K_{ij}}\)</span> summarises the degree of co-localisation of cell type <span class="math inline">\(j\)</span> with cell type <span class="math inline">\(i\)</span>, <span class="math inline">\(n_i\)</span> and <span class="math inline">\(n_j\)</span> are the number of cells of type <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span>, <span class="math inline">\(|W|\)</span> is the image area, <span class="math inline">\(d_{ij}\)</span> is the distance between two cells and <span class="math inline">\(e_{ij} (r)\)</span> is an edge correcting factor.</p>
<p>Specifically, the mean difference between the experimental function and the theoretical function is used as a measure for the level of localisation, defined as</p>
<p><span class="math display">\[
u = \sum_{r' = r_{\text{min}}}^{r_{\text{max}}} \widehat L_{ij, \text{Experimental}} (r') - \widehat L_{ij, \text{Poisson}} (r')
\]</span></p>
<p>where <span class="math inline">\(u\)</span> is the sum is taken over a discrete range of <span class="math inline">\(r\)</span> between <span class="math inline">\(r_{\text{min}}\)</span> and <span class="math inline">\(r_{\text{max}}\)</span>. Differences of the statistic <span class="math inline">\(u\)</span> between two conditions is modelled using a weighted linear model.</p>
</section>
<section id="test-for-changes-in-localisation-for-a-specific-pair-of-cells" class="level3" data-number="4.1.2">
<h3 data-number="4.1.2" class="anchored" data-anchor-id="test-for-changes-in-localisation-for-a-specific-pair-of-cells"><span class="header-section-number">4.1.2</span> Test for changes in localisation for a specific pair of cells</h3>
<p>Firstly, we can test whether one cell type tends to be more localised with another cell type in one condition compared to the other. This can be done using the <code>spicy()</code> function, where we specify the <code>condition</code> parameter.</p>
<p>In this example, we want to see whether or not neutrophils (<code>to</code>) tend to be found around CD8 T cells (<code>from</code>) in compartmentalised tumours compared to cold tumours. Given that there are 3 conditions, we can specify the desired conditions by setting the order of our <code>condition</code> factor. <code>spicy</code> will choose the first level of the factor as the base condition and the second level as the comparison condition. <code>spicy</code> will also naturally coerce the <code>condition</code> column into a factor if it is not already a factor. The column containing cell type annotations can be specified using the <code>cellTypeCol</code> argument. By default, <code>spicy</code> uses the column named <code>cellType</code> in the <code>SpatialExperiment</code> object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>spicyTestPair <span class="ot">&lt;-</span> <span class="fu">spicy</span>(</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  kerenSPE,</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">condition =</span> <span class="st">"tumour_type"</span>,</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">from =</span> <span class="st">"CD8_T_cell"</span>,</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">to =</span> <span class="st">"Neutrophils"</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="fu">topPairs</span>(spicyTestPair)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                        intercept coefficient      p.value   adj.pvalue
CD8_T_cell__Neutrophils  -109.081    112.0185 2.166646e-05 2.166646e-05
                              from          to
CD8_T_cell__Neutrophils CD8_T_cell Neutrophils</code></pre>
</div>
</div>
<p>We obtain a <code>spicy</code> object which details the results of the modelling performed. The <code>topPairs()</code> function can be used to obtain the associated coefficients and p-value.</p>
<p>As the <code>coefficient</code> in <code>spicyTestPair</code> is positive, we find that neutrophils are significantly more likely to be found near CD8 T cells in the compartmentalised tumours group compared to the cold tumour group.</p>
</section>
<section id="test-for-changes-in-localisation-for-all-pairwise-cell-combinations" class="level3" data-number="4.1.3">
<h3 data-number="4.1.3" class="anchored" data-anchor-id="test-for-changes-in-localisation-for-all-pairwise-cell-combinations"><span class="header-section-number">4.1.3</span> Test for changes in localisation for all pairwise cell combinations</h3>
<p>We can perform what we did above for all pairwise combinations of cell types by excluding the <code>from</code> and <code>to</code> parameters in <code>spicy()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>spicyTest <span class="ot">&lt;-</span> <span class="fu">spicy</span>(</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  kerenSPE,</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">condition =</span> <span class="st">"tumour_type"</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="fu">topPairs</span>(spicyTest)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                            intercept coefficient      p.value   adj.pvalue
Macrophages__dn_T_CD3       56.446064   -50.08474 1.080273e-07 3.035568e-05
dn_T_CD3__Macrophages       54.987151   -48.38664 2.194018e-07 3.082595e-05
Macrophages__DC_or_Mono     73.239404   -59.90361 5.224660e-06 4.893765e-04
DC_or_Mono__Macrophages     71.777087   -58.46833 7.431172e-06 5.220399e-04
dn_T_CD3__dn_T_CD3         -63.786032   100.61010 2.878804e-05 1.208706e-03
Neutrophils__dn_T_CD3      -63.141840    69.64356 2.891872e-05 1.208706e-03
dn_T_CD3__Neutrophils      -63.133725    70.15508 3.011012e-05 1.208706e-03
DC__Macrophages             96.893239   -92.55112 1.801300e-04 5.758129e-03
Macrophages__DC             96.896215   -93.25194 1.844241e-04 5.758129e-03
CD4_T_cell__Keratin_Tumour  -4.845037   -22.14995 2.834659e-04 7.409016e-03
                                  from             to
Macrophages__dn_T_CD3      Macrophages       dn_T_CD3
dn_T_CD3__Macrophages         dn_T_CD3    Macrophages
Macrophages__DC_or_Mono    Macrophages     DC_or_Mono
DC_or_Mono__Macrophages     DC_or_Mono    Macrophages
dn_T_CD3__dn_T_CD3            dn_T_CD3       dn_T_CD3
Neutrophils__dn_T_CD3      Neutrophils       dn_T_CD3
dn_T_CD3__Neutrophils         dn_T_CD3    Neutrophils
DC__Macrophages                     DC    Macrophages
Macrophages__DC            Macrophages             DC
CD4_T_cell__Keratin_Tumour  CD4_T_cell Keratin_Tumour</code></pre>
</div>
</div>
<p>Again, we obtain a <code>spicy</code> object which outlines the result of the linear models performed for each pairwise combination of cell types.</p>
<p>We can also examine the L-function metrics of individual images by using the convenient <code>bind()</code> function on our <code>spicyTest</code> results object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">bind</span>(spicyTest)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  imageID         condition Keratin_Tumour__Keratin_Tumour
1       1             mixed                      -2.300602
2       2             mixed                      -1.989699
3       3 compartmentalised                      11.373530
4       4 compartmentalised                      33.931133
5       5 compartmentalised                      17.922818
  dn_T_CD3__Keratin_Tumour B_cell__Keratin_Tumour
1                -5.298543             -20.827279
2               -16.020022               3.025815
3               -21.857447             -24.962913
4               -36.438476             -40.470221
5               -20.816783             -38.138076</code></pre>
</div>
</div>
<p>The results can be represented as a bubble plot using the <code>signifPlot()</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">signifPlot</span>(</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  spicyTest,</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">breaks =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">3</span>, <span class="dv">3</span>, <span class="dv">1</span>),</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">marksToPlot =</span> <span class="fu">c</span>(<span class="st">"Macrophages"</span>, <span class="st">"DC_or_Mono"</span>, <span class="st">"dn_T_CD3"</span>, <span class="st">"Neutrophils"</span>,</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>                  <span class="st">"CD8_T_cell"</span>, <span class="st">"Keratin_Tumour"</span>)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="04-cell_localisation_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Here, we can observe that the most significant relationships occur between macrophages and double negative CD3 T cells, suggesting that the two cell types are far more dispersed in compartmentalised tumours compared to cold tumours.</p>
<p>To examine a specific cell type-cell type relationship in more detail, we can use <code>spicyBoxplot()</code> and specify either <code>from = "Macrophages"</code> and <code>to = "dn_T_CD3"</code> or <code>rank = 1</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">spicyBoxPlot</span>(<span class="at">results =</span> spicyTest, </span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>             <span class="co"># from = "Macrophages",</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>             <span class="co"># to = "dn_T_CD3"</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">rank =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 2 rows containing non-finite outside the scale range
(`stat_boxplot()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="04-cell_localisation_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="linear-modelling-for-custom-metrics" class="level3" data-number="4.1.4">
<h3 data-number="4.1.4" class="anchored" data-anchor-id="linear-modelling-for-custom-metrics"><span class="header-section-number">4.1.4</span> Linear modelling for custom metrics</h3>
<p><code>spicyR</code> can also be applied to custom distance or abundance metrics. A kNN interactions graph can be generated with the function <code>buildSpatialGraph</code> from the <code>imcRtools</code> package. This generates a <code>colPairs</code> object inside of the <code>SpatialExperiment</code> object.</p>
<p><code>spicyR</code> provides the function <code>convPairs</code> for converting a <code>colPairs</code> object into an abundance matrix by calculating the average number of nearby cells types for every cell type for a given <code>k</code>. For example, if there exists on average 5 neutrophils for every macrophage in image 1, the column <code>Neutrophil__Macrophage</code> would have a value of 5 for image 1.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>kerenSPE <span class="ot">&lt;-</span> imcRtools<span class="sc">::</span><span class="fu">buildSpatialGraph</span>(kerenSPE, </span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">img_id =</span> <span class="st">"imageID"</span>, </span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">type =</span> <span class="st">"knn"</span>, <span class="at">k =</span> <span class="dv">20</span>,</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">coords =</span> <span class="fu">c</span>(<span class="st">"x"</span>, <span class="st">"y"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'sample_id's are duplicated across 'SpatialExperiment' objects to cbind; appending sample indices.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The returned object is ordered by the 'imageID' entry.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>pairAbundances <span class="ot">&lt;-</span> <span class="fu">convPairs</span>(kerenSPE,</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">colPair =</span> <span class="st">"knn_interaction_graph"</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pairAbundances[<span class="st">"B_cell__B_cell"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   B_cell__B_cell
1      12.7349608
10      0.2777778
11      0.0000000
12      1.3333333
13      1.2200957
14      0.0000000</code></pre>
</div>
</div>
<p>The custom distance or abundance metrics can then be included in the analysis with the <code>alternateResult</code> parameter. The <code>Statial</code> package contains other custom distance metrics which can be used with <code>spicy</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>spicyTestColPairs <span class="ot">&lt;-</span> <span class="fu">spicy</span>(</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>  kerenSPE,</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">condition =</span> <span class="st">"tumour_type"</span>,</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">alternateResult =</span> pairAbundances,</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">weights =</span> <span class="cn">FALSE</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a><span class="fu">topPairs</span>(spicyTestColPairs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                           intercept coefficient     p.value adj.pvalue
CD8_T_cell__Neutrophils  0.833333333  -0.7592968 0.002645466  0.3291833
B_cell__Tumour           0.001937984   0.2602822 0.004872664  0.3291833
Other_Immune__NK         0.012698413   0.2612881 0.005673068  0.3291833
Unidentified__CD8_T_cell 0.106626794   0.6387339 0.005906526  0.3291833
dn_T_CD3__NK             0.004242424   0.2148797 0.006317829  0.3291833
CD4_T_cell__Neutrophils  0.036213602   0.2947696 0.007902670  0.3291833
Tregs__CD4_T_cell        0.128876212   0.5726201 0.010207087  0.3291833
Endothelial__DC          0.008771930   0.3008523 0.011189533  0.3291833
Tumour__Neutrophils      0.021638939   0.2529045 0.011388850  0.3291833
Mesenchymal__Neutrophils 0.004504505   0.2494301 0.012761315  0.3291833
                                 from          to
CD8_T_cell__Neutrophils    CD8_T_cell Neutrophils
B_cell__Tumour                 B_cell      Tumour
Other_Immune__NK         Other_Immune          NK
Unidentified__CD8_T_cell Unidentified  CD8_T_cell
dn_T_CD3__NK                 dn_T_CD3          NK
CD4_T_cell__Neutrophils    CD4_T_cell Neutrophils
Tregs__CD4_T_cell               Tregs  CD4_T_cell
Endothelial__DC           Endothelial          DC
Tumour__Neutrophils            Tumour Neutrophils
Mesenchymal__Neutrophils  Mesenchymal Neutrophils</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">signifPlot</span>(</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  spicyTestColPairs,</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">breaks =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">3</span>, <span class="dv">3</span>, <span class="dv">1</span>),</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">marksToPlot =</span> <span class="fu">c</span>(<span class="st">"Macrophages"</span>, <span class="st">"dn_T_CD3"</span>, <span class="st">"CD4_T_cell"</span>, </span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>                  <span class="st">"B_cell"</span>, <span class="st">"DC_or_Mono"</span>, <span class="st">"Neutrophils"</span>, <span class="st">"CD8_T_cell"</span>)</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="04-cell_localisation_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="performing-survival-analysis" class="level3" data-number="4.1.5">
<h3 data-number="4.1.5" class="anchored" data-anchor-id="performing-survival-analysis"><span class="header-section-number">4.1.5</span> Performing survival analysis</h3>
<p><code>spicy</code> can also be used to perform survival analysis to asses whether changes in co-localisation between cell types are associated with survival probability. <code>spicy</code> requires the <code>SingleCellExperiment</code> object being used to contain a column called <code>survival</code> as a <code>Surv</code> object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>kerenSPE<span class="sc">$</span>event <span class="ot">=</span> <span class="dv">1</span> <span class="sc">-</span> kerenSPE<span class="sc">$</span>Censored</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>kerenSPE<span class="sc">$</span>survival <span class="ot">=</span> <span class="fu">Surv</span>(kerenSPE<span class="sc">$</span><span class="st">`</span><span class="at">Survival_days_capped*</span><span class="st">`</span>, kerenSPE<span class="sc">$</span>event)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can then perform survival analysis using the <code>spicy</code> function by specifying <code>condition = "survival"</code>. We can then access the corresponding coefficients and p-values by accessing the <code>survivalResults</code> slot in the <code>spicy</code> results object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Running survival analysis</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>spicySurvival <span class="ot">=</span> <span class="fu">spicy</span>(kerenSPE,</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">condition =</span> <span class="st">"survival"</span>)</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="co"># top 10 significant pairs</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(spicySurvival<span class="sc">$</span>survivalResults, <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 4
   test                       coef se.coef    p.value
   &lt;chr&gt;                     &lt;dbl&gt;   &lt;dbl&gt;      &lt;dbl&gt;
 1 Other_Immune__Tregs     0.0236  0.00866 0.00000893
 2 CD4_T_cell__Tregs       0.0177  0.00685 0.0000124 
 3 Tregs__Other_Immune     0.0237  0.00873 0.0000126 
 4 Tregs__CD4_T_cell       0.0171  0.00676 0.0000285 
 5 CD8_T_cell__CD8_T_cell  0.00605 0.00272 0.000332  
 6 Tumour__CD8_T_cell     -0.0305  0.0114  0.000617  
 7 CD8_T_cell__Tumour     -0.0305  0.0116  0.000721  
 8 CD4_T_cell__dn_T_CD3    0.00845 0.00353 0.000794  
 9 dn_T_CD3__CD4_T_cell    0.00840 0.00353 0.000937  
10 DC__Other_Immune       -0.0289  0.0123  0.00103   </code></pre>
</div>
</div>
</section>
<section id="accounting-for-tissue-inhomogeneity" class="level3" data-number="4.1.6">
<h3 data-number="4.1.6" class="anchored" data-anchor-id="accounting-for-tissue-inhomogeneity"><span class="header-section-number">4.1.6</span> Accounting for tissue inhomogeneity</h3>
<p>The <code>spicy</code> function can also account for tissue inhomogeneity to avoid false positives or negatives. This can be done by setting the <code>sigma =</code> parameter within the spicy function. By default, <code>sigma</code> is set to <code>NULL</code>, and <code>spicy</code> assumes a homogeneous tissue structure.</p>
<p>For example, when we examine the L-function for <code>Keratin_Tumour__Neutrophils</code> when <code>sigma = NULL</code> and <code>Rs = 100</code>, the value is positive, indicating attraction between the two cell types.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># filter SPE object to obtain image 24 data</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>kerenSubset <span class="ot">=</span> kerenSPE[, <span class="fu">colData</span>(kerenSPE)<span class="sc">$</span>imageID <span class="sc">==</span> <span class="st">"24"</span>]</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>pairwiseAssoc <span class="ot">=</span> <span class="fu">getPairwise</span>(kerenSubset, </span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>                            <span class="at">sigma =</span> <span class="cn">NULL</span>, </span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>                            <span class="at">Rs =</span> <span class="dv">100</span>) <span class="sc">|&gt;</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>()</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>pairwiseAssoc[[<span class="st">"Keratin_Tumour__Neutrophils"</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 10.88892</code></pre>
</div>
</div>
<p>When we specify <code>sigma = 20</code> and re-calculate the L-function, it indicates that there is no relationship between <code>Keratin_Tumour</code> and <code>Neutrophils</code>, i.e., there is no major attraction or dispersion, as it now takes into account tissue inhomogeneity.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>pairwiseAssoc <span class="ot">=</span> <span class="fu">getPairwise</span>(kerenSubset, </span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>                            <span class="at">sigma =</span> <span class="dv">20</span>, </span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>                            <span class="at">Rs =</span> <span class="dv">100</span>) <span class="sc">|&gt;</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>()</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>pairwiseAssoc[[<span class="st">"Keratin_Tumour__Neutrophils"</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.9024836</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># obtain colData for image 24</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>cData <span class="ot">=</span> <span class="fu">colData</span>(kerenSPE) <span class="sc">|&gt;</span> <span class="fu">as.data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>          dplyr<span class="sc">::</span><span class="fu">filter</span>(imageID <span class="sc">==</span> <span class="st">"24"</span>)</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="co"># obtain cells present in image 24</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>coords <span class="ot">=</span> <span class="fu">spatialCoords</span>(kerenSPE) <span class="sc">|&gt;</span> <span class="fu">as.data.frame</span>()</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>coords<span class="sc">$</span>cellID <span class="ot">=</span> <span class="fu">rownames</span>(coords)</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>coords <span class="ot">=</span> coords <span class="sc">|&gt;</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(cellID <span class="sc">%in%</span> cData<span class="sc">$</span>CellID)</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>cData<span class="sc">$</span>X <span class="ot">=</span> coords<span class="sc">$</span>x</span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>cData<span class="sc">$</span>Y <span class="ot">=</span> coords<span class="sc">$</span>y</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>cData <span class="ot">=</span> cData <span class="sc">|&gt;</span> </span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">cellTypeNew =</span> <span class="fu">ifelse</span>(cellType <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Keratin_Tumour"</span>, <span class="st">"Neutrophils"</span>), </span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a>                                     cellType, <span class="st">"Other"</span>))</span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a> pal <span class="ot">=</span> <span class="fu">setNames</span>(<span class="fu">c</span>(<span class="st">"#d6b11c"</span>, <span class="st">"#850f07"</span>), </span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a>                <span class="fu">c</span>(<span class="st">"Keratin_Tumour"</span>, <span class="st">"Neutrophils"</span>))</span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-20"><a href="#cb27-20" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb27-21"><a href="#cb27-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stat_density_2d</span>(<span class="at">data =</span> cData, <span class="fu">aes</span>(<span class="at">x =</span> X, <span class="at">y =</span> Y, <span class="at">fill =</span> <span class="fu">after_stat</span>(density)), </span>
<span id="cb27-22"><a href="#cb27-22" aria-hidden="true" tabindex="-1"></a>                    <span class="at">geom =</span> <span class="st">"raster"</span>, </span>
<span id="cb27-23"><a href="#cb27-23" aria-hidden="true" tabindex="-1"></a>                    <span class="at">contour =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb27-24"><a href="#cb27-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">data =</span> cData <span class="sc">|&gt;</span> <span class="fu">filter</span>(cellType <span class="sc">!=</span> <span class="st">"Other"</span>),</span>
<span id="cb27-25"><a href="#cb27-25" aria-hidden="true" tabindex="-1"></a>               <span class="fu">aes</span>(<span class="at">x =</span> X, <span class="at">y =</span> Y, <span class="at">colour =</span> cellTypeNew), <span class="at">size =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb27-26"><a href="#cb27-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> pal) <span class="sc">+</span></span>
<span id="cb27-27"><a href="#cb27-27" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_distiller</span>(<span class="at">palette =</span> <span class="st">"Blues"</span>, <span class="at">direction =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb27-28"><a href="#cb27-28" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb27-29"><a href="#cb27-29" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"image ID: 24"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="04-cell_localisation_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Plotting image 24 shows that the supposed co-localisation occurs due to the dense cluster of cells near the top of the image.</p>
</section>
<section id="mixed-effects-modelling" class="level3" data-number="4.1.7">
<h3 data-number="4.1.7" class="anchored" data-anchor-id="mixed-effects-modelling"><span class="header-section-number">4.1.7</span> Mixed effects modelling</h3>
<p><code>spicyR</code> supports mixed effects modelling when multiple images are obtained for each subject. In this case, <code>subject</code> is treated as a random effect and <code>condition</code> is treated as a fixed effect. To perform mixed effects modelling, we can specify the <code>subject</code> parameter in the <code>spicy</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>spicyMixedTest <span class="ot">&lt;-</span> <span class="fu">spicy</span>(</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  diabetesData,</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">condition =</span> <span class="st">"stage"</span>,</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">subject =</span> <span class="st">"case"</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="kontextual-context-aware-cell-localisation" class="level2" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="kontextual-context-aware-cell-localisation"><span class="header-section-number">4.2</span> Kontextual: Context aware cell localisation</h2>
<p><code>Kontextual</code> is a method for performing inference on cell localisation which explicitly defines the contexts in which spatial relationships between cells can be identified and interpreted. These contexts may represent landmarks, spatial domains, or groups of functionally similar cells which are consistent across regions. By modelling spatial relationships between cells relative to these contexts, <code>Kontextual</code> produces robust spatial quantifications that are not confounded by biases such as the choice of region to image and the tissue structure present in the images.</p>
<p>In this example we demonstrate how cell type hierarchies can be used as a means to derive appropriate “contexts” for the evaluation of cell localisation. We then demonstrate the types of conclusions which <code>Kontextual</code> enables.</p>
<section id="using-cell-type-hierarchies-to-define-a-context" class="level3" data-number="4.2.1">
<h3 data-number="4.2.1" class="anchored" data-anchor-id="using-cell-type-hierarchies-to-define-a-context"><span class="header-section-number">4.2.1</span> Using cell type hierarchies to define a “context”</h3>
<p>A cell type hierarchy may be used to define the “context” in which cell type relationships are evaluated within. A cell type hierarchy defines how cell types are functionally related to one another. The bottom of the hierarchy represents homogeneous populations of a cell type (child), the cell populations at the nodes of the hierarchy represent broader parent populations with shared generalised function. For example CD4 T cells may be considered a child population to the Immune parent population.</p>
<p>There are two ways to define the cell type hierarchy. First, they can be defined based on our biological understanding of the cell types. We can represent this by creating a named list containing the names of each parent and the associated vector of child cell types.</p>
<p><em>Note:</em> The <code>all</code> vector must be created to include cell types which do not have a parent e.g.&nbsp;the <em>undefined</em> cell type in this data set.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Examine all cell types in image</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(kerenSPE<span class="sc">$</span>cellType)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "Keratin_Tumour" "dn_T_CD3"       "B_cell"         "CD4_T_cell"    
 [5] "DC_or_Mono"     "Unidentified"   "Macrophages"    "CD8_T_cell"    
 [9] "Other_Immune"   "Endothelial"    "Mono_or_Neu"    "Mesenchymal"   
[13] "Neutrophils"    "NK"             "Tumour"         "DC"            
[17] "Tregs"         </code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Named list of parents and their child cell types</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>biologicalHierarchy <span class="ot">=</span> <span class="fu">list</span>(</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">"tumour"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"Keratin_Tumour"</span>, <span class="st">"Tumour"</span>),</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">"tcells"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"dn_T_CD3"</span>, <span class="st">"CD4_T_cell"</span>, <span class="st">"CD8_T_cell"</span>, <span class="st">"Tregs"</span>),</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>  <span class="st">"myeloid"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"DC_or_Mono"</span>, <span class="st">"DC"</span>, <span class="st">"Mono_or_Neu"</span>, <span class="st">"Macrophages"</span>, <span class="st">"Neutrophils"</span>),</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>  <span class="st">"tissue"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"Endothelial"</span>, <span class="st">"Mesenchymal"</span>)</span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Adding more broader immune parent populationse</span></span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a>biologicalHierarchy<span class="sc">$</span>immune <span class="ot">=</span> <span class="fu">c</span>(biologicalHierarchy<span class="sc">$</span>bcells,</span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a>                               biologicalHierarchy<span class="sc">$</span>tcells,</span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a>                               biologicalHierarchy<span class="sc">$</span>myeloid,</span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"NK"</span>, <span class="st">"Other_Immune"</span>, <span class="st">"B_cell"</span>)</span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-15"><a href="#cb31-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-16"><a href="#cb31-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Creating a vector for all cellTypes</span></span>
<span id="cb31-17"><a href="#cb31-17" aria-hidden="true" tabindex="-1"></a>all <span class="ot">&lt;-</span> <span class="fu">unique</span>(kerenSPE<span class="sc">$</span>cellType)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Alternatively, you can use the <code>treeKor</code> bioconductor package <a href="http://www.bioconductor.org/packages/release/bioc/html/treekoR.html">treekoR</a> to define these hierarchies in a data driven way.</p>
<p><em>Note:</em> These parent populations may not be accurate as we are using a small subset of the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate hierarchy using treekoR</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>kerenTree <span class="ot">&lt;-</span> treekoR<span class="sc">::</span><span class="fu">getClusterTree</span>(<span class="fu">t</span>(<span class="fu">assay</span>(kerenSPE, <span class="st">"intensities"</span>)),</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>                            kerenSPE<span class="sc">$</span>cellType,</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>                            <span class="at">hierarchy_method=</span><span class="st">"hopach"</span>,</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>                            <span class="at">hopach_K =</span> <span class="dv">1</span>)</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert treekoR result to a name list of parents and children.</span></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>treekorParents <span class="ot">=</span> <span class="fu">getParentPhylo</span>(kerenTree)</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>treekorParents</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$parent_1
 [1] "Keratin_Tumour" "DC_or_Mono"     "Unidentified"   "Macrophages"   
 [5] "Endothelial"    "Mono_or_Neu"    "Mesenchymal"    "Neutrophils"   
 [9] "Tumour"         "DC"            

$parent_2
[1] "dn_T_CD3"     "B_cell"       "CD4_T_cell"   "CD8_T_cell"   "Other_Immune"
[6] "NK"           "Tregs"       </code></pre>
</div>
</div>
</section>
<section id="application-on-triple-negative-breast-cancer-image." class="level3" data-number="4.2.2">
<h3 data-number="4.2.2" class="anchored" data-anchor-id="application-on-triple-negative-breast-cancer-image."><span class="header-section-number">4.2.2</span> Application on triple negative breast cancer image.</h3>
<p>Here we examine an image highlighted in the Keren et al.&nbsp;2018 manuscript where accounting for context information enables new conclusions. In image 6 of the Keren et al.&nbsp;dataset, we can see that <em>p53+ tumour cells</em> and <em>immune cells</em> are dispersed i.e.&nbsp;these two cell types are not mixing. However we can also see that <em>p53+ tumour cells</em> appear much more localised to <em>immune cells</em> relative to the tumour context (<em>tumour cells</em> and <em>p53+ tumour cells</em>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Lets define a new cell type vector</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>kerenSPE<span class="sc">$</span>cellTypeNew <span class="ot">&lt;-</span> kerenSPE<span class="sc">$</span>cellType</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Select for all cells that express higher than baseline level of p53</span></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>p53Pos <span class="ot">&lt;-</span> <span class="fu">assay</span>(kerenSPE)[<span class="st">"p53"</span>, ] <span class="sc">&gt;</span> <span class="sc">-</span><span class="fl">0.300460</span></span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Find p53+ tumour cells</span></span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>kerenSPE<span class="sc">$</span>cellTypeNew[kerenSPE<span class="sc">$</span>cellType <span class="sc">%in%</span> biologicalHierarchy<span class="sc">$</span>tumour] <span class="ot">&lt;-</span> <span class="st">"Tumour"</span></span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>kerenSPE<span class="sc">$</span>cellTypeNew[p53Pos <span class="sc">&amp;</span> kerenSPE<span class="sc">$</span>cellType <span class="sc">%in%</span> biologicalHierarchy<span class="sc">$</span>tumour] <span class="ot">&lt;-</span> <span class="st">"p53_Tumour"</span></span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Group all immune cells under the name "Immune"</span></span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a>kerenSPE<span class="sc">$</span>cellTypeNew[kerenSPE<span class="sc">$</span>cellType <span class="sc">%in%</span> biologicalHierarchy<span class="sc">$</span>immune] <span class="ot">&lt;-</span> <span class="st">"Immune"</span></span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a>kerenSPE<span class="sc">$</span>x <span class="ot">&lt;-</span> <span class="fu">spatialCoords</span>(kerenSPE)[,<span class="st">"x"</span>]</span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a>kerenSPE<span class="sc">$</span>y <span class="ot">&lt;-</span> <span class="fu">spatialCoords</span>(kerenSPE)[,<span class="st">"y"</span>]</span>
<span id="cb34-16"><a href="#cb34-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-17"><a href="#cb34-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot image 6</span></span>
<span id="cb34-18"><a href="#cb34-18" aria-hidden="true" tabindex="-1"></a>kerenSPE <span class="sc">|&gt;</span></span>
<span id="cb34-19"><a href="#cb34-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colData</span>() <span class="sc">|&gt;</span></span>
<span id="cb34-20"><a href="#cb34-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">|&gt;</span></span>
<span id="cb34-21"><a href="#cb34-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(imageID <span class="sc">==</span> <span class="st">"6"</span>) <span class="sc">|&gt;</span></span>
<span id="cb34-22"><a href="#cb34-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cellTypeNew <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Immune"</span>, <span class="st">"Tumour"</span>, <span class="st">"p53_Tumour"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb34-23"><a href="#cb34-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(cellTypeNew) <span class="sc">|&gt;</span></span>
<span id="cb34-24"><a href="#cb34-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> x<span class="fl">.1</span>, <span class="at">y =</span> y, <span class="at">color =</span> cellTypeNew)) <span class="sc">+</span></span>
<span id="cb34-25"><a href="#cb34-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb34-26"><a href="#cb34-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"Immune"</span> <span class="ot">=</span> <span class="st">"#505050"</span>, <span class="st">"p53_Tumour"</span> <span class="ot">=</span> <span class="st">"#64BC46"</span>, <span class="st">"Tumour"</span> <span class="ot">=</span> <span class="st">"#D6D6D6"</span>)) <span class="sc">+</span></span>
<span id="cb34-27"><a href="#cb34-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">colour =</span> <span class="fu">guide_legend</span>(<span class="at">title =</span> <span class="st">"Cell types"</span>, <span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">3</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="04-cell_localisation_files/figure-html/image6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><code>Kontextual</code> accepts a <code>SingleCellExperiment</code> object, a single image, or list of images from a <code>SingleCellExperiment</code> object, which gets passed into the <code>cells</code> argument. The two cell types which will be evaluated are specified in the <code>to</code> and <code>from</code> arguments. A parent population must also be specified in the <code>parent</code> argument, note the parent cell population must include the <code>to</code> cell type. The argument <code>r</code> will specify the radius which the cell relationship will be evaluated on. <code>Kontextual</code> supports parallel processing, the number of cores can be specified using the <code>cores</code> argument. <code>Kontextual</code> can take a single value or multiple values for each argument and will test all combinations of the arguments specified.</p>
<p>We can calculate these relationships across all images for a single radius (r = 100). Small radii will examine local spatial relationships, whereas larger radii will examine global spatial relationships.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>p53_Kontextual <span class="ot">&lt;-</span> <span class="fu">Kontextual</span>(</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">cells =</span> kerenSPE,</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">r =</span> <span class="dv">100</span>,</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">from =</span> <span class="st">"Immune"</span>,</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">to =</span> <span class="st">"p53_Tumour"</span>,</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">parent =</span> <span class="fu">c</span>(<span class="st">"p53_Tumour"</span>, <span class="st">"Tumour"</span>),</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">cellType =</span> <span class="st">"cellTypeNew"</span></span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>p53_Kontextual</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   imageID               test   original  kontextual   r inhomL
1        1 Immune__p53_Tumour -16.212016  -1.6815952 100  FALSE
2       10 Immune__p53_Tumour -14.715356  -1.7937407 100  FALSE
3       11 Immune__p53_Tumour -11.696597  -7.4615661 100  FALSE
4       12 Immune__p53_Tumour  -9.777271  -2.6287005 100  FALSE
5       13 Immune__p53_Tumour -15.613023  -3.9937364 100  FALSE
6       14 Immune__p53_Tumour -14.671281  -4.2879138 100  FALSE
7       15 Immune__p53_Tumour   8.369183  10.6710168 100  FALSE
8       16 Immune__p53_Tumour -41.081088 -20.9688333 100  FALSE
9       17 Immune__p53_Tumour  -6.331105   5.0017104 100  FALSE
10      18 Immune__p53_Tumour  -1.953366   0.5795853 100  FALSE
11      19 Immune__p53_Tumour -27.834450 -18.7433000 100  FALSE
12       2 Immune__p53_Tumour  -4.989150  -0.5330373 100  FALSE
13      20 Immune__p53_Tumour -20.580091  -9.2542544 100  FALSE
14      21 Immune__p53_Tumour -14.300802  -7.1425133 100  FALSE
15      22 Immune__p53_Tumour -13.673007 -12.9663547 100  FALSE
16      23 Immune__p53_Tumour  15.803493  37.3584378 100  FALSE
17      24 Immune__p53_Tumour -30.319961 -31.8146274 100  FALSE
18      25 Immune__p53_Tumour   6.262995   6.9429103 100  FALSE
19      26 Immune__p53_Tumour -38.190137 -24.6000029 100  FALSE
20      27 Immune__p53_Tumour  -2.373587   4.4044397 100  FALSE
21      28 Immune__p53_Tumour -70.058615 -33.4395839 100  FALSE
22      29 Immune__p53_Tumour -20.728463  -7.0172785 100  FALSE
23       3 Immune__p53_Tumour   1.719549  44.5060581 100  FALSE
24      31 Immune__p53_Tumour -12.306957  -5.2820792 100  FALSE
25      32 Immune__p53_Tumour -18.174569 -10.8972277 100  FALSE
26      33 Immune__p53_Tumour -19.750457 -19.6246151 100  FALSE
27      34 Immune__p53_Tumour -49.004947 -35.1320255 100  FALSE
28      35 Immune__p53_Tumour -75.980619 -66.2395276 100  FALSE
29      36 Immune__p53_Tumour -18.853398 -21.4398044 100  FALSE
30      37 Immune__p53_Tumour -43.624905 -27.7162991 100  FALSE
31      38 Immune__p53_Tumour -12.544687  -3.0415484 100  FALSE
32      39 Immune__p53_Tumour -19.293290  -4.5192485 100  FALSE
33       4 Immune__p53_Tumour         NA          NA 100  FALSE
34      40 Immune__p53_Tumour -37.744261 -27.9604962 100  FALSE
35      41 Immune__p53_Tumour -33.776940 -22.6113096 100  FALSE
36       5 Immune__p53_Tumour         NA          NA 100  FALSE
37       6 Immune__p53_Tumour -24.897348  -1.2724241 100  FALSE
38       7 Immune__p53_Tumour -13.068307   0.6361875 100  FALSE
39       8 Immune__p53_Tumour         NA          NA 100  FALSE
40       9 Immune__p53_Tumour -31.857501   1.0261067 100  FALSE</code></pre>
</div>
</div>
<p>The <code>kontextCurve</code> function plots the L-function value and Kontextual values over a range of radii. If the points lie above the red line (expected pattern) then localisation is indicated for that radius, if the points lie below the red line then dispersion is indicated.</p>
<p>As seen in the following plot the L-function produces negative values over a range of radii, indicating that <em>p53+ tumour cells</em> and <em>immune cells</em> are dispersed from one another. However by taking into account the tumour context, <code>Kontextual</code> shows positive values over some radii, indicating localisation between <em>p53+ tumour cells</em> and <em>immune cells</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>curves <span class="ot">&lt;-</span> <span class="fu">kontextCurve</span>(</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">cells =</span> kerenSPE,</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">from =</span> <span class="st">"Immune"</span>,</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">to =</span> <span class="st">"p53_Tumour"</span>,</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">parent =</span> <span class="fu">c</span>(<span class="st">"p53_Tumour"</span>, <span class="st">"Tumour"</span>),</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">rs =</span> <span class="fu">seq</span>(<span class="dv">50</span>, <span class="dv">510</span>, <span class="dv">50</span>),</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">image =</span> <span class="st">"6"</span>,</span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">cellType =</span> <span class="st">"cellTypeNew"</span>,</span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">cores =</span> nCores</span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a><span class="fu">kontextPlot</span>(curves)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="04-cell_localisation_files/figure-html/kontextCurve-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Alternatively all pairwise cell relationships and their corresponding parent in the dataset can be tested. A data frame with all pairwise combinations can be creating using the <code>parentCombinations</code> function. This function takes in a vector of all the cells, as well as the named list of parents and children created earlier in the <code>parentList</code> argument. As shown below the output is a data frame specifying the <code>to</code>, <code>from</code>, and <code>parent</code> arguments for <code>Kontextual</code>.</p>
<p><em>Note:</em> the output of <code>getPhyloParent</code> may also be using the in the <code>parentList</code> argument, for example if you wanted to use the treekoR defined hierarchy instead.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get all relationships between cell types and their parents</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>parentDf <span class="ot">&lt;-</span> <span class="fu">parentCombinations</span>(</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">all =</span> all,</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">parentList =</span> biologicalHierarchy</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="calculating-all-pairwise-relationships" class="level3" data-number="4.2.3">
<h3 data-number="4.2.3" class="anchored" data-anchor-id="calculating-all-pairwise-relationships"><span class="header-section-number">4.2.3</span> Calculating all pairwise relationships</h3>
<p>Rather than specifying <code>to</code>, <code>from</code>, and <code>parent</code> in <code>Kontextual</code>, the output from <code>parentCombinations</code> can be inputed into <code>Kontextual</code> using the <code>parentDf</code> argument, to examine all pairwise relationships in the dataset. This chunk will take a significant amount of time to run, for demonstration the results have been saved and are loaded in.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Running Kontextual on all relationships across all images.</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>kerenKontextual <span class="ot">&lt;-</span> <span class="fu">Kontextual</span>(</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">cells =</span> kerenSPE,</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">parentDf =</span> parentDf,</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">r =</span> <span class="dv">100</span>,</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">cores =</span> nCores</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For every pairwise relationship (named accordingly: <code>from__to__parent</code>) <code>Kontextual</code> outputs the L-function values (original) and the Kontextual value. The relationships where the L-function and Kontextual disagree (e.g.&nbsp;one metric is positive and the other is negative) represent relationships where adding context information results in different conclusions on the spatial relationship between the two cell types.</p>
</section>
<section id="associating-the-relationships-with-survival-outcomes." class="level3" data-number="4.2.4">
<h3 data-number="4.2.4" class="anchored" data-anchor-id="associating-the-relationships-with-survival-outcomes."><span class="header-section-number">4.2.4</span> Associating the relationships with survival outcomes.</h3>
<p>To examine whether the features obtained from <code>Statial</code> are associated with patient outcomes or groupings, we can use the <code>spicy</code> function from the <code>spicyR</code> package.</p>
<p>In addition to this, the Kontextual results must be converted from a <code>data.frame</code> to a wide <code>matrix</code>, this can be done using <code>prepMatrix</code>. Note, to extract the original L-function values, specify <code>column = "original"</code> in <code>prepMatrix</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Converting Kontextual result into data matrix</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>kontextMat <span class="ot">&lt;-</span> <span class="fu">prepMatrix</span>(kerenKontextual)</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Ensuring rownames of kontextMat match up with the image IDs of the SCE object</span></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>kontextMat <span class="ot">&lt;-</span> kontextMat[kerenSPE<span class="sc">$</span>imageID <span class="sc">|&gt;</span> <span class="fu">unique</span>(), ]</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Replace NAs with 0</span></span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>kontextMat[<span class="fu">is.na</span>(kontextMat)] <span class="ot">&lt;-</span> <span class="dv">0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Finally, both the <code>SingleCellExperiment</code> object and the Kontextual matrix are passed into the <code>spicy</code> function, with <code>condition = "survival"</code>. The resulting coefficients and p values can be obtained by accessing the <code>survivalResults</code> name.</p>
<p><em>Note:</em> You can specify additional covariates and include a subject id for mixed effects survival modelling, see \code{ for more information.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Running survival analysis</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>survivalResults <span class="ot">=</span> <span class="fu">spicy</span>(<span class="at">cells =</span> kerenSPE,</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">alternateResult =</span> kontextMat,</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">condition =</span> <span class="st">"survival"</span>,</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>                        <span class="at">weights =</span> <span class="cn">TRUE</span>)</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(survivalResults<span class="sc">$</span>survivalResults, <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 4
   test                                    coef se.coef     p.value
   &lt;chr&gt;                                  &lt;dbl&gt;   &lt;dbl&gt;       &lt;dbl&gt;
 1 dn_T_CD3__Tregs__immune               0.0187 0.00716 0.000000572
 2 Other_Immune__Tregs__immune           0.0255 0.00885 0.000000727
 3 Tregs__dn_T_CD3__immune               0.0189 0.00715 0.00000169 
 4 Neutrophils__CD8_T_cell__immune       0.0178 0.00651 0.00000172 
 5 DC__dn_T_CD3__immune                 -0.0195 0.00661 0.00000710 
 6 DC__dn_T_CD3__tcells                 -0.0171 0.00758 0.0000167  
 7 Other_Immune__Keratin_Tumour__tumour  0.137  0.0406  0.0000212  
 8 CD4_T_cell__Tregs__immune             0.0216 0.00901 0.0000306  
 9 Other_Immune__DC__myeloid            -0.0181 0.00664 0.0000492  
10 Other_Immune__DC__immune             -0.0186 0.00669 0.0000656  </code></pre>
</div>
</div>
<p>The survival results can also be visualised using the <code>signifPlot</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">signifPlot</span>(survivalResults)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="04-cell_localisation_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>As we can see from the results <code>DC__NK__immune</code> is the one of the most significant pairwise relationship which contributes to patient survival. That is the relationship between dendritic cells and natural killer cells, relative to the parent population of immune cells. We can see that there is a negative coefficient associated with this relationship, which tells us an increase in localisation of these cell types relative to immune cells leads to better survival outcomes for patients.</p>
<p>The association between <code>DC__NK__immune</code> and survival can also be visualised on a Kaplan-Meier curve. First, we extract survival data from the <code>SingleCellExperiment</code> object and create a survival vector.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extracting survival data</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>survData <span class="ot">&lt;-</span> kerenSPE <span class="sc">|&gt;</span></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colData</span>() <span class="sc">|&gt;</span></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">|&gt;</span></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(imageID, survival) <span class="sc">|&gt;</span></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unique</span>()</span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Creating survival vector</span></span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>kerenSurv <span class="ot">&lt;-</span> survData<span class="sc">$</span>survival</span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(kerenSurv) <span class="ot">&lt;-</span> survData<span class="sc">$</span>imageID</span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-12"><a href="#cb44-12" aria-hidden="true" tabindex="-1"></a>kerenSurv</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    1     2     3     4     5     6     7     8     9    10    11    12    13 
 2612   745 3130+ 2523+ 1683+ 2275+   584   946 3767+ 3822+ 3774+ 4353+  1072 
   14    15    16    17    18    19    20    21    22    23    24    25    26 
4145+  1754   530  2842 5063+ 3725+ 4761+   635   NA?    91   194 4785+ 4430+ 
   27    28    29    31    32    33    34    35    36    37    38    39    40 
 3658 3767+  1319  1009 1568+ 1738+ 2832+ 2759+ 3063+ 2853+   NA? 2096+  3573 
   41 
3355+ </code></pre>
</div>
</div>
<p>Next, we extract the Kontextual values of this relationship across all images. We then determine if dendritic and natural killer cells are relatively attracted or avoiding in each image by comparing the Kontextual value in each image to the median Kontextual value.</p>
<p>Finally, we plot a Kaplan-Meier curve using the <code>ggsurvfit</code> package. As shown below, when dendritic and natural killer cells are more localised to one another relative to the immune cell population, patients tend to have better survival outcomes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Selecting most significant relationship</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>survRelationship <span class="ot">&lt;-</span> kontextMat[[<span class="st">"DC__NK__immune"</span>]]</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>survRelationship <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(survRelationship <span class="sc">&gt;</span> <span class="fu">median</span>(survRelationship), <span class="st">"Localised"</span>, <span class="st">"Dispersed"</span>)</span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Plotting Kaplan-Meier curve</span></span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a><span class="fu">survfit2</span>(kerenSurv <span class="sc">~</span> survRelationship) <span class="sc">|&gt;</span></span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggsurvfit</span>() <span class="sc">+</span></span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"DC__NK__immune"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="04-cell_localisation_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="480"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="sessioninfo" class="level2" data-number="4.3">
<h2 data-number="4.3" class="anchored" data-anchor-id="sessioninfo"><span class="header-section-number">4.3</span> sessionInfo</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Debian GNU/Linux 12 (bookworm)

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.21.so;  LAPACK version 3.11.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

time zone: Australia/Sydney
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] ggsurvfit_1.1.0             treekoR_1.14.0             
 [3] tibble_3.2.1                survival_3.7-0             
 [5] dplyr_1.1.4                 imcRtools_1.12.0           
 [7] SpatialDatasets_1.4.0       ExperimentHub_2.14.0       
 [9] AnnotationHub_3.14.0        BiocFileCache_2.14.0       
[11] dbplyr_2.5.0                SpatialExperiment_1.16.0   
[13] SingleCellExperiment_1.28.1 SummarizedExperiment_1.36.0
[15] Biobase_2.66.0              GenomicRanges_1.58.0       
[17] GenomeInfoDb_1.42.0         IRanges_2.40.0             
[19] S4Vectors_0.44.0            BiocGenerics_0.52.0        
[21] MatrixGenerics_1.18.0       matrixStats_1.4.1          
[23] ggplot2_3.5.1               Statial_1.8.0              
[25] spicyR_1.18.0              

loaded via a namespace (and not attached):
  [1] vroom_1.6.5                 tiff_0.1-12                
  [3] goftest_1.2-3               DT_0.33                    
  [5] Biostrings_2.74.0           HDF5Array_1.34.0           
  [7] TH.data_1.1-2               vctrs_0.6.5                
  [9] spatstat.random_3.3-2       digest_0.6.37              
 [11] png_0.1-8                   shape_1.4.6.1              
 [13] proxy_0.4-27                ggrepel_0.9.6              
 [15] deldir_2.0-4                magick_2.8.5               
 [17] MASS_7.3-61                 reshape2_1.4.4             
 [19] httpuv_1.6.15               foreach_1.5.2              
 [21] withr_3.0.2                 ggfun_0.1.7                
 [23] xfun_0.49                   ggpubr_0.6.0               
 [25] memoise_2.0.1               RTriangle_1.6-0.14         
 [27] cytomapper_1.18.0           ggbeeswarm_0.7.2           
 [29] RProtoBufLib_2.18.0         systemfonts_1.1.0          
 [31] tidytree_0.4.6              zoo_1.8-12                 
 [33] GlobalOptions_0.1.2         Formula_1.2-5              
 [35] KEGGREST_1.46.0             promises_1.3.0             
 [37] httr_1.4.7                  rstatix_0.7.2              
 [39] rhdf5filters_1.18.0         rhdf5_2.50.0               
 [41] rstudioapi_0.17.1           UCSC.utils_1.2.0           
 [43] units_0.8-5                 generics_0.1.3             
 [45] concaveman_1.1.0            curl_6.0.0                 
 [47] zlibbioc_1.52.0             ggraph_2.2.1               
 [49] polyclip_1.10-7             GenomeInfoDbData_1.2.13    
 [51] SparseArray_1.6.0           fftwtools_0.9-11           
 [53] xtable_1.8-4                stringr_1.5.1              
 [55] doParallel_1.0.17           evaluate_1.0.1             
 [57] S4Arrays_1.6.0              hms_1.1.3                  
 [59] colorspace_2.1-1            filelock_1.0.3             
 [61] spatstat.data_3.1-2         magrittr_2.0.3             
 [63] readr_2.1.5                 later_1.3.2                
 [65] viridis_0.6.5               ggtree_3.14.0              
 [67] lattice_0.22-6              spatstat.geom_3.3-3        
 [69] XML_3.99-0.17               scuttle_1.16.0             
 [71] ggupset_0.4.0               class_7.3-22               
 [73] svgPanZoom_0.3.4            pillar_1.9.0               
 [75] nlme_3.1-165                iterators_1.0.14           
 [77] EBImage_4.48.0              compiler_4.4.1             
 [79] beachmat_2.22.0             stringi_1.8.4              
 [81] sf_1.0-19                   tensor_1.5                 
 [83] minqa_1.2.8                 ClassifyR_3.10.0           
 [85] plyr_1.8.9                  crayon_1.5.3               
 [87] abind_1.4-8                 gridGraphics_0.5-1         
 [89] locfit_1.5-9.10             sp_2.1-4                   
 [91] graphlayouts_1.2.0          bit_4.5.0                  
 [93] terra_1.7-83                sandwich_3.1-1             
 [95] codetools_0.2-20            multcomp_1.4-26            
 [97] e1071_1.7-16                GetoptLong_1.0.5           
 [99] plotly_4.10.4               mime_0.12                  
[101] MultiAssayExperiment_1.32.0 splines_4.4.1              
[103] circlize_0.4.16             Rcpp_1.0.13-1              
[105] knitr_1.49                  blob_1.2.4                 
[107] utf8_1.2.4                  clue_0.3-65                
[109] BiocVersion_3.20.0          lme4_1.1-35.5              
[111] fs_1.6.5                    nnls_1.6                   
[113] ggplotify_0.1.2             ggsignif_0.6.4             
[115] Matrix_1.7-0                scam_1.2-17                
[117] statmod_1.5.0               tzdb_0.4.0                 
[119] svglite_2.1.3               tweenr_2.0.3               
[121] pkgconfig_2.0.3             pheatmap_1.0.12            
[123] tools_4.4.1                 cachem_1.1.0               
[125] RSQLite_2.3.7               viridisLite_0.4.2          
[127] DBI_1.2.3                   numDeriv_2016.8-1.1        
[129] fastmap_1.2.0               rmarkdown_2.29             
[131] scales_1.3.0                grid_4.4.1                 
[133] shinydashboard_0.7.2        broom_1.0.7                
[135] patchwork_1.3.0             BiocManager_1.30.25        
[137] carData_3.0-5               farver_2.1.2               
[139] tidygraph_1.3.1             mgcv_1.9-1                 
[141] yaml_2.3.10                 ggthemes_5.1.0             
[143] cli_3.6.3                   purrr_1.0.2                
[145] hopach_2.66.0               lifecycle_1.0.4            
[147] mvtnorm_1.3-2               backports_1.5.0            
[149] BiocParallel_1.40.0         cytolib_2.18.0             
[151] gtable_0.3.6                rjson_0.2.23               
[153] parallel_4.4.1              ape_5.8                    
[155] limma_3.62.1                jsonlite_1.8.9             
[157] edgeR_4.4.0                 bitops_1.0-9               
[159] bit64_4.5.2                 Rtsne_0.17                 
[161] FlowSOM_2.14.0              yulab.utils_0.1.8          
[163] spatstat.utils_3.1-1        BiocNeighbors_2.0.0        
[165] ranger_0.17.0               flowCore_2.18.0            
[167] bdsmatrix_1.3-7             spatstat.univar_3.1-1      
[169] lazyeval_0.2.2              shiny_1.9.1                
[171] ConsensusClusterPlus_1.70.0 htmltools_0.5.8.1          
[173] diffcyt_1.26.0              rappdirs_0.3.3             
[175] distances_0.1.11            glue_1.8.0                 
[177] XVector_0.46.0              RCurl_1.98-1.16            
[179] treeio_1.30.0               classInt_0.4-10            
[181] coxme_2.2-22                jpeg_0.1-10                
[183] gridExtra_2.3               boot_1.3-30                
[185] igraph_2.1.1                R6_2.5.1                   
[187] tidyr_1.3.1                 ggiraph_0.8.10             
[189] labeling_0.4.3              ggh4x_0.2.8                
[191] cluster_2.1.6               Rhdf5lib_1.28.0            
[193] aplot_0.2.3                 nloptr_2.1.1               
[195] DelayedArray_0.32.0         tidyselect_1.2.1           
[197] vipor_0.4.7                 ggforce_0.4.2              
[199] raster_3.6-30               car_3.1-3                  
[201] AnnotationDbi_1.68.0        munsell_0.5.1              
[203] KernSmooth_2.23-24          data.table_1.16.2          
[205] htmlwidgets_1.6.4           ComplexHeatmap_2.22.0      
[207] RColorBrewer_1.1-3          rlang_1.1.4                
[209] spatstat.sparse_3.1-0       spatstat.explore_3.3-3     
[211] lmerTest_3.1-3              uuid_1.2-1                 
[213] colorRamps_2.3.4            ggnewscale_0.5.0           
[215] fansi_1.0.6                 beeswarm_0.4.0             </code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./03-cell_annotation.html" class="pagination-link" aria-label="Cell Annotation">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Cell Annotation</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./05-cellular_niches.html" class="pagination-link" aria-label="Spatial domains">
        <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Spatial domains</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>