<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.39">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>6&nbsp; Cell Annotation â€“ Spatial analysis playbook</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./04a-cell_localisation.html" rel="next">
<link href="./02-quality_control.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-e26003cea8cd680ca0c55a263523d882.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-19fe84b14861edf6ae29e3008defacb0.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><link rel="stylesheet" href="styles.css">
</head>
<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="./index.html" class="navbar-brand navbar-brand-logo">
    <img src="./images/USydLogo.svg" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Spatial analysis playbook</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./03-cell_annotation.html">Cell annotation</a></li><li class="breadcrumb-item"><a href="./03-cell_annotation.html"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Cell Annotation</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Overview</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Datasets</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./getting_started.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Getting Started</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Processing</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01-processing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Cell segmentation and pre-processing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02-quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Quality Control</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Cell annotation</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03-cell_annotation.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Cell Annotation</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Cell localisation</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04a-cell_localisation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Cell localisation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04b-cell_localisation_parent.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Cell localisation</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">Spatial domains</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05-cellular_niches.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Spatial domains</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true">
 <span class="menu-text">Changes in marker expression</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06-changes_in_marker_expression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Marker expression</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true">
 <span class="menu-text">Classification</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07-classification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Classification</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" role="navigation" aria-expanded="true">
 <span class="menu-text">Case studies</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-8" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08-case_study1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Case Study: Head and Neck Squamous Cell Carcinoma (Ferguson et al., 2022)</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-9" role="navigation" aria-expanded="true">
 <span class="menu-text">References</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-9" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-9" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./09-references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#clustering-vs-annotation" id="toc-clustering-vs-annotation" class="nav-link active" data-scroll-target="#clustering-vs-annotation"><span class="header-section-number">6.1</span> Clustering vs Annotation</a></li>
  <li>
<a href="#clustering-with-fusesom" id="toc-clustering-with-fusesom" class="nav-link" data-scroll-target="#clustering-with-fusesom"><span class="header-section-number">6.2</span> Clustering with FuseSOM</a>
  <ul class="collapse">
<li><a href="#fusesom-matrix-input" id="toc-fusesom-matrix-input" class="nav-link" data-scroll-target="#fusesom-matrix-input"><span class="header-section-number">6.2.1</span> <code>FuseSOM</code> Matrix Input</a></li>
  <li><a href="#using-fusesom-to-estimate-the-number-of-clusters" id="toc-using-fusesom-to-estimate-the-number-of-clusters" class="nav-link" data-scroll-target="#using-fusesom-to-estimate-the-number-of-clusters"><span class="header-section-number">6.2.2</span> Using <code>FuseSOM</code> to estimate the number of clusters</a></li>
  <li><a href="#fusesom-with-single-cell-epxeriment-object-as-input" id="toc-fusesom-with-single-cell-epxeriment-object-as-input" class="nav-link" data-scroll-target="#fusesom-with-single-cell-epxeriment-object-as-input"><span class="header-section-number">6.2.3</span> <code>FuseSOM</code> with Single Cell Epxeriment object as input</a></li>
  <li><a href="#using-fusesom-to-estimate-the-number-of-clusters-for-singlecellexperiment-objects" id="toc-using-fusesom-to-estimate-the-number-of-clusters-for-singlecellexperiment-objects" class="nav-link" data-scroll-target="#using-fusesom-to-estimate-the-number-of-clusters-for-singlecellexperiment-objects"><span class="header-section-number">6.2.4</span> Using <code>FuseSOM</code> to estimate the number of clusters for <code>SingleCellExperiment</code> objects</a></li>
  </ul>
</li>
  <li><a href="#scclassify-cell-annotation" id="toc-scclassify-cell-annotation" class="nav-link" data-scroll-target="#scclassify-cell-annotation"><span class="header-section-number">6.3</span> scClassify: Cell annotation</a></li>
  <li><a href="#choosing-between-clustering-and-annotation" id="toc-choosing-between-clustering-and-annotation" class="nav-link" data-scroll-target="#choosing-between-clustering-and-annotation"><span class="header-section-number">6.4</span> Choosing between clustering and annotation</a></li>
  <li><a href="#sessioninfo" id="toc-sessioninfo" class="nav-link" data-scroll-target="#sessioninfo"><span class="header-section-number">6.5</span> sessionInfo</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./03-cell_annotation.html">Cell annotation</a></li><li class="breadcrumb-item"><a href="./03-cell_annotation.html"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Cell Annotation</span></a></li></ol></nav><div class="quarto-title">
<h1 class="title">
<span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Cell Annotation</span>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><p>Labeling the identity of your cells is a key step in any spatial processing protocol in order to determine differential cell type compositions and changes which occur in specific cell types during disease. However, the method by which this is done can differ from study to study. Here, we demonstrate the use of two packages: <code>FuseSOM</code> for cell type clustering and <code>scClassify</code> for cell type annotation.</p>
<p>Steps:</p>
<ol type="1">
<li>Clustering vs annotation</li>
<li>Clustering with FuseSOM</li>
<li>Cluster annotation with pheatmap</li>
<li>Cell annotation with scClassify</li>
<li>Selecting a reference dataset with scClassify</li>
</ol>
<section id="clustering-vs-annotation" class="level2" data-number="6.1"><h2 data-number="6.1" class="anchored" data-anchor-id="clustering-vs-annotation">
<span class="header-section-number">6.1</span> Clustering vs Annotation</h2>
<p>Clustering is an unsupervised method of labelling cells. An algorithm identifies clusters of similar cells based on marker expression patterns, and the resulting clusters need to be manually identified based on biological domain knowledge. Cell annotation is a supervised method which requires a separate, reference dataset. The algorithm uses the reference dataset to assign a cell type label to each cell in the dataset. There are advantages and disadvantages to both. We will first demonstrate the use of both <code>FuseSOM</code> and <code>scClassify</code>, and then discuss how to choose between clustering and annotation.</p>
</section><section id="clustering-with-fusesom" class="level2" data-number="6.2"><h2 data-number="6.2" class="anchored" data-anchor-id="clustering-with-fusesom">
<span class="header-section-number">6.2</span> Clustering with FuseSOM</h2>
<p><a href="https://www.bioconductor.org/packages/release/bioc/html/FuseSOM.html">FuseSOM</a> is an unsupervised clustering tool for highly multiplexed in situ imaging cytometry assays. It combines a <code>Self Organiszing Map</code> architecture and a <code>MultiView</code> integration of correlation-based metrics for robustness and high accuracy. It has been streamlined to accept multiple data structures including <code>SingleCellExperiment</code> objects, <code>SpatialExperiment</code> objects, and <code>DataFrames</code>.</p>
<section id="fusesom-matrix-input" class="level3" data-number="6.2.1"><h3 data-number="6.2.1" class="anchored" data-anchor-id="fusesom-matrix-input">
<span class="header-section-number">6.2.1</span> <code>FuseSOM</code> Matrix Input</h3>
<p>To demonstrate the functionality of <code>FuseSOM</code>, we will use the Risom 2022 dataset, which profiles the spatial landscape of ductal carcinoma in situ (DCIS). We will be using the markers used in the original study to perform clustering.</p>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># load FuseSOM</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">FuseSOM</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scClassify</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/lmweber/STexampleData">STexampleData</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: ExperimentHub</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: BiocGenerics</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    Position, rank, rbind, Reduce, rownames, sapply, saveRDS, setdiff,
    table, tapply, union, unique, unsplit, which.max, which.min</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: AnnotationHub</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: BiocFileCache</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: dbplyr</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: SingleCellExperiment</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: SummarizedExperiment</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: MatrixGenerics</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: matrixStats</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'MatrixGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:matrixStats':

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    rowWeightedSds, rowWeightedVars</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: GenomicRanges</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: stats4</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: S4Vectors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'S4Vectors'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:utils':

    findMatches</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    expand.grid, I, unname</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: IRanges</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: GenomeInfoDb</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: Biobase</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Welcome to Bioconductor

    Vignettes contain introductory material; view with
    'browseVignettes()'. To cite Bioconductor, see
    'citation("Biobase")', and for packages 'citation("pkgname")'.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'Biobase'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:MatrixGenerics':

    rowMedians</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:matrixStats':

    anyMissing, rowMedians</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:ExperimentHub':

    cache</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:AnnotationHub':

    cache</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: SpatialExperiment</code></pre>
</div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/yanyachen/MLmetrics">MLmetrics</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'MLmetrics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:base':

    Recall</code></pre>
</div>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/simpleSeg/">simpleSeg</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load in the data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"risom_dat"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># define the markers of interest</span></span>
<span><span class="va">risomMarkers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'CD45'</span>,<span class="st">'SMA'</span>,<span class="st">'CK7'</span>,<span class="st">'CK5'</span>,<span class="st">'VIM'</span>,<span class="st">'CD31'</span>,<span class="st">'PanKRT'</span>,<span class="st">'ECAD'</span>,</span>
<span>                   <span class="st">'Tryptase'</span>,<span class="st">'MPO'</span>,<span class="st">'CD20'</span>,<span class="st">'CD3'</span>,<span class="st">'CD8'</span>,<span class="st">'CD4'</span>,<span class="st">'CD14'</span>,<span class="st">'CD68'</span>,<span class="st">'FAP'</span>,</span>
<span>                   <span class="st">'CD36'</span>,<span class="st">'CD11c'</span>,<span class="st">'HLADRDPDQ'</span>,<span class="st">'P63'</span>,<span class="st">'CD44'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># we will be using the manual_gating_phenotype as the true cell type to gauge </span></span>
<span><span class="co"># performance</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">risom_dat</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">risom_dat</span><span class="op">)</span> <span class="op">==</span> <span class="st">'manual_gating_phenotype'</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'CellType'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now that we have loaded the data and defined the markers of interest, we can run the <code>FuseSOM</code> algorithm using the <code><a href="https://rdrr.io/pkg/FuseSOM/man/runFuseSOM.html">runFuseSOM()</a></code> function. We specify the number of clusters to be 23 based on prior domain knowledge. The output contains the cluster labels as well as the <code>Self Organizing Map</code> model.</p>
<div class="cell">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">risomRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/runFuseSOM.html">runFuseSOM</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">risom_dat</span>, markers <span class="op">=</span> <span class="va">risomMarkers</span>, </span>
<span>                        numClusters <span class="op">=</span> <span class="fl">23</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>You have provided a dataset of class data.frame</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Everything looks good. Now running the FuseSOM algorithm</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Now Generating the Self Organizing Map Grid</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Optimal Grid Size is: 8</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Now Running the Self Organizing Map Model</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Now Clustering the Prototypes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required namespace: fastcluster</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Now Mapping Clusters to the Original Data</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The Prototypes have been Clustered and Mapped Successfully</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The FuseSOM algorithm has completed successfully</code></pre>
</div>
</div>
<p>Lets look at the distribution of the clusters.</p>
<div class="cell">
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># get the distribution of the clusters</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">risomRes</span><span class="op">$</span><span class="va">clusters</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">risomRes</span><span class="op">$</span><span class="va">clusters</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
  cluster_1  cluster_10  cluster_11  cluster_12  cluster_13  cluster_14 
0.323602021 0.035968538 0.005439775 0.021443334 0.061100586 0.026596050 
 cluster_15  cluster_16  cluster_17  cluster_18  cluster_19   cluster_2 
0.020582156 0.032624297 0.024931106 0.076128143 0.015802618 0.014927087 
 cluster_20  cluster_21  cluster_22  cluster_23   cluster_3   cluster_4 
0.049962682 0.009185900 0.051771156 0.066913538 0.004923068 0.014108968 
  cluster_5   cluster_6   cluster_7   cluster_8   cluster_9 
0.040776783 0.064444827 0.020854863 0.010032725 0.007879780 </code></pre>
</div>
</div>
<p>It appears that 32% of cells have been assigned to <code>cluster_1</code>. Next, lets generate a heatmap of the marker expression for each cluster.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">risomHeat</span> <span class="op">&lt;-</span> <span class="fu">FuseSOM</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/markerHeatmap.html">markerHeatmap</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">risom_dat</span>, markers <span class="op">=</span> <span class="va">risomMarkers</span>,</span>
<span>                            clusters <span class="op">=</span> <span class="va">risomRes</span><span class="op">$</span><span class="va">clusters</span>, clusterMarkers <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="03-cell_annotation_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>few comments analysing heatmap?</li>
</ul></section><section id="using-fusesom-to-estimate-the-number-of-clusters" class="level3" data-number="6.2.2"><h3 data-number="6.2.2" class="anchored" data-anchor-id="using-fusesom-to-estimate-the-number-of-clusters">
<span class="header-section-number">6.2.2</span> Using <code>FuseSOM</code> to estimate the number of clusters</h3>
<p>When the number of expected cell typess or clusters is not known beforehand, the <code><a href="https://rdrr.io/pkg/FuseSOM/man/estimateNumCluster.html">estimateNumCluster()</a></code> function can be used to estimate the number of clusters. Two methods have been developed to calculate the number of clusters:</p>
<ol type="1">
<li>Discriminant based method:
<ul>
<li>A method developed in house based on discriminant based maximum clusterability projection pursuit</li>
</ul>
</li>
<li>Distance based methods which includes:
<ul>
<li>The Gap Statistic</li>
<li>The Jump Statistic</li>
<li>The Slope Statistic</li>
<li>The Within Cluster Dissimilarity Statistic</li>
<li>The Silhouette Statistic</li>
</ul>
</li>
</ol>
<p>We run <code><a href="https://rdrr.io/pkg/FuseSOM/man/estimateNumCluster.html">estimateNumCluster()</a></code> and specify <code>method = c("Discriminant", "Distance")</code> to use both approaches.</p>
<div class="cell">
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># lets estimate the number of clusters using all the methods</span></span>
<span><span class="co"># original clustering has 23 clusters so we will set kseq from 2:25</span></span>
<span><span class="co"># we pass it the SOM model generated in the previous step</span></span>
<span><span class="va">risomKest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/estimateNumCluster.html">estimateNumCluster</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">risomRes</span><span class="op">$</span><span class="va">model</span>, kSeq <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">25</span>, </span>
<span>                                  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Discriminant"</span>, <span class="st">"Distance"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Now Computing the Number of Clusters using Discriminant Analysis</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Now Computing The Number Of Clusters Using Distance Analysis</code></pre>
</div>
</div>
<p>We can then use this result to determine the best number of clusters for this dataset based on the different metrics.</p>
<div class="cell">
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># what is the best number of clusters determined by the discriminant method?</span></span>
<span><span class="va">risomKest</span><span class="op">$</span><span class="va">Discriminant</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 10</code></pre>
</div>
</div>
<p>According to the Discriminant method, the optimal number of clusters is 7.</p>
<p>We can use the <code><a href="https://rdrr.io/pkg/FuseSOM/man/optiPlot.html">optiPlot()</a></code> function to generate an elbow plot with the optimal value for the number of clusters for the distance based methods.</p>
<div class="cell">
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># we can plot the results using the optiplot function</span></span>
<span><span class="va">pSlope</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/optiPlot.html">optiPlot</a></span><span class="op">(</span><span class="va">risomKest</span>, method <span class="op">=</span> <span class="st">'slope'</span><span class="op">)</span></span>
<span><span class="va">pSlope</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="03-cell_annotation_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pJump</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/optiPlot.html">optiPlot</a></span><span class="op">(</span><span class="va">risomKest</span>, method <span class="op">=</span> <span class="st">'jump'</span><span class="op">)</span></span>
<span><span class="va">pJump</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="03-cell_annotation_files/figure-html/unnamed-chunk-7-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pWcd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/optiPlot.html">optiPlot</a></span><span class="op">(</span><span class="va">risomKest</span>, method <span class="op">=</span> <span class="st">'wcd'</span><span class="op">)</span></span>
<span><span class="va">pWcd</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="03-cell_annotation_files/figure-html/unnamed-chunk-7-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pGap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/optiPlot.html">optiPlot</a></span><span class="op">(</span><span class="va">risomKest</span>, method <span class="op">=</span> <span class="st">'gap'</span><span class="op">)</span></span>
<span><span class="va">pGap</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="03-cell_annotation_files/figure-html/unnamed-chunk-7-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pSil</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/optiPlot.html">optiPlot</a></span><span class="op">(</span><span class="va">risomKest</span>, method <span class="op">=</span> <span class="st">'silhouette'</span><span class="op">)</span></span>
<span><span class="va">pSil</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="03-cell_annotation_files/figure-html/unnamed-chunk-7-5.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>From the plots, we see that the <code>Jump</code> statistic almost perfectly captures the correct number of clusters. The <code>Gap</code> statistic is a close second with 15 clusters. All the other methods significantly underestimate the number of clusters.</p>
</section><section id="fusesom-with-single-cell-epxeriment-object-as-input" class="level3" data-number="6.2.3"><h3 data-number="6.2.3" class="anchored" data-anchor-id="fusesom-with-single-cell-epxeriment-object-as-input">
<span class="header-section-number">6.2.3</span> <code>FuseSOM</code> with Single Cell Epxeriment object as input</h3>
<p>The <code>FuseSOM</code> algorithm is also equipped to take in a <code>SingleCellExperiment</code> object as input. The results of the pipeline will be written to either the metada or the colData fields.</p>
<p>First, we create a <code>SingleCellExperiment</code> object using the Risom 2022 data.</p>
<div class="cell">
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create an SCE object using Risom 2022 data</span></span>
<span><span class="va">colDat</span> <span class="op">&lt;-</span> <span class="va">risom_dat</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/sets.html">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">risom_dat</span><span class="op">)</span>, <span class="va">risomMarkers</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html">SingleCellExperiment</a></span><span class="op">(</span>assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">risom_dat</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                                 colData <span class="op">=</span> <span class="va">colDat</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SingleCellExperiment 
dim: 23 69672 
metadata(0):
assays(1): counts
rownames(23): CD45 SMA ... CD44 CellType
rowData names(0):
colnames: NULL
colData names(1): X
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
</div>
<p>Next, we pass it to the <code><a href="https://rdrr.io/pkg/FuseSOM/man/runFuseSOM.html">runFuseSOM()</a></code> function. Here, we can provide the assay in which the data is stored (<code>counts</code>) and specify the column to store the clusters in using <code>clusterCol = "clusters"</code>. The <code>Self Organizing Map</code> that is generated will be stored in the metadata field.</p>
<div class="cell">
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">risomRessce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/runFuseSOM.html">runFuseSOM</a></span><span class="op">(</span><span class="va">sce</span>, markers <span class="op">=</span> <span class="va">risomMarkers</span>, clusterCol <span class="op">=</span> <span class="st">"clusters"</span>,</span>
<span>                          assay <span class="op">=</span> <span class="st">'counts'</span>, numClusters <span class="op">=</span> <span class="fl">23</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>You have provided a dataset of class SingleCellExperiment</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Everything looks good. Now running the FuseSOM algorithm</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Now Generating the Self Organizing Map Grid</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Optimal Grid Size is: 8</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Now Running the Self Organizing Map Model</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Now Clustering the Prototypes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Now Mapping Clusters to the Original Data</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The Prototypes have been Clustered and Mapped Successfully</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The FuseSOM algorithm has completed successfully</code></pre>
</div>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">risomRessce</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "X"        "clusters"</code></pre>
</div>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">risomRessce</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "SOM"</code></pre>
</div>
</div>
<p>Notice how the there is now a <code>clusters</code> column in the <code>colData</code> and a SOM field in the metadata.</p>
<p>If necessary, you can run <code><a href="https://rdrr.io/pkg/FuseSOM/man/runFuseSOM.html">runFuseSOM()</a></code> with a new cluster number and specify a new <code>clusterCol</code>. If <code>clusterCol</code> contains a new name, the new clusters will be stored in the new column. Otherwise, it will overwrite the the current <code>clusters</code> column. Running FuseSOM on the same object will overwrite the SOM field in the metadata.</p>
<p>Just like before, we can plot a heatmap of the resulting clusters across all markers.</p>
<div class="cell">
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">risom_dat</span><span class="op">[</span>, <span class="va">risomMarkers</span><span class="op">]</span> <span class="co"># get the original data used</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">risomRessce</span><span class="op">)</span><span class="op">$</span><span class="va">clusters</span> <span class="co"># extract the clusters from the SCE object</span></span>
<span></span>
<span><span class="co"># generate the heatmap</span></span>
<span><span class="va">risomHeatsce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/markerHeatmap.html">markerHeatmap</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">risom_dat</span>, markers <span class="op">=</span> <span class="va">risomMarkers</span>,</span>
<span>                            clusters <span class="op">=</span> <span class="va">clusters</span>, clusterMarkers <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="03-cell_annotation_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="using-fusesom-to-estimate-the-number-of-clusters-for-singlecellexperiment-objects" class="level3" data-number="6.2.4"><h3 data-number="6.2.4" class="anchored" data-anchor-id="using-fusesom-to-estimate-the-number-of-clusters-for-singlecellexperiment-objects">
<span class="header-section-number">6.2.4</span> Using <code>FuseSOM</code> to estimate the number of clusters for <code>SingleCellExperiment</code> objects</h3>
<p>Just like before, we will use <code><a href="https://rdrr.io/pkg/FuseSOM/man/estimateNumCluster.html">estimateNumCluster()</a></code> on our Risom <code>SingleCellExperiment</code> object.</p>
<div class="cell">
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># lets estimate the number of clusters using all the methods</span></span>
<span><span class="co"># original clustering has 23 clusters so we will set kseq from 2:25</span></span>
<span><span class="va">risomRessce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/estimateNumCluster.html">estimateNumCluster</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">risomRessce</span>, kSeq <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">25</span>, </span>
<span>                                  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Discriminant"</span>, <span class="st">"Distance"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>You have provided a dataset of class: SingleCellExperiment</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Now Computing the Number of Clusters using Discriminant Analysis</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Now Computing The Number Of Clusters Using Distance Analysis</code></pre>
</div>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">risomRessce</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "SOM"               "clusterEstimation"</code></pre>
</div>
</div>
<p>The metadata now contains a <code>clusterEstimation</code> field which holds the results from the <code><a href="https://rdrr.io/pkg/FuseSOM/man/estimateNumCluster.html">estimateNumCluster()</a></code> function.</p>
<p>We can assess the results of cluster estimation as below.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># what is the best number of clusters determined by the discriminant method?</span></span>
<span><span class="fu">metadata</span><span class="op">(</span><span class="va">risomRessce</span><span class="op">)</span><span class="op">$</span><span class="va">clusterEstimation</span><span class="op">$</span><span class="va">Discriminant</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 8</code></pre>
</div>
</div>
<p>According to the discrminant method, the optimal number of clusters is 10.</p>
<div class="cell">
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># we can plot the results using the optiplot function</span></span>
<span><span class="va">pSlope</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/optiPlot.html">optiPlot</a></span><span class="op">(</span><span class="va">risomRessce</span>, method <span class="op">=</span> <span class="st">'slope'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>You have provided a dataset of class: SingleCellExperiment</code></pre>
</div>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pSlope</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="03-cell_annotation_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pJump</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/optiPlot.html">optiPlot</a></span><span class="op">(</span><span class="va">risomRessce</span>, method <span class="op">=</span> <span class="st">'jump'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>You have provided a dataset of class: SingleCellExperiment</code></pre>
</div>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pJump</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="03-cell_annotation_files/figure-html/unnamed-chunk-13-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pWcd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/optiPlot.html">optiPlot</a></span><span class="op">(</span><span class="va">risomRessce</span>, method <span class="op">=</span> <span class="st">'wcd'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>You have provided a dataset of class: SingleCellExperiment</code></pre>
</div>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pWcd</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="03-cell_annotation_files/figure-html/unnamed-chunk-13-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pGap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/optiPlot.html">optiPlot</a></span><span class="op">(</span><span class="va">risomRessce</span>, method <span class="op">=</span> <span class="st">'gap'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>You have provided a dataset of class: SingleCellExperiment</code></pre>
</div>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pGap</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="03-cell_annotation_files/figure-html/unnamed-chunk-13-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pSil</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/optiPlot.html">optiPlot</a></span><span class="op">(</span><span class="va">risomRessce</span>, method <span class="op">=</span> <span class="st">'silhouette'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>You have provided a dataset of class: SingleCellExperiment</code></pre>
</div>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pSil</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="03-cell_annotation_files/figure-html/unnamed-chunk-13-5.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Again, we see that the <code>Jump</code> statistic almost perfectly captures the correct number of clusters with 24 clusters. The <code>Gap</code> method is a close second with 15 clusters. All the other methods significantly underestimate the number of clusters.</p>
</section></section><section id="scclassify-cell-annotation" class="level2" data-number="6.3"><h2 data-number="6.3" class="anchored" data-anchor-id="scclassify-cell-annotation">
<span class="header-section-number">6.3</span> scClassify: Cell annotation</h2>
<p><a href="https://new.bioconductor.org/packages/release/bioc/html/scClassify.html">scClassify</a> is a supervised multiscale classification framework for scRNAseq data based on ensemble learning and cell type hierarchies. scClassify first constructs a cell type tree using the reference dataset, and then builds an ensemble weighted kNN classifier at each node. Users can select any combination of six correlation/similarity metrics and five marker selection methods to build the kNN model. This results in up to 30 base classifiers that can be constructed for each node of the cell type tree.</p>
<p>To demonstrate the use of scClassify, we will use the Stickels 2021 dataset, which examines mouse brain tissue from the hippocampus and surrounding regions. It is available through the <code>STexampleData</code> package. We will select samples 21-24 for demonstration purposes.</p>
<div class="cell">
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">stickelsSPE</span> <span class="op">=</span> <span class="fu">STexampleData</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/STexampleData/man/STexampleData.html">SlideSeqV2_mouseHPC</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># remove missing celltype values</span></span>
<span><span class="va">stickelsSPE</span> <span class="op">=</span> <span class="va">stickelsSPE</span><span class="op">[</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">stickelsSPE</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">stickelsSPE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SpatialExperiment 
dim: 23264 15003 
metadata(0):
assays(1): counts
rownames(23264): 0610005C13Rik 0610007P14Rik ... n-R5s40 n-R5s95
rowData names(1): gene_name
colnames(15003): GATGCGACTCCTCG ACAGGGTTTATCGA ... CGTCGGATGTAGGG
  GCGCTCAGTGCTCT
colData names(3): barcode_id sample_id celltype
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):
spatialCoords names(2) : xcoord ycoord
imgData names(0):</code></pre>
</div>
</div>
<p>The dataset contains 17 unique cell types.</p>
<div class="cell">
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">stickelsSPE</span><span class="op">$</span><span class="va">celltype</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "CA3"                   "CA1"                   "Choroid"              
 [4] "Oligodendrocyte"       "Denate"                "Entorihinal"          
 [7] "Ependymal"             "Interneuron"           "Astrocyte"            
[10] "Neuron.Slc17a6"        "Neurogenesis"          "Mural"                
[13] "Endothelial_Stalk"     "Polydendrocyte"        "Endothelial_Tip"      
[16] "Microglia_Macrophages" "Cajal_Retzius"        </code></pre>
</div>
</div>
<p>As scClassify requires log-normalised expression data, we will use <code>normalizeCounts</code> from the <code>scuttle</code> package to transform and normalise our expression data.</p>
<div class="cell">
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># log normalisation</span></span>
<span><span class="va">norm_assay</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/normalizeCounts.html">normalizeCounts</a></span><span class="op">(</span><span class="va">stickelsSPE</span>, transform <span class="op">=</span> <span class="st">"log"</span>, BPPARAM <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">8</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># store normalised expression data in the "norm" slot</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">stickelsSPE</span>, <span class="st">"norm"</span><span class="op">)</span> <span class="op">=</span> <span class="va">norm_assay</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We will first split the dataset into an 80% training or reference dataset and a 20% testing or query dataset.</p>
<div class="cell">
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># split into 80% reference and 20% query</span></span>
<span><span class="va">ref_idx</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">stickelsSPE</span><span class="op">)</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">0.8</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">stickelsSPE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ref_spe</span> <span class="op">=</span> <span class="va">stickelsSPE</span><span class="op">[</span>, <span class="va">ref_idx</span><span class="op">]</span></span>
<span><span class="va">query_spe</span> <span class="op">=</span> <span class="va">stickelsSPE</span><span class="op">[</span>, <span class="op">-</span><span class="va">ref_idx</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># access expression data as matrix</span></span>
<span><span class="va">ref_mat</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">ref_spe</span>, <span class="st">"norm"</span><span class="op">)</span></span>
<span><span class="va">query_mat</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">query_spe</span>, <span class="st">"norm"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>scClassify</code> is the main classification function within scClassify. It requires a reference expression matrix (<code>exprsMat_train</code>) and the corresponding vector of cell type labels as input (<code>cellTypes_train</code>). We then specify our query matrix (<code>exprsMat_test</code>) and the corresponding cell type labels (<code>cellTypes_test</code>) in the form of lists. Below, we have also specified the following customisations for building the model -</p>
<ul>
<li>
<code>tree = HOPACH</code>: use the HOPACH algorithm to build a cell type hierarchy.</li>
<li>
<code>algorithm = WKNN</code>: use the weighted KNN model</li>
<li>
<code>selectFeatures = c("limma")</code>: identify differentially expressed (DE) genes using <code>limma</code>, which will be used as features to build the WKNN model. Other options to select features include differentially variable (DV) genes, differentially distributed (DD), bimodally distributed (BI) and differentially expressed proportions.</li>
<li>
<code>c("pearson", "spearman", "cosine")</code>: use Pearson and Spearman correlation to measure similarity between cells. Other options include Kendallâ€™s correlation, cosine distance, Jaccard distance, Manhattan distance, and weighted rank correlation.</li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">scClassify_res</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scClassify/man/scClassify.html">scClassify</a></span><span class="op">(</span>exprsMat_train <span class="op">=</span> <span class="va">ref_mat</span>,</span>
<span>                            cellTypes_train <span class="op">=</span> <span class="va">ref_spe</span><span class="op">$</span><span class="va">celltype</span>,</span>
<span>                            exprsMat_test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>stickels <span class="op">=</span> <span class="va">query_mat</span><span class="op">)</span>,</span>
<span>                            cellTypes_test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>stickels <span class="op">=</span> <span class="va">query_spe</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span>,</span>
<span>                            tree <span class="op">=</span> <span class="st">"HOPACH"</span>,</span>
<span>                            algorithm <span class="op">=</span> <span class="st">"WKNN"</span>,</span>
<span>                            selectFeatures <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"limma"</span><span class="op">)</span>,</span>
<span>                            similarity <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"pearson"</span>, <span class="st">"spearman"</span><span class="op">)</span>,</span>
<span>                            returnList <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                            parallel <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                            BPPARAM <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>,</span>
<span>                            verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The output is a list containing all model building and testing information. The <code>testRes</code> slot contains the results of testing the model on the query dataset, and the <code>trainRes</code> contains the model built from the reference dataset.</p>
<div class="cell">
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">scClassify_res</span><span class="op">$</span><span class="va">trainRes</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can then examine the cell type tree generated by the reference dataset.</p>
<div class="cell">
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scClassify/man/plotCellTypeTree.html">plotCellTypeTree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/scClassify/man/cellTypeTree.html">cellTypeTree</a></span><span class="op">(</span><span class="va">scClassify_res</span><span class="op">$</span><span class="va">trainRes</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The predictions for each cell can be accessed through the <code>ensembleRes</code> slot in the testing results.</p>
<div class="cell">
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pred_mat</span> <span class="op">=</span> <span class="va">scClassify_res</span><span class="op">$</span><span class="va">testRes</span><span class="op">$</span><span class="va">stickels</span><span class="op">$</span><span class="va">ensembleRes</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">pred_mat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We will evaluate scClassifyâ€™s performance using 4 metrics - accuracy, precision, recall, and F1 Score.</p>
<div class="cell">
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># accuracy</span></span>
<span><span class="fu">MLmetrics</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MLmetrics/man/Accuracy.html">Accuracy</a></span><span class="op">(</span><span class="va">test_spe</span><span class="op">$</span><span class="va">celltype</span>, <span class="va">pred_mat</span><span class="op">$</span><span class="va">cellTypes</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># precision</span></span>
<span><span class="fu">MLmetrics</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MLmetrics/man/Precision.html">Precision</a></span><span class="op">(</span><span class="va">test_spe</span><span class="op">$</span><span class="va">celltype</span>, <span class="va">pred_mat</span><span class="op">$</span><span class="va">cellTypes</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># recall</span></span>
<span><span class="fu">MLmetrics</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MLmetrics/man/Recall.html">Recall</a></span><span class="op">(</span><span class="va">test_spe</span><span class="op">$</span><span class="va">celltype</span>, <span class="va">pred_mat</span><span class="op">$</span><span class="va">cellTypes</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># recall</span></span>
<span><span class="fu">MLmetrics</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MLmetrics/man/F1_Score.html">F1_Score</a></span><span class="op">(</span><span class="va">test_spe</span><span class="op">$</span><span class="va">celltype</span>, <span class="va">pred_mat</span><span class="op">$</span><span class="va">cellTypes</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="choosing-between-clustering-and-annotation" class="level2" data-number="6.4"><h2 data-number="6.4" class="anchored" data-anchor-id="choosing-between-clustering-and-annotation">
<span class="header-section-number">6.4</span> Choosing between clustering and annotation</h2>
</section><section id="sessioninfo" class="level2" data-number="6.5"><h2 data-number="6.5" class="anchored" data-anchor-id="sessioninfo">
<span class="header-section-number">6.5</span> sessionInfo</h2>
<div class="cell">
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Debian GNU/Linux 12 (bookworm)

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.21.so;  LAPACK version 3.11.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

time zone: Australia/Sydney
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] scuttle_1.16.0              simpleSeg_1.8.0            
 [3] MLmetrics_1.1.3             STexampleData_1.14.0       
 [5] SpatialExperiment_1.16.0    SingleCellExperiment_1.28.1
 [7] SummarizedExperiment_1.36.0 Biobase_2.66.0             
 [9] GenomicRanges_1.58.0        GenomeInfoDb_1.42.1        
[11] IRanges_2.40.0              S4Vectors_0.44.0           
[13] MatrixGenerics_1.18.0       matrixStats_1.4.1          
[15] ExperimentHub_2.14.0        AnnotationHub_3.14.0       
[17] BiocFileCache_2.14.0        dbplyr_2.5.0               
[19] BiocGenerics_0.52.0         scClassify_1.18.0          
[21] FuseSOM_1.8.0              

loaded via a namespace (and not attached):
  [1] segmented_2.1-3           fs_1.6.5                 
  [3] bitops_1.0-9              EBImage_4.48.0           
  [5] httr_1.4.7                hopach_2.66.0            
  [7] RColorBrewer_1.1-3        prabclus_2.3-4           
  [9] DataVisualizations_1.3.2  tools_4.4.1              
 [11] backports_1.5.0           utf8_1.2.4               
 [13] R6_2.5.1                  vegan_2.6-8              
 [15] HDF5Array_1.34.0          lazyeval_0.2.2           
 [17] mgcv_1.9-1                rhdf5filters_1.18.0      
 [19] permute_0.9-7             withr_3.0.2              
 [21] sp_2.1-4                  analogue_0.17-7          
 [23] gridExtra_2.3             cli_3.6.3                
 [25] profileModel_0.6.1        labeling_0.4.3           
 [27] diptest_0.77-1            robustbase_0.99-4-1      
 [29] brglm_0.7.2               nnls_1.6                 
 [31] spatstat.data_3.1-4       proxy_0.4-27             
 [33] systemfonts_1.1.0         yulab.utils_0.1.8        
 [35] svglite_2.1.3             limma_3.62.1             
 [37] rstudioapi_0.17.1         RSQLite_2.3.9            
 [39] generics_0.1.3            gridGraphics_0.5-1       
 [41] car_3.1-3                 dplyr_1.1.4              
 [43] Matrix_1.7-1              ggbeeswarm_0.7.2         
 [45] fansi_1.0.6               abind_1.4-8              
 [47] terra_1.7-83              lifecycle_1.0.4          
 [49] yaml_2.3.10               carData_3.0-5            
 [51] rhdf5_2.50.0              SparseArray_1.6.0        
 [53] grid_4.4.1                blob_1.2.4               
 [55] promises_1.3.2            crayon_1.5.3             
 [57] shinydashboard_0.7.2      lattice_0.22-6           
 [59] beachmat_2.22.0           KEGGREST_1.46.0          
 [61] magick_2.8.5              cytomapper_1.18.0        
 [63] pillar_1.9.0              knitr_1.49               
 [65] rjson_0.2.23              fpc_2.2-13               
 [67] codetools_0.2-20          glue_1.8.0               
 [69] FCPS_1.3.4                spatstat.univar_3.1-1    
 [71] data.table_1.16.4         vctrs_0.6.5              
 [73] png_0.1-8                 gtable_0.3.6             
 [75] kernlab_0.9-33            cachem_1.1.0             
 [77] xfun_0.49                 princurve_2.1.6          
 [79] S4Arrays_1.6.0            mime_0.12                
 [81] coop_0.6-3                tidygraph_1.3.1          
 [83] survival_3.7-0            pheatmap_1.0.12          
 [85] statmod_1.5.0             nlme_3.1-166             
 [87] bit64_4.5.2               filelock_1.0.3           
 [89] svgPanZoom_0.3.4          vipor_0.4.7              
 [91] colorspace_2.1-1          DBI_1.2.3                
 [93] nnet_7.3-19               raster_3.6-30            
 [95] mnormt_2.1.1              tidyselect_1.2.1         
 [97] proxyC_0.4.1              bit_4.5.0.1              
 [99] compiler_4.4.1            curl_6.0.1               
[101] DelayedArray_0.32.0       plotly_4.10.4            
[103] scales_1.3.0              DEoptimR_1.1-3-1         
[105] psych_2.4.6.26            rappdirs_0.3.3           
[107] tiff_0.1-12               stringr_1.5.1            
[109] digest_0.6.37             fftwtools_0.9-11         
[111] spatstat.utils_3.1-1      mixtools_2.0.0           
[113] rmarkdown_2.29            XVector_0.46.0           
[115] htmltools_0.5.8.1         pkgconfig_2.0.3          
[117] jpeg_0.1-10               Cepo_1.12.0              
[119] sparseMatrixStats_1.18.0  fastmap_1.2.0            
[121] rlang_1.1.4               htmlwidgets_1.6.4        
[123] UCSC.utils_1.2.0          shiny_1.9.1              
[125] DelayedMatrixStats_1.28.0 farver_2.1.2             
[127] jsonlite_1.8.9            BiocParallel_1.40.0      
[129] mclust_6.1.1              RCurl_1.98-1.16          
[131] magrittr_2.0.3            modeltools_0.2-23        
[133] Formula_1.2-5             GenomeInfoDbData_1.2.13  
[135] ggplotify_0.1.2           patchwork_1.3.0          
[137] Rhdf5lib_1.28.0           munsell_0.5.1            
[139] Rcpp_1.0.13-1             viridis_0.6.5            
[141] stringi_1.8.4             ggraph_2.2.1             
[143] zlibbioc_1.52.0           MASS_7.3-61              
[145] flexmix_2.3-19            plyr_1.8.9               
[147] parallel_4.4.1            ggrepel_0.9.6            
[149] deldir_2.0-4              Biostrings_2.74.0        
[151] graphlayouts_1.2.1        splines_4.4.1            
[153] locfit_1.5-9.10           igraph_2.1.1             
[155] ggpubr_0.6.0              fastcluster_1.2.6        
[157] spatstat.geom_3.3-4       ggsignif_0.6.4           
[159] reshape2_1.4.4            BiocVersion_3.20.0       
[161] evaluate_1.0.1            BiocManager_1.30.25      
[163] tweenr_2.0.3              httpuv_1.6.15            
[165] tidyr_1.3.1               purrr_1.0.2              
[167] polyclip_1.10-7           ggplot2_3.5.1            
[169] ggforce_0.4.2             broom_1.0.7              
[171] xtable_1.8-4              rstatix_0.7.2            
[173] later_1.4.1               class_7.3-22             
[175] viridisLite_0.4.2         minpack.lm_1.2-4         
[177] tibble_3.2.1              memoise_2.0.1            
[179] beeswarm_0.4.0            AnnotationDbi_1.68.0     
[181] cluster_2.1.6            </code></pre>
</div>
</div>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./02-quality_control.html" class="pagination-link" aria-label="Quality Control">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Quality Control</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./04a-cell_localisation.html" class="pagination-link" aria-label="Cell localisation">
        <span class="nav-page-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Cell localisation</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>